{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f6fa3073-a023-4d2e-9804-67a421f6fe19",
   "metadata": {},
   "source": [
    "# What to expect\n",
    "\n",
    "In this notebook, we will run some of the initial quality filtering and mapping with downsampled data from the *Schistosoma* dataset. If you need to revisit the presentation where we introduced RNAseq and the example dataset, you can find it on Learn, in the \"Workshop 1\" folder. You will also find the original paper where this dataset was published, and some review articles about RNA-Seq analysis methods. These are not compulsory reading, but may be of interest and are worth a look.\n",
    "\n",
    "In places we will provide the code to run an analysis step first, and then describe what it is doing. Some of these steps may take a few minutes, and this time can be spent reading and understanding the process as it runs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97fc3dd1-7baf-46cc-a3f2-0d712ab4728c",
   "metadata": {},
   "source": [
    "## The command line\n",
    "\n",
    "Whilst the first and second year courses focused on teaching coding in python, another key skill in biology is to run specialized existing software. Some of these can be installed as python modules, but many real-world tools are run \"on the command line\". This means that they run like an application or program, but the user types commands in a \"shell\" or \"terminal\" instead of clicking/swiping in an interactive window. Within the notebook environment, these commands can be run in 3 ways:\n",
    "\n",
    "* adding `%%bash` to the top of a cell\n",
    "* adding a `!` to the start of the command in a cell\n",
    "* (sometimes) if the command can only be interpreted in bash, jupyter sometimes doesn't need to be told. There is a really excellent (short) primer to the command line which can be found [here](https://cyverse-leptin-rna-seq-lesson-dev.readthedocs-hosted.com/en/latest/section-3.html) and gives a description of the most common commands.\n",
    "\n",
    "During this (and subsequent) workshops, we will combine code that is written on the command line (for which the full command will be provided) and questions which will require you to use the python coding you have already learnt."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a3323b6f-de79-4db7-b080-4a6e7242b925",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /Users/rmcolq/Work/apps/miniconda3/envs/pathbio3/lib/python3.10/site-packages (1.83)\n",
      "Requirement already satisfied: numpy in /Users/rmcolq/Work/apps/miniconda3/envs/pathbio3/lib/python3.10/site-packages (from biopython) (1.26.4)\n",
      "Retrieving notices: ...working... /Users/rmcolq/Work/apps/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'repo.anaconda.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings\n",
      "  warnings.warn(\n",
      "/Users/rmcolq/Work/apps/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'repo.anaconda.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings\n",
      "  warnings.warn(\n",
      "/Users/rmcolq/Work/apps/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'conda.anaconda.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings\n",
      "  warnings.warn(\n",
      "/Users/rmcolq/Work/apps/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'conda.anaconda.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings\n",
      "  warnings.warn(\n",
      "done\n",
      "/Users/rmcolq/Work/apps/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'conda.anaconda.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings\n",
      "  warnings.warn(\n",
      "Channels:\n",
      " - conda-forge\n",
      " - bioconda\n",
      " - defaults\n",
      " - anaconda\n",
      "Platform: osx-64\n",
      "Collecting package metadata (repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /Users/rmcolq/Work/apps/miniconda3/envs/pathbio3\n",
      "\n",
      "  added / updated specs:\n",
      "    - bioconda::fastqc\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    git-lfs-3.5.1              |       hecd8cb5_0         4.5 MB  anaconda\n",
      "    scikit-learn-1.5.1         |  py310h207f725_0        10.6 MB  anaconda\n",
      "    star-2.7.11b               |       hd576cc4_2         2.9 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        17.9 MB\n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  ca-certificates                       2024.6.2-h8857fd0_0 --> 2024.7.4-h8857fd0_0 \n",
      "  certifi                             2024.6.2-pyhd8ed1ab_0 --> 2024.7.4-pyhd8ed1ab_0 \n",
      "  git-lfs                                 2.13.3-hecd8cb5_0 --> 3.5.1-hecd8cb5_0 \n",
      "  openssl                                  3.3.1-h87427d6_1 --> 3.3.1-h87427d6_2 \n",
      "  scikit-learn                        1.3.0-py310h3ea8b11_0 --> 1.5.1-py310h207f725_0 \n",
      "  star                                   2.7.11b-h94ec1f5_1 --> 2.7.11b-hd576cc4_2 \n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "/Users/rmcolq/Work/apps/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'conda.anaconda.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings\n",
      "  warnings.warn(\n",
      "Channels:\n",
      " - conda-forge\n",
      " - bioconda\n",
      " - defaults\n",
      " - anaconda\n",
      "Platform: osx-64\n",
      "Collecting package metadata (repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Let's start by installing dependancies we will use\n",
    "! pip install biopython\n",
    "! conda install --yes --quiet bioconda::fastqc\n",
    "! conda install --yes --quiet bioconda::trim-galore conda-forge::pigz\n",
    "! conda install --yes --quiet bioconda::star"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fead3dd2-9c1d-4474-8ac8-0343b9ade22e",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "In the code above:\n",
    "\n",
    "`conda` - Package manager. We use it to install software\n",
    "\n",
    "`--yes` - confirm that we want to install all the dependencies\n",
    "\n",
    "`--quiet` - do not show extra output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9226ed08-811d-4aea-883a-2e98b58f494b",
   "metadata": {},
   "source": [
    "# The raw data\n",
    "\n",
    "The data is stored in `data/Schistosoma_mansoni`. Here we find four elements:\n",
    "\n",
    "1. `README` file - contains basic information about the data in this folder.\n",
    "2. `list_ids` file - contains the ids for the reads in this dataset. Each id corresponds to one sample (eg Schistosomula 3h post infection, replicate 1)\n",
    "3. `reference` folder - contains the reference genome and transcriptome\n",
    "4. `subsampled` folder - contains the raw data file for the subset of sequences we have taken for this workshop\n",
    "\n",
    "Let's see what we have in this examples dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "baabc7b3-2e41-4063-b02d-e9863170f442",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR022872\n",
      "ERR022873\n",
      "ERR022874\n",
      "ERR022875\n",
      "ERR022876\n",
      "ERR022877\n",
      "ERR022878\n",
      "ERR022879\n",
      "ERR022880\n",
      "ERR022881\n",
      "ERR022882\n",
      "ERR022883\n"
     ]
    }
   ],
   "source": [
    "# Print the list of ids for this example dataset\n",
    "! cat data/Schistosoma_mansoni/list_ids.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75f5b880-0fe9-43df-89e6-a6271e3a8f9a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "In the code above:\n",
    "\n",
    "`cat` - display the contents of a file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4635b11",
   "metadata": {},
   "source": [
    "# FASTQ structure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27949128-43ab-4fb0-a96d-6bbcbb11905a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Questions:\n",
    "1. The raw data files look like this `<accession>_<1|2>.fastq.gz`. What does the \".gz\" in the file name mean?\n",
    "2. Pick one of the files and open it. What does it look like? Define the [FASTQ format](https://en.wikipedia.org/wiki/FASTQ_format)?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdfde4ef-d522-4a6e-ba5b-2208eda89fce",
   "metadata": {},
   "source": [
    "<details>\n",
    "<summary><i>Hint</i></summary>\n",
    "\n",
    "1. \".gz\" is what is known as a file extension.\n",
    "2. To view the file you first need to uncompress it. In Data Exploration (week 1, class 2) you uncompressed \".gz\" files using `gunzip`. This time you want to save (`keep`) the uncompressed version of the file too. Once you have created the uncompressed version, you can open by double clicking in the file browser on the left hand side, or can download it and open it with a text editor.\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2ad181f7-718a-451a-becd-a1c5fe93dcc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq already exists -- do you wish to overwrite (y or n)? ^C\n"
     ]
    }
   ],
   "source": [
    "! gunzip --keep data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdd8eec5-a33c-43d8-86c0-3252afe5d33a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Answers:\n",
    "\n",
    "1. These files had been compressed using an algorithm called gzip\n",
    "2. This file contains read sequences. Each read is recorded in 4 lines:\n",
    "   - Field 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line).\n",
    "   - Field 2 is the raw sequence letters.\n",
    "   - Field 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again.\n",
    "   - Field 4 encodes the quality values for the sequence in Field 2, and must contain the same number of symbols as letters in the sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443f4988",
   "metadata": {},
   "source": [
    "For standard file formats for biological data (like FASTQ), there are often tools or librarys which make it easier to interract with them. These usually check that the file is formatted as expected. They also make it much easier to find information in the file without having to look up the exact structure of the file format yourself. A commonly used python module is [SeqIO](https://biopython.org/wiki/SeqIO) from the python library [biopython](https://biopython.org/).\n",
    "\n",
    "In your investigations on the FASTQ format you will have seen that scores are used to indicate how likely it is that a base reported in a sequencing read is in error. This is the [Phred score](https://learn.gencore.bio.nyu.edu/ngs-file-formats/quality-scores/). Now let's use the biopython library to investigate the sequence qualities in this file. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3789dd1e-d03a-4d74-88e5-1e439d5bb58b",
   "metadata": {},
   "source": [
    "For each read in a fastq file, SeqIO creates a [record](https://biopython.org/wiki/SeqRecord) object, with information about the read. The following code prints the record and Phred score for the first sequence in our FASTQ file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2554bc9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: ERR022872.17\n",
      "Name: ERR022872.17\n",
      "Description: ERR022872.17 IL9_3012:1:1:2:473/1\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('CNTCACCACAACCCAGCAGACCTTTACACTATGTATCTTCTTTAGNTNANANNA...AAC')\n",
      "[24, 4, 23, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 32, 32, 32, 32, 28, 29, 31, 32, 32, 32, 32, 32, 32, 33, 33, 33, 33, 32, 32, 32, 33, 33, 32, 32, 32, 32, 32, 33, 33, 32, 16, 4, 25, 4, 22, 4, 23, 4, 4, 24, 32, 33, 32, 32, 32, 33, 32, 32, 32, 33, 33, 32, 32, 32, 32, 34, 32, 31, 31, 31, 29, 32]\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "for index, record in enumerate(SeqIO.parse(\"data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq\", \"fastq\")):\n",
    "    print(record)\n",
    "    print(record.letter_annotations[\"phred_quality\"])\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9f672eb-3a72-468f-a405-04cfbaa3de54",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "In the code above:\n",
    "\n",
    "`SeqIO.parse(\"file name\",\"file format\")` - is a function that reads the file and gives the desired output\n",
    "\n",
    "`break` - is needed to stop the loop, otherwise it would do it for every single read in the file, which is not what we want just now\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07fc0545-c052-4a15-a8e9-2372ceca1a9e",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "3. Choose one of the FASTQ files and use a for loop to find out what the highest and lowest Phred scores are in your chosen file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4fe4113-150b-4199-bba2-ea9f63f60f64",
   "metadata": {},
   "source": [
    "<details>\n",
    "<summary><i>Hint</i></summary>\n",
    "\n",
    "For each record you need to:\n",
    " - find out the maximum and minimum Phred score in the record\n",
    " - update the variables\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c9619f83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Highest Phred score: 35\n",
      "Lowest Phred score: 4\n"
     ]
    }
   ],
   "source": [
    "#First define variables. All Phred scores will be bigger than 0 and less than 1000\n",
    "highest_phred=0\n",
    "lowest_phred=1000\n",
    "\n",
    "#Write the loop\n",
    "for record in SeqIO.parse(\"data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq\", \"fastq\"):\n",
    "    quality_scores = record.letter_annotations[\"phred_quality\"]\n",
    "    record_highest = max(quality_scores)\n",
    "    record_lowest = min(quality_scores)\n",
    "    highest_phred = max(highest_phred, record_highest)\n",
    "    lowest_phred = min(lowest_phred, record_lowest)\n",
    "    \n",
    "#Print the output\n",
    "print(f\"Highest Phred score: {highest_phred}\")\n",
    "print(f\"Lowest Phred score: {lowest_phred}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c8d82a5-eeef-4268-9429-47a747f1e701",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Answers:\n",
    "\n",
    "    #First define variables. All Phred scores will be bigger than 0 and less than 1000\n",
    "    highest_phred=0\n",
    "    lowest_phred=1000\n",
    "\n",
    "    #Write the loop\n",
    "    for record in SeqIO.parse(\"data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq\", \"fastq\"):\n",
    "        quality_scores = record.letter_annotations[\"phred_quality\"]\n",
    "        record_highest = max(quality_scores)\n",
    "        record_lowest = min(quality_scores)\n",
    "        highest_phred = max(highest_phred, record_highest)\n",
    "        lowest_phred = min(lowest_phred, record_lowest)\n",
    "    \n",
    "    #Print the output\n",
    "    print(f\"Highest Phred score: {highest_phred}\")\n",
    "    print(f\"Lowest Phred score: {lowest_phred}\")\n",
    "\n",
    "\n",
    "Highest Phred score: 35\n",
    "Lowest Phred score: 4\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eb1a206-ebcf-4efd-b12d-980a664cce67",
   "metadata": {},
   "source": [
    "# Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12ee69f8-e98f-4b0c-bc57-44b265e00be0",
   "metadata": {},
   "source": [
    "Hopefully in the previous exercises it has become clear that sequencing DNA is not error-free - each base of the sequence is read with a degree of uncertainty and error - and that our data definitely has some low quality reads (and parts of reads). Each sequencing machine/method results in different error profiles. To improve our analysis we first want to filter the lowest quality reads. One tool commonly used to profile the amounts of error is [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "90d4f6da-1ddf-4bbd-b368-9452a87bc17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the output directory\n",
    "! mkdir -p analysis/Schistosoma_mansoni/qc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4cc7c7e-510f-40e3-8e38-0b3c906f1128",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "In the code above:\n",
    "\n",
    "`mkdir` - command to create a new folder\n",
    "\n",
    "`-p` - flag to create nested folders"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6c319c0-72ce-4988-887f-cbd679158770",
   "metadata": {},
   "source": [
    "Now let's do the FASTQC for each file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3e667b0d-bf74-456f-aa2a-cf32d63c10e6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022872_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Approx 5% complete for ERR022872_1.fastq.gz\n",
      "Approx 10% complete for ERR022872_1.fastq.gz\n",
      "Approx 15% complete for ERR022872_1.fastq.gz\n",
      "Approx 20% complete for ERR022872_1.fastq.gz\n",
      "Approx 25% complete for ERR022872_1.fastq.gz\n",
      "Approx 30% complete for ERR022872_1.fastq.gz\n",
      "Approx 35% complete for ERR022872_1.fastq.gz\n",
      "Approx 40% complete for ERR022872_1.fastq.gz\n",
      "Approx 45% complete for ERR022872_1.fastq.gz\n",
      "Approx 50% complete for ERR022872_1.fastq.gz\n",
      "Approx 55% complete for ERR022872_1.fastq.gz\n",
      "Approx 60% complete for ERR022872_1.fastq.gz\n",
      "Approx 65% complete for ERR022872_1.fastq.gz\n",
      "Approx 70% complete for ERR022872_1.fastq.gz\n",
      "Approx 75% complete for ERR022872_1.fastq.gz\n",
      "Approx 80% complete for ERR022872_1.fastq.gz\n",
      "Approx 85% complete for ERR022872_1.fastq.gz\n",
      "Approx 90% complete for ERR022872_1.fastq.gz\n",
      "Approx 95% complete for ERR022872_1.fastq.gz\n",
      "Approx 100% complete for ERR022872_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022872_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022872_2.fastq.gz\n",
      "Approx 5% complete for ERR022872_2.fastq.gz\n",
      "Approx 10% complete for ERR022872_2.fastq.gz\n",
      "Approx 15% complete for ERR022872_2.fastq.gz\n",
      "Approx 20% complete for ERR022872_2.fastq.gz\n",
      "Approx 25% complete for ERR022872_2.fastq.gz\n",
      "Approx 30% complete for ERR022872_2.fastq.gz\n",
      "Approx 35% complete for ERR022872_2.fastq.gz\n",
      "Approx 40% complete for ERR022872_2.fastq.gz\n",
      "Approx 45% complete for ERR022872_2.fastq.gz\n",
      "Approx 50% complete for ERR022872_2.fastq.gz\n",
      "Approx 55% complete for ERR022872_2.fastq.gz\n",
      "Approx 60% complete for ERR022872_2.fastq.gz\n",
      "Approx 65% complete for ERR022872_2.fastq.gz\n",
      "Approx 70% complete for ERR022872_2.fastq.gz\n",
      "Approx 75% complete for ERR022872_2.fastq.gz\n",
      "Approx 80% complete for ERR022872_2.fastq.gz\n",
      "Approx 85% complete for ERR022872_2.fastq.gz\n",
      "Approx 90% complete for ERR022872_2.fastq.gz\n",
      "Approx 95% complete for ERR022872_2.fastq.gz\n",
      "Approx 100% complete for ERR022872_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022872_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022873_1.fastq.gz\n",
      "Approx 5% complete for ERR022873_1.fastq.gz\n",
      "Approx 10% complete for ERR022873_1.fastq.gz\n",
      "Approx 15% complete for ERR022873_1.fastq.gz\n",
      "Approx 20% complete for ERR022873_1.fastq.gz\n",
      "Approx 25% complete for ERR022873_1.fastq.gz\n",
      "Approx 30% complete for ERR022873_1.fastq.gz\n",
      "Approx 35% complete for ERR022873_1.fastq.gz\n",
      "Approx 40% complete for ERR022873_1.fastq.gz\n",
      "Approx 45% complete for ERR022873_1.fastq.gz\n",
      "Approx 50% complete for ERR022873_1.fastq.gz\n",
      "Approx 55% complete for ERR022873_1.fastq.gz\n",
      "Approx 60% complete for ERR022873_1.fastq.gz\n",
      "Approx 65% complete for ERR022873_1.fastq.gz\n",
      "Approx 70% complete for ERR022873_1.fastq.gz\n",
      "Approx 75% complete for ERR022873_1.fastq.gz\n",
      "Approx 80% complete for ERR022873_1.fastq.gz\n",
      "Approx 85% complete for ERR022873_1.fastq.gz\n",
      "Approx 90% complete for ERR022873_1.fastq.gz\n",
      "Approx 95% complete for ERR022873_1.fastq.gz\n",
      "Approx 100% complete for ERR022873_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022873_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022873_2.fastq.gz\n",
      "Approx 5% complete for ERR022873_2.fastq.gz\n",
      "Approx 10% complete for ERR022873_2.fastq.gz\n",
      "Approx 15% complete for ERR022873_2.fastq.gz\n",
      "Approx 20% complete for ERR022873_2.fastq.gz\n",
      "Approx 25% complete for ERR022873_2.fastq.gz\n",
      "Approx 30% complete for ERR022873_2.fastq.gz\n",
      "Approx 35% complete for ERR022873_2.fastq.gz\n",
      "Approx 40% complete for ERR022873_2.fastq.gz\n",
      "Approx 45% complete for ERR022873_2.fastq.gz\n",
      "Approx 50% complete for ERR022873_2.fastq.gz\n",
      "Approx 55% complete for ERR022873_2.fastq.gz\n",
      "Approx 60% complete for ERR022873_2.fastq.gz\n",
      "Approx 65% complete for ERR022873_2.fastq.gz\n",
      "Approx 70% complete for ERR022873_2.fastq.gz\n",
      "Approx 75% complete for ERR022873_2.fastq.gz\n",
      "Approx 80% complete for ERR022873_2.fastq.gz\n",
      "Approx 85% complete for ERR022873_2.fastq.gz\n",
      "Approx 90% complete for ERR022873_2.fastq.gz\n",
      "Approx 95% complete for ERR022873_2.fastq.gz\n",
      "Approx 100% complete for ERR022873_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022873_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022874_1.fastq.gz\n",
      "Approx 5% complete for ERR022874_1.fastq.gz\n",
      "Approx 10% complete for ERR022874_1.fastq.gz\n",
      "Approx 15% complete for ERR022874_1.fastq.gz\n",
      "Approx 20% complete for ERR022874_1.fastq.gz\n",
      "Approx 25% complete for ERR022874_1.fastq.gz\n",
      "Approx 30% complete for ERR022874_1.fastq.gz\n",
      "Approx 35% complete for ERR022874_1.fastq.gz\n",
      "Approx 40% complete for ERR022874_1.fastq.gz\n",
      "Approx 45% complete for ERR022874_1.fastq.gz\n",
      "Approx 50% complete for ERR022874_1.fastq.gz\n",
      "Approx 55% complete for ERR022874_1.fastq.gz\n",
      "Approx 60% complete for ERR022874_1.fastq.gz\n",
      "Approx 65% complete for ERR022874_1.fastq.gz\n",
      "Approx 70% complete for ERR022874_1.fastq.gz\n",
      "Approx 75% complete for ERR022874_1.fastq.gz\n",
      "Approx 80% complete for ERR022874_1.fastq.gz\n",
      "Approx 85% complete for ERR022874_1.fastq.gz\n",
      "Approx 90% complete for ERR022874_1.fastq.gz\n",
      "Approx 95% complete for ERR022874_1.fastq.gz\n",
      "Approx 100% complete for ERR022874_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022874_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022874_2.fastq.gz\n",
      "Approx 5% complete for ERR022874_2.fastq.gz\n",
      "Approx 10% complete for ERR022874_2.fastq.gz\n",
      "Approx 15% complete for ERR022874_2.fastq.gz\n",
      "Approx 20% complete for ERR022874_2.fastq.gz\n",
      "Approx 25% complete for ERR022874_2.fastq.gz\n",
      "Approx 30% complete for ERR022874_2.fastq.gz\n",
      "Approx 35% complete for ERR022874_2.fastq.gz\n",
      "Approx 40% complete for ERR022874_2.fastq.gz\n",
      "Approx 45% complete for ERR022874_2.fastq.gz\n",
      "Approx 50% complete for ERR022874_2.fastq.gz\n",
      "Approx 55% complete for ERR022874_2.fastq.gz\n",
      "Approx 60% complete for ERR022874_2.fastq.gz\n",
      "Approx 65% complete for ERR022874_2.fastq.gz\n",
      "Approx 70% complete for ERR022874_2.fastq.gz\n",
      "Approx 75% complete for ERR022874_2.fastq.gz\n",
      "Approx 80% complete for ERR022874_2.fastq.gz\n",
      "Approx 85% complete for ERR022874_2.fastq.gz\n",
      "Approx 90% complete for ERR022874_2.fastq.gz\n",
      "Approx 95% complete for ERR022874_2.fastq.gz\n",
      "Approx 100% complete for ERR022874_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022874_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022875_1.fastq.gz\n",
      "Approx 5% complete for ERR022875_1.fastq.gz\n",
      "Approx 10% complete for ERR022875_1.fastq.gz\n",
      "Approx 15% complete for ERR022875_1.fastq.gz\n",
      "Approx 20% complete for ERR022875_1.fastq.gz\n",
      "Approx 25% complete for ERR022875_1.fastq.gz\n",
      "Approx 30% complete for ERR022875_1.fastq.gz\n",
      "Approx 35% complete for ERR022875_1.fastq.gz\n",
      "Approx 40% complete for ERR022875_1.fastq.gz\n",
      "Approx 45% complete for ERR022875_1.fastq.gz\n",
      "Approx 50% complete for ERR022875_1.fastq.gz\n",
      "Approx 55% complete for ERR022875_1.fastq.gz\n",
      "Approx 60% complete for ERR022875_1.fastq.gz\n",
      "Approx 65% complete for ERR022875_1.fastq.gz\n",
      "Approx 70% complete for ERR022875_1.fastq.gz\n",
      "Approx 75% complete for ERR022875_1.fastq.gz\n",
      "Approx 80% complete for ERR022875_1.fastq.gz\n",
      "Approx 85% complete for ERR022875_1.fastq.gz\n",
      "Approx 90% complete for ERR022875_1.fastq.gz\n",
      "Approx 95% complete for ERR022875_1.fastq.gz\n",
      "Approx 100% complete for ERR022875_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022875_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022875_2.fastq.gz\n",
      "Approx 5% complete for ERR022875_2.fastq.gz\n",
      "Approx 10% complete for ERR022875_2.fastq.gz\n",
      "Approx 15% complete for ERR022875_2.fastq.gz\n",
      "Approx 20% complete for ERR022875_2.fastq.gz\n",
      "Approx 25% complete for ERR022875_2.fastq.gz\n",
      "Approx 30% complete for ERR022875_2.fastq.gz\n",
      "Approx 35% complete for ERR022875_2.fastq.gz\n",
      "Approx 40% complete for ERR022875_2.fastq.gz\n",
      "Approx 45% complete for ERR022875_2.fastq.gz\n",
      "Approx 50% complete for ERR022875_2.fastq.gz\n",
      "Approx 55% complete for ERR022875_2.fastq.gz\n",
      "Approx 60% complete for ERR022875_2.fastq.gz\n",
      "Approx 65% complete for ERR022875_2.fastq.gz\n",
      "Approx 70% complete for ERR022875_2.fastq.gz\n",
      "Approx 75% complete for ERR022875_2.fastq.gz\n",
      "Approx 80% complete for ERR022875_2.fastq.gz\n",
      "Approx 85% complete for ERR022875_2.fastq.gz\n",
      "Approx 90% complete for ERR022875_2.fastq.gz\n",
      "Approx 95% complete for ERR022875_2.fastq.gz\n",
      "Approx 100% complete for ERR022875_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022875_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022876_1.fastq.gz\n",
      "Approx 5% complete for ERR022876_1.fastq.gz\n",
      "Approx 10% complete for ERR022876_1.fastq.gz\n",
      "Approx 15% complete for ERR022876_1.fastq.gz\n",
      "Approx 20% complete for ERR022876_1.fastq.gz\n",
      "Approx 25% complete for ERR022876_1.fastq.gz\n",
      "Approx 30% complete for ERR022876_1.fastq.gz\n",
      "Approx 35% complete for ERR022876_1.fastq.gz\n",
      "Approx 40% complete for ERR022876_1.fastq.gz\n",
      "Approx 45% complete for ERR022876_1.fastq.gz\n",
      "Approx 50% complete for ERR022876_1.fastq.gz\n",
      "Approx 55% complete for ERR022876_1.fastq.gz\n",
      "Approx 60% complete for ERR022876_1.fastq.gz\n",
      "Approx 65% complete for ERR022876_1.fastq.gz\n",
      "Approx 70% complete for ERR022876_1.fastq.gz\n",
      "Approx 75% complete for ERR022876_1.fastq.gz\n",
      "Approx 80% complete for ERR022876_1.fastq.gz\n",
      "Approx 85% complete for ERR022876_1.fastq.gz\n",
      "Approx 90% complete for ERR022876_1.fastq.gz\n",
      "Approx 95% complete for ERR022876_1.fastq.gz\n",
      "Approx 100% complete for ERR022876_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022876_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022876_2.fastq.gz\n",
      "Approx 5% complete for ERR022876_2.fastq.gz\n",
      "Approx 10% complete for ERR022876_2.fastq.gz\n",
      "Approx 15% complete for ERR022876_2.fastq.gz\n",
      "Approx 20% complete for ERR022876_2.fastq.gz\n",
      "Approx 25% complete for ERR022876_2.fastq.gz\n",
      "Approx 30% complete for ERR022876_2.fastq.gz\n",
      "Approx 35% complete for ERR022876_2.fastq.gz\n",
      "Approx 40% complete for ERR022876_2.fastq.gz\n",
      "Approx 45% complete for ERR022876_2.fastq.gz\n",
      "Approx 50% complete for ERR022876_2.fastq.gz\n",
      "Approx 55% complete for ERR022876_2.fastq.gz\n",
      "Approx 60% complete for ERR022876_2.fastq.gz\n",
      "Approx 65% complete for ERR022876_2.fastq.gz\n",
      "Approx 70% complete for ERR022876_2.fastq.gz\n",
      "Approx 75% complete for ERR022876_2.fastq.gz\n",
      "Approx 80% complete for ERR022876_2.fastq.gz\n",
      "Approx 85% complete for ERR022876_2.fastq.gz\n",
      "Approx 90% complete for ERR022876_2.fastq.gz\n",
      "Approx 95% complete for ERR022876_2.fastq.gz\n",
      "Approx 100% complete for ERR022876_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022876_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022877_1.fastq.gz\n",
      "Approx 5% complete for ERR022877_1.fastq.gz\n",
      "Approx 10% complete for ERR022877_1.fastq.gz\n",
      "Approx 15% complete for ERR022877_1.fastq.gz\n",
      "Approx 20% complete for ERR022877_1.fastq.gz\n",
      "Approx 25% complete for ERR022877_1.fastq.gz\n",
      "Approx 30% complete for ERR022877_1.fastq.gz\n",
      "Approx 35% complete for ERR022877_1.fastq.gz\n",
      "Approx 40% complete for ERR022877_1.fastq.gz\n",
      "Approx 45% complete for ERR022877_1.fastq.gz\n",
      "Approx 50% complete for ERR022877_1.fastq.gz\n",
      "Approx 55% complete for ERR022877_1.fastq.gz\n",
      "Approx 60% complete for ERR022877_1.fastq.gz\n",
      "Approx 65% complete for ERR022877_1.fastq.gz\n",
      "Approx 70% complete for ERR022877_1.fastq.gz\n",
      "Approx 75% complete for ERR022877_1.fastq.gz\n",
      "Approx 80% complete for ERR022877_1.fastq.gz\n",
      "Approx 85% complete for ERR022877_1.fastq.gz\n",
      "Approx 90% complete for ERR022877_1.fastq.gz\n",
      "Approx 95% complete for ERR022877_1.fastq.gz\n",
      "Approx 100% complete for ERR022877_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022877_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022877_2.fastq.gz\n",
      "Approx 5% complete for ERR022877_2.fastq.gz\n",
      "Approx 10% complete for ERR022877_2.fastq.gz\n",
      "Approx 15% complete for ERR022877_2.fastq.gz\n",
      "Approx 20% complete for ERR022877_2.fastq.gz\n",
      "Approx 25% complete for ERR022877_2.fastq.gz\n",
      "Approx 30% complete for ERR022877_2.fastq.gz\n",
      "Approx 35% complete for ERR022877_2.fastq.gz\n",
      "Approx 40% complete for ERR022877_2.fastq.gz\n",
      "Approx 45% complete for ERR022877_2.fastq.gz\n",
      "Approx 50% complete for ERR022877_2.fastq.gz\n",
      "Approx 55% complete for ERR022877_2.fastq.gz\n",
      "Approx 60% complete for ERR022877_2.fastq.gz\n",
      "Approx 65% complete for ERR022877_2.fastq.gz\n",
      "Approx 70% complete for ERR022877_2.fastq.gz\n",
      "Approx 75% complete for ERR022877_2.fastq.gz\n",
      "Approx 80% complete for ERR022877_2.fastq.gz\n",
      "Approx 85% complete for ERR022877_2.fastq.gz\n",
      "Approx 90% complete for ERR022877_2.fastq.gz\n",
      "Approx 95% complete for ERR022877_2.fastq.gz\n",
      "Approx 100% complete for ERR022877_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022877_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022878_1.fastq.gz\n",
      "Approx 5% complete for ERR022878_1.fastq.gz\n",
      "Approx 10% complete for ERR022878_1.fastq.gz\n",
      "Approx 15% complete for ERR022878_1.fastq.gz\n",
      "Approx 20% complete for ERR022878_1.fastq.gz\n",
      "Approx 25% complete for ERR022878_1.fastq.gz\n",
      "Approx 30% complete for ERR022878_1.fastq.gz\n",
      "Approx 35% complete for ERR022878_1.fastq.gz\n",
      "Approx 40% complete for ERR022878_1.fastq.gz\n",
      "Approx 45% complete for ERR022878_1.fastq.gz\n",
      "Approx 50% complete for ERR022878_1.fastq.gz\n",
      "Approx 55% complete for ERR022878_1.fastq.gz\n",
      "Approx 60% complete for ERR022878_1.fastq.gz\n",
      "Approx 65% complete for ERR022878_1.fastq.gz\n",
      "Approx 70% complete for ERR022878_1.fastq.gz\n",
      "Approx 75% complete for ERR022878_1.fastq.gz\n",
      "Approx 80% complete for ERR022878_1.fastq.gz\n",
      "Approx 85% complete for ERR022878_1.fastq.gz\n",
      "Approx 90% complete for ERR022878_1.fastq.gz\n",
      "Approx 95% complete for ERR022878_1.fastq.gz\n",
      "Approx 100% complete for ERR022878_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022878_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022878_2.fastq.gz\n",
      "Approx 5% complete for ERR022878_2.fastq.gz\n",
      "Approx 10% complete for ERR022878_2.fastq.gz\n",
      "Approx 15% complete for ERR022878_2.fastq.gz\n",
      "Approx 20% complete for ERR022878_2.fastq.gz\n",
      "Approx 25% complete for ERR022878_2.fastq.gz\n",
      "Approx 30% complete for ERR022878_2.fastq.gz\n",
      "Approx 35% complete for ERR022878_2.fastq.gz\n",
      "Approx 40% complete for ERR022878_2.fastq.gz\n",
      "Approx 45% complete for ERR022878_2.fastq.gz\n",
      "Approx 50% complete for ERR022878_2.fastq.gz\n",
      "Approx 55% complete for ERR022878_2.fastq.gz\n",
      "Approx 60% complete for ERR022878_2.fastq.gz\n",
      "Approx 65% complete for ERR022878_2.fastq.gz\n",
      "Approx 70% complete for ERR022878_2.fastq.gz\n",
      "Approx 75% complete for ERR022878_2.fastq.gz\n",
      "Approx 80% complete for ERR022878_2.fastq.gz\n",
      "Approx 85% complete for ERR022878_2.fastq.gz\n",
      "Approx 90% complete for ERR022878_2.fastq.gz\n",
      "Approx 95% complete for ERR022878_2.fastq.gz\n",
      "Approx 100% complete for ERR022878_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022878_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022879_1.fastq.gz\n",
      "Approx 5% complete for ERR022879_1.fastq.gz\n",
      "Approx 10% complete for ERR022879_1.fastq.gz\n",
      "Approx 15% complete for ERR022879_1.fastq.gz\n",
      "Approx 20% complete for ERR022879_1.fastq.gz\n",
      "Approx 25% complete for ERR022879_1.fastq.gz\n",
      "Approx 30% complete for ERR022879_1.fastq.gz\n",
      "Approx 35% complete for ERR022879_1.fastq.gz\n",
      "Approx 40% complete for ERR022879_1.fastq.gz\n",
      "Approx 45% complete for ERR022879_1.fastq.gz\n",
      "Approx 50% complete for ERR022879_1.fastq.gz\n",
      "Approx 55% complete for ERR022879_1.fastq.gz\n",
      "Approx 60% complete for ERR022879_1.fastq.gz\n",
      "Approx 65% complete for ERR022879_1.fastq.gz\n",
      "Approx 70% complete for ERR022879_1.fastq.gz\n",
      "Approx 75% complete for ERR022879_1.fastq.gz\n",
      "Approx 80% complete for ERR022879_1.fastq.gz\n",
      "Approx 85% complete for ERR022879_1.fastq.gz\n",
      "Approx 90% complete for ERR022879_1.fastq.gz\n",
      "Approx 95% complete for ERR022879_1.fastq.gz\n",
      "Approx 100% complete for ERR022879_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022879_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022879_2.fastq.gz\n",
      "Approx 5% complete for ERR022879_2.fastq.gz\n",
      "Approx 10% complete for ERR022879_2.fastq.gz\n",
      "Approx 15% complete for ERR022879_2.fastq.gz\n",
      "Approx 20% complete for ERR022879_2.fastq.gz\n",
      "Approx 25% complete for ERR022879_2.fastq.gz\n",
      "Approx 30% complete for ERR022879_2.fastq.gz\n",
      "Approx 35% complete for ERR022879_2.fastq.gz\n",
      "Approx 40% complete for ERR022879_2.fastq.gz\n",
      "Approx 45% complete for ERR022879_2.fastq.gz\n",
      "Approx 50% complete for ERR022879_2.fastq.gz\n",
      "Approx 55% complete for ERR022879_2.fastq.gz\n",
      "Approx 60% complete for ERR022879_2.fastq.gz\n",
      "Approx 65% complete for ERR022879_2.fastq.gz\n",
      "Approx 70% complete for ERR022879_2.fastq.gz\n",
      "Approx 75% complete for ERR022879_2.fastq.gz\n",
      "Approx 80% complete for ERR022879_2.fastq.gz\n",
      "Approx 85% complete for ERR022879_2.fastq.gz\n",
      "Approx 90% complete for ERR022879_2.fastq.gz\n",
      "Approx 95% complete for ERR022879_2.fastq.gz\n",
      "Approx 100% complete for ERR022879_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022879_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022880_1.fastq.gz\n",
      "Approx 5% complete for ERR022880_1.fastq.gz\n",
      "Approx 10% complete for ERR022880_1.fastq.gz\n",
      "Approx 15% complete for ERR022880_1.fastq.gz\n",
      "Approx 20% complete for ERR022880_1.fastq.gz\n",
      "Approx 25% complete for ERR022880_1.fastq.gz\n",
      "Approx 30% complete for ERR022880_1.fastq.gz\n",
      "Approx 35% complete for ERR022880_1.fastq.gz\n",
      "Approx 40% complete for ERR022880_1.fastq.gz\n",
      "Approx 45% complete for ERR022880_1.fastq.gz\n",
      "Approx 50% complete for ERR022880_1.fastq.gz\n",
      "Approx 55% complete for ERR022880_1.fastq.gz\n",
      "Approx 60% complete for ERR022880_1.fastq.gz\n",
      "Approx 65% complete for ERR022880_1.fastq.gz\n",
      "Approx 70% complete for ERR022880_1.fastq.gz\n",
      "Approx 75% complete for ERR022880_1.fastq.gz\n",
      "Approx 80% complete for ERR022880_1.fastq.gz\n",
      "Approx 85% complete for ERR022880_1.fastq.gz\n",
      "Approx 90% complete for ERR022880_1.fastq.gz\n",
      "Approx 95% complete for ERR022880_1.fastq.gz\n",
      "Approx 100% complete for ERR022880_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022880_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022880_2.fastq.gz\n",
      "Approx 5% complete for ERR022880_2.fastq.gz\n",
      "Approx 10% complete for ERR022880_2.fastq.gz\n",
      "Approx 15% complete for ERR022880_2.fastq.gz\n",
      "Approx 20% complete for ERR022880_2.fastq.gz\n",
      "Approx 25% complete for ERR022880_2.fastq.gz\n",
      "Approx 30% complete for ERR022880_2.fastq.gz\n",
      "Approx 35% complete for ERR022880_2.fastq.gz\n",
      "Approx 40% complete for ERR022880_2.fastq.gz\n",
      "Approx 45% complete for ERR022880_2.fastq.gz\n",
      "Approx 50% complete for ERR022880_2.fastq.gz\n",
      "Approx 55% complete for ERR022880_2.fastq.gz\n",
      "Approx 60% complete for ERR022880_2.fastq.gz\n",
      "Approx 65% complete for ERR022880_2.fastq.gz\n",
      "Approx 70% complete for ERR022880_2.fastq.gz\n",
      "Approx 75% complete for ERR022880_2.fastq.gz\n",
      "Approx 80% complete for ERR022880_2.fastq.gz\n",
      "Approx 85% complete for ERR022880_2.fastq.gz\n",
      "Approx 90% complete for ERR022880_2.fastq.gz\n",
      "Approx 95% complete for ERR022880_2.fastq.gz\n",
      "Approx 100% complete for ERR022880_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022880_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022881_1.fastq.gz\n",
      "Approx 5% complete for ERR022881_1.fastq.gz\n",
      "Approx 10% complete for ERR022881_1.fastq.gz\n",
      "Approx 15% complete for ERR022881_1.fastq.gz\n",
      "Approx 20% complete for ERR022881_1.fastq.gz\n",
      "Approx 25% complete for ERR022881_1.fastq.gz\n",
      "Approx 30% complete for ERR022881_1.fastq.gz\n",
      "Approx 35% complete for ERR022881_1.fastq.gz\n",
      "Approx 40% complete for ERR022881_1.fastq.gz\n",
      "Approx 45% complete for ERR022881_1.fastq.gz\n",
      "Approx 50% complete for ERR022881_1.fastq.gz\n",
      "Approx 55% complete for ERR022881_1.fastq.gz\n",
      "Approx 60% complete for ERR022881_1.fastq.gz\n",
      "Approx 65% complete for ERR022881_1.fastq.gz\n",
      "Approx 70% complete for ERR022881_1.fastq.gz\n",
      "Approx 75% complete for ERR022881_1.fastq.gz\n",
      "Approx 80% complete for ERR022881_1.fastq.gz\n",
      "Approx 85% complete for ERR022881_1.fastq.gz\n",
      "Approx 90% complete for ERR022881_1.fastq.gz\n",
      "Approx 95% complete for ERR022881_1.fastq.gz\n",
      "Approx 100% complete for ERR022881_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022881_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022881_2.fastq.gz\n",
      "Approx 5% complete for ERR022881_2.fastq.gz\n",
      "Approx 10% complete for ERR022881_2.fastq.gz\n",
      "Approx 15% complete for ERR022881_2.fastq.gz\n",
      "Approx 20% complete for ERR022881_2.fastq.gz\n",
      "Approx 25% complete for ERR022881_2.fastq.gz\n",
      "Approx 30% complete for ERR022881_2.fastq.gz\n",
      "Approx 35% complete for ERR022881_2.fastq.gz\n",
      "Approx 40% complete for ERR022881_2.fastq.gz\n",
      "Approx 45% complete for ERR022881_2.fastq.gz\n",
      "Approx 50% complete for ERR022881_2.fastq.gz\n",
      "Approx 55% complete for ERR022881_2.fastq.gz\n",
      "Approx 60% complete for ERR022881_2.fastq.gz\n",
      "Approx 65% complete for ERR022881_2.fastq.gz\n",
      "Approx 70% complete for ERR022881_2.fastq.gz\n",
      "Approx 75% complete for ERR022881_2.fastq.gz\n",
      "Approx 80% complete for ERR022881_2.fastq.gz\n",
      "Approx 85% complete for ERR022881_2.fastq.gz\n",
      "Approx 90% complete for ERR022881_2.fastq.gz\n",
      "Approx 95% complete for ERR022881_2.fastq.gz\n",
      "Approx 100% complete for ERR022881_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022881_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022882_1.fastq.gz\n",
      "Approx 5% complete for ERR022882_1.fastq.gz\n",
      "Approx 10% complete for ERR022882_1.fastq.gz\n",
      "Approx 15% complete for ERR022882_1.fastq.gz\n",
      "Approx 20% complete for ERR022882_1.fastq.gz\n",
      "Approx 25% complete for ERR022882_1.fastq.gz\n",
      "Approx 30% complete for ERR022882_1.fastq.gz\n",
      "Approx 35% complete for ERR022882_1.fastq.gz\n",
      "Approx 40% complete for ERR022882_1.fastq.gz\n",
      "Approx 45% complete for ERR022882_1.fastq.gz\n",
      "Approx 50% complete for ERR022882_1.fastq.gz\n",
      "Approx 55% complete for ERR022882_1.fastq.gz\n",
      "Approx 60% complete for ERR022882_1.fastq.gz\n",
      "Approx 65% complete for ERR022882_1.fastq.gz\n",
      "Approx 70% complete for ERR022882_1.fastq.gz\n",
      "Approx 75% complete for ERR022882_1.fastq.gz\n",
      "Approx 80% complete for ERR022882_1.fastq.gz\n",
      "Approx 85% complete for ERR022882_1.fastq.gz\n",
      "Approx 90% complete for ERR022882_1.fastq.gz\n",
      "Approx 95% complete for ERR022882_1.fastq.gz\n",
      "Approx 100% complete for ERR022882_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022882_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022882_2.fastq.gz\n",
      "Approx 5% complete for ERR022882_2.fastq.gz\n",
      "Approx 10% complete for ERR022882_2.fastq.gz\n",
      "Approx 15% complete for ERR022882_2.fastq.gz\n",
      "Approx 20% complete for ERR022882_2.fastq.gz\n",
      "Approx 25% complete for ERR022882_2.fastq.gz\n",
      "Approx 30% complete for ERR022882_2.fastq.gz\n",
      "Approx 35% complete for ERR022882_2.fastq.gz\n",
      "Approx 40% complete for ERR022882_2.fastq.gz\n",
      "Approx 45% complete for ERR022882_2.fastq.gz\n",
      "Approx 50% complete for ERR022882_2.fastq.gz\n",
      "Approx 55% complete for ERR022882_2.fastq.gz\n",
      "Approx 60% complete for ERR022882_2.fastq.gz\n",
      "Approx 65% complete for ERR022882_2.fastq.gz\n",
      "Approx 70% complete for ERR022882_2.fastq.gz\n",
      "Approx 75% complete for ERR022882_2.fastq.gz\n",
      "Approx 80% complete for ERR022882_2.fastq.gz\n",
      "Approx 85% complete for ERR022882_2.fastq.gz\n",
      "Approx 90% complete for ERR022882_2.fastq.gz\n",
      "Approx 95% complete for ERR022882_2.fastq.gz\n",
      "Approx 100% complete for ERR022882_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022882_2.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022883_1.fastq.gz\n",
      "Approx 5% complete for ERR022883_1.fastq.gz\n",
      "Approx 10% complete for ERR022883_1.fastq.gz\n",
      "Approx 15% complete for ERR022883_1.fastq.gz\n",
      "Approx 20% complete for ERR022883_1.fastq.gz\n",
      "Approx 25% complete for ERR022883_1.fastq.gz\n",
      "Approx 30% complete for ERR022883_1.fastq.gz\n",
      "Approx 35% complete for ERR022883_1.fastq.gz\n",
      "Approx 40% complete for ERR022883_1.fastq.gz\n",
      "Approx 45% complete for ERR022883_1.fastq.gz\n",
      "Approx 50% complete for ERR022883_1.fastq.gz\n",
      "Approx 55% complete for ERR022883_1.fastq.gz\n",
      "Approx 60% complete for ERR022883_1.fastq.gz\n",
      "Approx 65% complete for ERR022883_1.fastq.gz\n",
      "Approx 70% complete for ERR022883_1.fastq.gz\n",
      "Approx 75% complete for ERR022883_1.fastq.gz\n",
      "Approx 80% complete for ERR022883_1.fastq.gz\n",
      "Approx 85% complete for ERR022883_1.fastq.gz\n",
      "Approx 90% complete for ERR022883_1.fastq.gz\n",
      "Approx 95% complete for ERR022883_1.fastq.gz\n",
      "Approx 100% complete for ERR022883_1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022883_1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022883_2.fastq.gz\n",
      "Approx 5% complete for ERR022883_2.fastq.gz\n",
      "Approx 10% complete for ERR022883_2.fastq.gz\n",
      "Approx 15% complete for ERR022883_2.fastq.gz\n",
      "Approx 20% complete for ERR022883_2.fastq.gz\n",
      "Approx 25% complete for ERR022883_2.fastq.gz\n",
      "Approx 30% complete for ERR022883_2.fastq.gz\n",
      "Approx 35% complete for ERR022883_2.fastq.gz\n",
      "Approx 40% complete for ERR022883_2.fastq.gz\n",
      "Approx 45% complete for ERR022883_2.fastq.gz\n",
      "Approx 50% complete for ERR022883_2.fastq.gz\n",
      "Approx 55% complete for ERR022883_2.fastq.gz\n",
      "Approx 60% complete for ERR022883_2.fastq.gz\n",
      "Approx 65% complete for ERR022883_2.fastq.gz\n",
      "Approx 70% complete for ERR022883_2.fastq.gz\n",
      "Approx 75% complete for ERR022883_2.fastq.gz\n",
      "Approx 80% complete for ERR022883_2.fastq.gz\n",
      "Approx 85% complete for ERR022883_2.fastq.gz\n",
      "Approx 90% complete for ERR022883_2.fastq.gz\n",
      "Approx 95% complete for ERR022883_2.fastq.gz\n",
      "Approx 100% complete for ERR022883_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022883_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for accession in $(cat data/Schistosoma_mansoni/list_ids.txt)\n",
    "do \n",
    "    fastqc data/Schistosoma_mansoni/subsampled/$accession*.fastq.gz --noextract -o analysis/Schistosoma_mansoni/qc\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34fbc054-2251-480c-8e81-13b08236cfa7",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "In the code above:\n",
    "\n",
    "`for X in Y; do ... done` - this the structure for a loop in bash; we are asking that for each X (in this case accession n) in Y (in this case the list of ids), the program does something (in this case fastqc)\n",
    "\n",
    "`--noextract` - Do not uncompress the output file after creating it\n",
    "\n",
    "`-o` - create all output files in the output directory specified next"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbaf1b55-86fd-4e61-b4f6-c06aa5203e93",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Open the output directory. You will see that two new files have been created for each read. Open one of the html files and have a look. [Here](https://hbctraining.github.io/Training-modules/planning_successful_rnaseq/lessons/QC_raw_data.html) you can find guidance on what each graph means. \n",
    "\n",
    "Questions:\n",
    "\n",
    "4. Is there a pattern to where the errors occur in these reads?\n",
    "5. Are there any overrepresented sequences? What are they from?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "404255c3-6f21-4f60-959c-ec1b6445f7be",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    \n",
    "Answers:\n",
    "\n",
    "4. The error rate increases as you get closer to the end of the read\n",
    "5. Several sequences from Illumina Paired End PCR Primers and Illumina Universal Adaptors are overrepresented. These are short regions of sequence which were attached to the reads during the sequencing process."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ff8b54f-e244-4be9-bbd9-01c9398b844a",
   "metadata": {},
   "source": [
    "# Trimming"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd6204fe-0750-4eb1-83b1-6b999c0e33df",
   "metadata": {},
   "source": [
    "When we generate millions of reads in a sequencing experiment, we are able to average multiple observations at the same location. However if there are portions of the reads which are low quality these may still affect the average and so we want to remove these regions. As you will have seen in the previous section, the adaptor sequences used to generate the DNA library may also be over-represented in the reads and could cause contamination.\n",
    "\n",
    "The next step in an RNA-Seq analysis is therefore to trim poor quality regions and adaptor sequences. [TrimGalore](https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md) is a popular software for trimming sequencing reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1a9eb730-ed9c-47bb-9baa-02a2d9d14f73",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022872<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t66884\tAGATCGGAAGAGC\t1000000\t6.69\n",
      "Nextera\t2\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t2\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 66884). Second best hit was Nextera (count: 2)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022872_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 28.674 s (28.674 s/read; 2.09 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   390,904 (39.1%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               3,653,748 bp (4.8%)\n",
      "Total written (filtered):     69,752,754 bp (91.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 390904 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 27.7%\n",
      "  C: 27.8%\n",
      "  G: 24.6%\n",
      "  T: 18.9%\n",
      "  none/other: 1.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t226480\t250000.0\t0\t226480\n",
      "2\t61580\t62500.0\t0\t61580\n",
      "3\t16371\t15625.0\t0\t16371\n",
      "4\t5622\t3906.2\t0\t5622\n",
      "5\t2022\t976.6\t0\t2022\n",
      "6\t2517\t244.1\t0\t2517\n",
      "7\t1817\t61.0\t0\t1817\n",
      "8\t1913\t15.3\t0\t1913\n",
      "9\t1302\t3.8\t0\t1198 104\n",
      "10\t2078\t1.0\t1\t1968 110\n",
      "11\t1727\t0.2\t1\t1669 58\n",
      "12\t2451\t0.1\t1\t2377 74\n",
      "13\t1206\t0.0\t1\t1160 46\n",
      "14\t2495\t0.0\t1\t2372 123\n",
      "15\t4226\t0.0\t1\t4035 191\n",
      "16\t1930\t0.0\t1\t1843 87\n",
      "17\t1823\t0.0\t1\t1770 53\n",
      "18\t1165\t0.0\t1\t1132 33\n",
      "19\t1759\t0.0\t1\t1698 61\n",
      "20\t2055\t0.0\t1\t1985 70\n",
      "21\t1141\t0.0\t1\t1096 45\n",
      "22\t1390\t0.0\t1\t1336 54\n",
      "23\t2455\t0.0\t1\t2355 100\n",
      "24\t1638\t0.0\t1\t1566 72\n",
      "25\t2100\t0.0\t1\t2042 58\n",
      "26\t2090\t0.0\t1\t2002 88\n",
      "27\t1228\t0.0\t1\t1194 34\n",
      "28\t2205\t0.0\t1\t2129 76\n",
      "29\t1890\t0.0\t1\t1824 66\n",
      "30\t729\t0.0\t1\t676 53\n",
      "31\t1591\t0.0\t1\t1517 74\n",
      "32\t1111\t0.0\t1\t1068 43\n",
      "33\t1455\t0.0\t1\t1403 52\n",
      "34\t2386\t0.0\t1\t2302 84\n",
      "35\t2775\t0.0\t1\t2688 87\n",
      "36\t1025\t0.0\t1\t975 50\n",
      "37\t4261\t0.0\t1\t4127 134\n",
      "38\t4902\t0.0\t1\t4768 134\n",
      "39\t884\t0.0\t1\t854 30\n",
      "40\t98\t0.0\t1\t79 19\n",
      "41\t242\t0.0\t1\t219 23\n",
      "42\t330\t0.0\t1\t293 37\n",
      "43\t484\t0.0\t1\t464 20\n",
      "44\t210\t0.0\t1\t151 59\n",
      "45\t476\t0.0\t1\t242 234\n",
      "46\t383\t0.0\t1\t342 41\n",
      "47\t604\t0.0\t1\t558 46\n",
      "48\t411\t0.0\t1\t375 36\n",
      "49\t629\t0.0\t1\t581 48\n",
      "50\t299\t0.0\t1\t244 55\n",
      "51\t61\t0.0\t1\t41 20\n",
      "52\t75\t0.0\t1\t38 37\n",
      "53\t276\t0.0\t1\t185 91\n",
      "54\t358\t0.0\t1\t276 82\n",
      "55\t243\t0.0\t1\t178 65\n",
      "56\t939\t0.0\t1\t792 147\n",
      "57\t2355\t0.0\t1\t2242 113\n",
      "58\t128\t0.0\t1\t9 119\n",
      "59\t101\t0.0\t1\t14 87\n",
      "60\t45\t0.0\t1\t24 21\n",
      "61\t95\t0.0\t1\t33 62\n",
      "62\t99\t0.0\t1\t61 38\n",
      "63\t83\t0.0\t1\t0 83\n",
      "64\t256\t0.0\t1\t2 254\n",
      "65\t248\t0.0\t1\t7 241\n",
      "66\t79\t0.0\t1\t0 79\n",
      "67\t326\t0.0\t1\t1 325\n",
      "68\t254\t0.0\t1\t0 254\n",
      "69\t20\t0.0\t1\t0 20\n",
      "70\t1\t0.0\t1\t0 1\n",
      "71\t3\t0.0\t1\t0 3\n",
      "72\t3\t0.0\t1\t0 3\n",
      "73\t33\t0.0\t1\t0 33\n",
      "74\t20\t0.0\t1\t0 20\n",
      "75\t96\t0.0\t1\t0 96\n",
      "76\t746\t0.0\t1\t0 746\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022872_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022872_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022872_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022872_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022872_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.968 s (26.968 s/read; 2.22 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   395,332 (39.5%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               5,785,619 bp (7.6%)\n",
      "Total written (filtered):     68,186,295 bp (89.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 395332 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.7%\n",
      "  C: 28.6%\n",
      "  G: 22.1%\n",
      "  T: 19.6%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t239800\t250000.0\t0\t239800\n",
      "2\t66627\t62500.0\t0\t66627\n",
      "3\t17103\t15625.0\t0\t17103\n",
      "4\t4975\t3906.2\t0\t4975\n",
      "5\t1495\t976.6\t0\t1495\n",
      "6\t2113\t244.1\t0\t2113\n",
      "7\t1874\t61.0\t0\t1874\n",
      "8\t1821\t15.3\t0\t1821\n",
      "9\t1831\t3.8\t0\t1696 135\n",
      "10\t2269\t1.0\t1\t2150 119\n",
      "11\t1879\t0.2\t1\t1807 72\n",
      "12\t3121\t0.1\t1\t3001 120\n",
      "13\t1056\t0.0\t1\t995 61\n",
      "14\t3371\t0.0\t1\t3232 139\n",
      "15\t951\t0.0\t1\t920 31\n",
      "16\t941\t0.0\t1\t895 46\n",
      "17\t2725\t0.0\t1\t2614 111\n",
      "18\t900\t0.0\t1\t860 40\n",
      "19\t2944\t0.0\t1\t2830 114\n",
      "20\t1050\t0.0\t1\t1011 39\n",
      "21\t805\t0.0\t1\t780 25\n",
      "22\t1833\t0.0\t1\t1750 83\n",
      "23\t2141\t0.0\t1\t2039 102\n",
      "24\t2726\t0.0\t1\t2580 146\n",
      "25\t1352\t0.0\t1\t1311 41\n",
      "26\t1636\t0.0\t1\t1579 57\n",
      "27\t1151\t0.0\t1\t1087 64\n",
      "28\t2162\t0.0\t1\t2069 93\n",
      "29\t1266\t0.0\t1\t1215 51\n",
      "30\t2903\t0.0\t1\t2761 142\n",
      "31\t354\t0.0\t1\t339 15\n",
      "32\t2095\t0.0\t1\t2019 76\n",
      "33\t426\t0.0\t1\t407 19\n",
      "34\t1135\t0.0\t1\t1092 43\n",
      "35\t1445\t0.0\t1\t1379 66\n",
      "36\t1719\t0.0\t1\t1666 53\n",
      "37\t1220\t0.0\t1\t1179 41\n",
      "38\t580\t0.0\t1\t557 23\n",
      "39\t915\t0.0\t1\t881 34\n",
      "40\t453\t0.0\t1\t437 16\n",
      "41\t1182\t0.0\t1\t1155 27\n",
      "42\t1533\t0.0\t1\t1473 60\n",
      "43\t151\t0.0\t1\t140 11\n",
      "44\t1010\t0.0\t1\t976 34\n",
      "45\t2247\t0.0\t1\t2157 90\n",
      "46\t80\t0.0\t1\t70 10\n",
      "47\t22\t0.0\t1\t19 3\n",
      "48\t61\t0.0\t1\t56 5\n",
      "49\t34\t0.0\t1\t31 3\n",
      "50\t39\t0.0\t1\t38 1\n",
      "51\t27\t0.0\t1\t27\n",
      "52\t3\t0.0\t1\t3\n",
      "53\t2\t0.0\t1\t1 1\n",
      "54\t12\t0.0\t1\t11 1\n",
      "55\t14\t0.0\t1\t13 1\n",
      "56\t21\t0.0\t1\t18 3\n",
      "57\t1277\t0.0\t1\t1152 125\n",
      "58\t25\t0.0\t1\t24 1\n",
      "59\t3\t0.0\t1\t2 1\n",
      "60\t17\t0.0\t1\t17\n",
      "61\t13\t0.0\t1\t13\n",
      "62\t374\t0.0\t1\t355 19\n",
      "63\t5\t0.0\t1\t5\n",
      "64\t5\t0.0\t1\t5\n",
      "65\t7\t0.0\t1\t5 2\n",
      "66\t2\t0.0\t1\t1 1\n",
      "67\t2\t0.0\t1\t2\n",
      "73\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022872_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022872_1_trimmed.fq.gz and ERR022872_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022872_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022872_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022872_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022872_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022872_R1_trimmed.fq.gz, file_2: ERR022872_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022872_R1_trimmed.fq.gz and ERR022872_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022872_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022872_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 39451 (3.95%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022872_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022872_val_1.fq.gz\n",
      "Approx 5% complete for ERR022872_val_1.fq.gz\n",
      "Approx 10% complete for ERR022872_val_1.fq.gz\n",
      "Approx 15% complete for ERR022872_val_1.fq.gz\n",
      "Approx 20% complete for ERR022872_val_1.fq.gz\n",
      "Approx 25% complete for ERR022872_val_1.fq.gz\n",
      "Approx 30% complete for ERR022872_val_1.fq.gz\n",
      "Approx 35% complete for ERR022872_val_1.fq.gz\n",
      "Approx 40% complete for ERR022872_val_1.fq.gz\n",
      "Approx 45% complete for ERR022872_val_1.fq.gz\n",
      "Approx 50% complete for ERR022872_val_1.fq.gz\n",
      "Approx 55% complete for ERR022872_val_1.fq.gz\n",
      "Approx 60% complete for ERR022872_val_1.fq.gz\n",
      "Approx 65% complete for ERR022872_val_1.fq.gz\n",
      "Approx 70% complete for ERR022872_val_1.fq.gz\n",
      "Approx 75% complete for ERR022872_val_1.fq.gz\n",
      "Approx 80% complete for ERR022872_val_1.fq.gz\n",
      "Approx 85% complete for ERR022872_val_1.fq.gz\n",
      "Approx 90% complete for ERR022872_val_1.fq.gz\n",
      "Approx 95% complete for ERR022872_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022872_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022872_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022872_val_2.fq.gz\n",
      "Approx 5% complete for ERR022872_val_2.fq.gz\n",
      "Approx 10% complete for ERR022872_val_2.fq.gz\n",
      "Approx 15% complete for ERR022872_val_2.fq.gz\n",
      "Approx 20% complete for ERR022872_val_2.fq.gz\n",
      "Approx 25% complete for ERR022872_val_2.fq.gz\n",
      "Approx 30% complete for ERR022872_val_2.fq.gz\n",
      "Approx 35% complete for ERR022872_val_2.fq.gz\n",
      "Approx 40% complete for ERR022872_val_2.fq.gz\n",
      "Approx 45% complete for ERR022872_val_2.fq.gz\n",
      "Approx 50% complete for ERR022872_val_2.fq.gz\n",
      "Approx 55% complete for ERR022872_val_2.fq.gz\n",
      "Approx 60% complete for ERR022872_val_2.fq.gz\n",
      "Approx 65% complete for ERR022872_val_2.fq.gz\n",
      "Approx 70% complete for ERR022872_val_2.fq.gz\n",
      "Approx 75% complete for ERR022872_val_2.fq.gz\n",
      "Approx 80% complete for ERR022872_val_2.fq.gz\n",
      "Approx 85% complete for ERR022872_val_2.fq.gz\n",
      "Approx 90% complete for ERR022872_val_2.fq.gz\n",
      "Approx 95% complete for ERR022872_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022872_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022872_R1_trimmed.fq.gz and ERR022872_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022873<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022873_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t37821\tAGATCGGAAGAGC\t1000000\t3.78\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 37821). Second best hit was Nextera (count: 0)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022873_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022873_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022873_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022873_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 27.999 s (27.999 s/read; 2.14 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   356,784 (35.7%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               3,129,413 bp (4.1%)\n",
      "Total written (filtered):     71,329,521 bp (93.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 356784 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 28.0%\n",
      "  C: 28.9%\n",
      "  G: 21.6%\n",
      "  T: 21.2%\n",
      "  none/other: 0.3%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t214185\t250000.0\t0\t214185\n",
      "2\t65762\t62500.0\t0\t65762\n",
      "3\t16056\t15625.0\t0\t16056\n",
      "4\t6309\t3906.2\t0\t6309\n",
      "5\t2193\t976.6\t0\t2193\n",
      "6\t2609\t244.1\t0\t2609\n",
      "7\t2605\t61.0\t0\t2605\n",
      "8\t1719\t15.3\t0\t1719\n",
      "9\t654\t3.8\t0\t609 45\n",
      "10\t2036\t1.0\t1\t1880 156\n",
      "11\t1290\t0.2\t1\t1219 71\n",
      "12\t2046\t0.1\t1\t1935 111\n",
      "13\t1439\t0.0\t1\t1335 104\n",
      "14\t2168\t0.0\t1\t1978 190\n",
      "15\t2700\t0.0\t1\t2470 230\n",
      "16\t1844\t0.0\t1\t1699 145\n",
      "17\t1777\t0.0\t1\t1680 97\n",
      "18\t1503\t0.0\t1\t1316 187\n",
      "19\t1180\t0.0\t1\t1100 80\n",
      "20\t2113\t0.0\t1\t1959 154\n",
      "21\t1592\t0.0\t1\t1403 189\n",
      "22\t992\t0.0\t1\t899 93\n",
      "23\t1899\t0.0\t1\t1710 189\n",
      "24\t1848\t0.0\t1\t1650 198\n",
      "25\t1521\t0.0\t1\t1416 105\n",
      "26\t991\t0.0\t1\t924 67\n",
      "27\t907\t0.0\t1\t832 75\n",
      "28\t1666\t0.0\t1\t1543 123\n",
      "29\t1516\t0.0\t1\t1331 185\n",
      "30\t809\t0.0\t1\t691 118\n",
      "31\t1582\t0.0\t1\t1466 116\n",
      "32\t676\t0.0\t1\t643 33\n",
      "33\t1243\t0.0\t1\t1182 61\n",
      "34\t845\t0.0\t1\t813 32\n",
      "35\t1178\t0.0\t1\t1117 61\n",
      "36\t552\t0.0\t1\t508 44\n",
      "37\t1272\t0.0\t1\t1198 74\n",
      "38\t566\t0.0\t1\t543 23\n",
      "39\t563\t0.0\t1\t546 17\n",
      "40\t87\t0.0\t1\t80 7\n",
      "41\t368\t0.0\t1\t349 19\n",
      "42\t316\t0.0\t1\t284 32\n",
      "43\t349\t0.0\t1\t329 20\n",
      "44\t150\t0.0\t1\t138 12\n",
      "45\t137\t0.0\t1\t93 44\n",
      "46\t79\t0.0\t1\t65 14\n",
      "47\t104\t0.0\t1\t93 11\n",
      "48\t78\t0.0\t1\t67 11\n",
      "49\t27\t0.0\t1\t25 2\n",
      "50\t45\t0.0\t1\t35 10\n",
      "51\t20\t0.0\t1\t7 13\n",
      "52\t15\t0.0\t1\t5 10\n",
      "53\t37\t0.0\t1\t7 30\n",
      "54\t42\t0.0\t1\t3 39\n",
      "55\t11\t0.0\t1\t1 10\n",
      "56\t6\t0.0\t1\t4 2\n",
      "57\t14\t0.0\t1\t10 4\n",
      "58\t14\t0.0\t1\t0 14\n",
      "59\t7\t0.0\t1\t2 5\n",
      "60\t4\t0.0\t1\t1 3\n",
      "61\t12\t0.0\t1\t3 9\n",
      "62\t1\t0.0\t1\t0 1\n",
      "63\t14\t0.0\t1\t0 14\n",
      "64\t13\t0.0\t1\t0 13\n",
      "65\t7\t0.0\t1\t1 6\n",
      "66\t3\t0.0\t1\t0 3\n",
      "67\t36\t0.0\t1\t1 35\n",
      "68\t19\t0.0\t1\t0 19\n",
      "69\t3\t0.0\t1\t0 3\n",
      "70\t1\t0.0\t1\t0 1\n",
      "72\t1\t0.0\t1\t0 1\n",
      "73\t42\t0.0\t1\t0 42\n",
      "74\t39\t0.0\t1\t0 39\n",
      "75\t50\t0.0\t1\t0 50\n",
      "76\t227\t0.0\t1\t0 227\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022873_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022873_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022873_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022873_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022873_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.629 s (26.629 s/read; 2.25 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   394,873 (39.5%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               5,272,233 bp (6.9%)\n",
      "Total written (filtered):     69,162,733 bp (91.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 394873 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.5%\n",
      "  C: 27.1%\n",
      "  G: 20.7%\n",
      "  T: 21.5%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t246772\t250000.0\t0\t246772\n",
      "2\t70920\t62500.0\t0\t70920\n",
      "3\t17158\t15625.0\t0\t17158\n",
      "4\t6555\t3906.2\t0\t6555\n",
      "5\t1238\t976.6\t0\t1238\n",
      "6\t2129\t244.1\t0\t2129\n",
      "7\t2107\t61.0\t0\t2107\n",
      "8\t1919\t15.3\t0\t1919\n",
      "9\t1334\t3.8\t0\t1286 48\n",
      "10\t2326\t1.0\t1\t2184 142\n",
      "11\t1460\t0.2\t1\t1372 88\n",
      "12\t2972\t0.1\t1\t2731 241\n",
      "13\t855\t0.0\t1\t796 59\n",
      "14\t2595\t0.0\t1\t2355 240\n",
      "15\t1889\t0.0\t1\t1804 85\n",
      "16\t698\t0.0\t1\t663 35\n",
      "17\t2187\t0.0\t1\t2015 172\n",
      "18\t1078\t0.0\t1\t986 92\n",
      "19\t2429\t0.0\t1\t2100 329\n",
      "20\t1647\t0.0\t1\t1463 184\n",
      "21\t677\t0.0\t1\t639 38\n",
      "22\t1550\t0.0\t1\t1369 181\n",
      "23\t1622\t0.0\t1\t1521 101\n",
      "24\t2301\t0.0\t1\t2172 129\n",
      "25\t1424\t0.0\t1\t1364 60\n",
      "26\t1628\t0.0\t1\t1538 90\n",
      "27\t662\t0.0\t1\t628 34\n",
      "28\t1874\t0.0\t1\t1731 143\n",
      "29\t860\t0.0\t1\t818 42\n",
      "30\t2047\t0.0\t1\t1912 135\n",
      "31\t452\t0.0\t1\t431 21\n",
      "32\t1603\t0.0\t1\t1526 77\n",
      "33\t750\t0.0\t1\t719 31\n",
      "34\t756\t0.0\t1\t724 32\n",
      "35\t1148\t0.0\t1\t1036 112\n",
      "36\t905\t0.0\t1\t842 63\n",
      "37\t961\t0.0\t1\t930 31\n",
      "38\t373\t0.0\t1\t334 39\n",
      "39\t735\t0.0\t1\t702 33\n",
      "40\t218\t0.0\t1\t214 4\n",
      "41\t535\t0.0\t1\t515 20\n",
      "42\t574\t0.0\t1\t552 22\n",
      "43\t116\t0.0\t1\t109 7\n",
      "44\t292\t0.0\t1\t275 17\n",
      "45\t415\t0.0\t1\t397 18\n",
      "46\t17\t0.0\t1\t16 1\n",
      "47\t7\t0.0\t1\t7\n",
      "48\t49\t0.0\t1\t45 4\n",
      "49\t9\t0.0\t1\t9\n",
      "50\t5\t0.0\t1\t1 4\n",
      "51\t17\t0.0\t1\t15 2\n",
      "52\t2\t0.0\t1\t2\n",
      "54\t1\t0.0\t1\t0 1\n",
      "55\t1\t0.0\t1\t1\n",
      "56\t2\t0.0\t1\t0 2\n",
      "57\t4\t0.0\t1\t4\n",
      "58\t4\t0.0\t1\t3 1\n",
      "59\t2\t0.0\t1\t1 1\n",
      "61\t1\t0.0\t1\t0 1\n",
      "62\t1\t0.0\t1\t1\n",
      "64\t1\t0.0\t1\t1\n",
      "65\t1\t0.0\t1\t1\n",
      "67\t1\t0.0\t1\t0 1\n",
      "72\t1\t0.0\t1\t0 1\n",
      "73\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022873_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022873_1_trimmed.fq.gz and ERR022873_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022873_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022873_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022873_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022873_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022873_R1_trimmed.fq.gz, file_2: ERR022873_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022873_R1_trimmed.fq.gz and ERR022873_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022873_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022873_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 44357 (4.44%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022873_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022873_val_1.fq.gz\n",
      "Approx 5% complete for ERR022873_val_1.fq.gz\n",
      "Approx 10% complete for ERR022873_val_1.fq.gz\n",
      "Approx 15% complete for ERR022873_val_1.fq.gz\n",
      "Approx 20% complete for ERR022873_val_1.fq.gz\n",
      "Approx 25% complete for ERR022873_val_1.fq.gz\n",
      "Approx 30% complete for ERR022873_val_1.fq.gz\n",
      "Approx 35% complete for ERR022873_val_1.fq.gz\n",
      "Approx 40% complete for ERR022873_val_1.fq.gz\n",
      "Approx 45% complete for ERR022873_val_1.fq.gz\n",
      "Approx 50% complete for ERR022873_val_1.fq.gz\n",
      "Approx 55% complete for ERR022873_val_1.fq.gz\n",
      "Approx 60% complete for ERR022873_val_1.fq.gz\n",
      "Approx 65% complete for ERR022873_val_1.fq.gz\n",
      "Approx 70% complete for ERR022873_val_1.fq.gz\n",
      "Approx 75% complete for ERR022873_val_1.fq.gz\n",
      "Approx 80% complete for ERR022873_val_1.fq.gz\n",
      "Approx 85% complete for ERR022873_val_1.fq.gz\n",
      "Approx 90% complete for ERR022873_val_1.fq.gz\n",
      "Approx 95% complete for ERR022873_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022873_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022873_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022873_val_2.fq.gz\n",
      "Approx 5% complete for ERR022873_val_2.fq.gz\n",
      "Approx 10% complete for ERR022873_val_2.fq.gz\n",
      "Approx 15% complete for ERR022873_val_2.fq.gz\n",
      "Approx 20% complete for ERR022873_val_2.fq.gz\n",
      "Approx 25% complete for ERR022873_val_2.fq.gz\n",
      "Approx 30% complete for ERR022873_val_2.fq.gz\n",
      "Approx 35% complete for ERR022873_val_2.fq.gz\n",
      "Approx 40% complete for ERR022873_val_2.fq.gz\n",
      "Approx 45% complete for ERR022873_val_2.fq.gz\n",
      "Approx 50% complete for ERR022873_val_2.fq.gz\n",
      "Approx 55% complete for ERR022873_val_2.fq.gz\n",
      "Approx 60% complete for ERR022873_val_2.fq.gz\n",
      "Approx 65% complete for ERR022873_val_2.fq.gz\n",
      "Approx 70% complete for ERR022873_val_2.fq.gz\n",
      "Approx 75% complete for ERR022873_val_2.fq.gz\n",
      "Approx 80% complete for ERR022873_val_2.fq.gz\n",
      "Approx 85% complete for ERR022873_val_2.fq.gz\n",
      "Approx 90% complete for ERR022873_val_2.fq.gz\n",
      "Approx 95% complete for ERR022873_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022873_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022873_R1_trimmed.fq.gz and ERR022873_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022874<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022874_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t37771\tAGATCGGAAGAGC\t1000000\t3.78\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 37771). Second best hit was Nextera (count: 1)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022874_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022874_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022874_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022874_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 28.850 s (28.850 s/read; 2.08 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   346,093 (34.6%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               3,227,786 bp (4.2%)\n",
      "Total written (filtered):     71,246,629 bp (93.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 346093 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 26.7%\n",
      "  C: 30.4%\n",
      "  G: 23.6%\n",
      "  T: 19.1%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t204144\t250000.0\t0\t204144\n",
      "2\t69171\t62500.0\t0\t69171\n",
      "3\t15677\t15625.0\t0\t15677\n",
      "4\t5423\t3906.2\t0\t5423\n",
      "5\t1980\t976.6\t0\t1980\n",
      "6\t2395\t244.1\t0\t2395\n",
      "7\t2176\t61.0\t0\t2176\n",
      "8\t1537\t15.3\t0\t1537\n",
      "9\t477\t3.8\t0\t419 58\n",
      "10\t1751\t1.0\t1\t1633 118\n",
      "11\t1059\t0.2\t1\t1013 46\n",
      "12\t1692\t0.1\t1\t1611 81\n",
      "13\t1172\t0.0\t1\t1103 69\n",
      "14\t2050\t0.0\t1\t1919 131\n",
      "15\t2113\t0.0\t1\t1972 141\n",
      "16\t1649\t0.0\t1\t1545 104\n",
      "17\t1643\t0.0\t1\t1560 83\n",
      "18\t1374\t0.0\t1\t1191 183\n",
      "19\t1022\t0.0\t1\t958 64\n",
      "20\t1835\t0.0\t1\t1729 106\n",
      "21\t1444\t0.0\t1\t1279 165\n",
      "22\t873\t0.0\t1\t806 67\n",
      "23\t1919\t0.0\t1\t1754 165\n",
      "24\t1620\t0.0\t1\t1463 157\n",
      "25\t1582\t0.0\t1\t1495 87\n",
      "26\t979\t0.0\t1\t929 50\n",
      "27\t784\t0.0\t1\t721 63\n",
      "28\t1424\t0.0\t1\t1331 93\n",
      "29\t1723\t0.0\t1\t1548 175\n",
      "30\t779\t0.0\t1\t659 120\n",
      "31\t1655\t0.0\t1\t1540 115\n",
      "32\t672\t0.0\t1\t648 24\n",
      "33\t1430\t0.0\t1\t1368 62\n",
      "34\t837\t0.0\t1\t811 26\n",
      "35\t1533\t0.0\t1\t1463 70\n",
      "36\t621\t0.0\t1\t570 51\n",
      "37\t1676\t0.0\t1\t1604 72\n",
      "38\t869\t0.0\t1\t845 24\n",
      "39\t760\t0.0\t1\t747 13\n",
      "40\t98\t0.0\t1\t93 5\n",
      "41\t489\t0.0\t1\t475 14\n",
      "42\t410\t0.0\t1\t385 25\n",
      "43\t491\t0.0\t1\t474 17\n",
      "44\t165\t0.0\t1\t151 14\n",
      "45\t133\t0.0\t1\t126 7\n",
      "46\t113\t0.0\t1\t107 6\n",
      "47\t139\t0.0\t1\t134 5\n",
      "48\t124\t0.0\t1\t113 11\n",
      "49\t48\t0.0\t1\t42 6\n",
      "50\t80\t0.0\t1\t73 7\n",
      "51\t19\t0.0\t1\t17 2\n",
      "52\t14\t0.0\t1\t12 2\n",
      "53\t17\t0.0\t1\t9 8\n",
      "54\t17\t0.0\t1\t12 5\n",
      "55\t8\t0.0\t1\t6 2\n",
      "56\t10\t0.0\t1\t9 1\n",
      "57\t39\t0.0\t1\t38 1\n",
      "58\t12\t0.0\t1\t8 4\n",
      "59\t9\t0.0\t1\t8 1\n",
      "60\t8\t0.0\t1\t5 3\n",
      "61\t19\t0.0\t1\t19\n",
      "62\t13\t0.0\t1\t12 1\n",
      "63\t5\t0.0\t1\t2 3\n",
      "64\t6\t0.0\t1\t3 3\n",
      "65\t6\t0.0\t1\t5 1\n",
      "66\t9\t0.0\t1\t9\n",
      "67\t11\t0.0\t1\t3 8\n",
      "68\t5\t0.0\t1\t1 4\n",
      "69\t1\t0.0\t1\t1\n",
      "71\t1\t0.0\t1\t0 1\n",
      "73\t7\t0.0\t1\t0 7\n",
      "74\t5\t0.0\t1\t1 4\n",
      "75\t10\t0.0\t1\t0 10\n",
      "76\t32\t0.0\t1\t0 32\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022874_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022874_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022874_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022874_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022874_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.800 s (26.800 s/read; 2.24 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   386,158 (38.6%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               6,358,271 bp (8.4%)\n",
      "Total written (filtered):     68,054,077 bp (89.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 386158 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.2%\n",
      "  C: 29.2%\n",
      "  G: 22.1%\n",
      "  T: 19.4%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t239720\t250000.0\t0\t239720\n",
      "2\t72376\t62500.0\t0\t72376\n",
      "3\t16989\t15625.0\t0\t16989\n",
      "4\t5844\t3906.2\t0\t5844\n",
      "5\t1156\t976.6\t0\t1156\n",
      "6\t1767\t244.1\t0\t1767\n",
      "7\t1736\t61.0\t0\t1736\n",
      "8\t1677\t15.3\t0\t1677\n",
      "9\t1316\t3.8\t0\t1223 93\n",
      "10\t1962\t1.0\t1\t1856 106\n",
      "11\t1298\t0.2\t1\t1245 53\n",
      "12\t2554\t0.1\t1\t2379 175\n",
      "13\t704\t0.0\t1\t674 30\n",
      "14\t2264\t0.0\t1\t2110 154\n",
      "15\t1508\t0.0\t1\t1471 37\n",
      "16\t551\t0.0\t1\t522 29\n",
      "17\t1979\t0.0\t1\t1876 103\n",
      "18\t851\t0.0\t1\t799 52\n",
      "19\t2203\t0.0\t1\t1944 259\n",
      "20\t1407\t0.0\t1\t1267 140\n",
      "21\t509\t0.0\t1\t490 19\n",
      "22\t1454\t0.0\t1\t1296 158\n",
      "23\t1567\t0.0\t1\t1495 72\n",
      "24\t2001\t0.0\t1\t1913 88\n",
      "25\t1419\t0.0\t1\t1374 45\n",
      "26\t1610\t0.0\t1\t1543 67\n",
      "27\t636\t0.0\t1\t606 30\n",
      "28\t1762\t0.0\t1\t1664 98\n",
      "29\t851\t0.0\t1\t816 35\n",
      "30\t2192\t0.0\t1\t2110 82\n",
      "31\t350\t0.0\t1\t337 13\n",
      "32\t1836\t0.0\t1\t1795 41\n",
      "33\t716\t0.0\t1\t699 17\n",
      "34\t813\t0.0\t1\t789 24\n",
      "35\t1276\t0.0\t1\t1219 57\n",
      "36\t1113\t0.0\t1\t1056 57\n",
      "37\t1234\t0.0\t1\t1203 31\n",
      "38\t622\t0.0\t1\t587 35\n",
      "39\t1008\t0.0\t1\t976 32\n",
      "40\t270\t0.0\t1\t263 7\n",
      "41\t812\t0.0\t1\t780 32\n",
      "42\t827\t0.0\t1\t791 36\n",
      "43\t138\t0.0\t1\t136 2\n",
      "44\t373\t0.0\t1\t354 19\n",
      "45\t645\t0.0\t1\t625 20\n",
      "46\t19\t0.0\t1\t19\n",
      "47\t11\t0.0\t1\t11\n",
      "48\t72\t0.0\t1\t69 3\n",
      "49\t19\t0.0\t1\t18 1\n",
      "50\t11\t0.0\t1\t7 4\n",
      "51\t33\t0.0\t1\t30 3\n",
      "54\t4\t0.0\t1\t4\n",
      "56\t2\t0.0\t1\t2\n",
      "57\t33\t0.0\t1\t24 9\n",
      "58\t12\t0.0\t1\t11 1\n",
      "59\t2\t0.0\t1\t2\n",
      "60\t3\t0.0\t1\t2 1\n",
      "61\t4\t0.0\t1\t3 1\n",
      "62\t10\t0.0\t1\t9 1\n",
      "63\t6\t0.0\t1\t4 2\n",
      "64\t6\t0.0\t1\t6\n",
      "65\t9\t0.0\t1\t8 1\n",
      "66\t3\t0.0\t1\t3\n",
      "67\t1\t0.0\t1\t1\n",
      "69\t2\t0.0\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022874_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022874_1_trimmed.fq.gz and ERR022874_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022874_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022874_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022874_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022874_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022874_R1_trimmed.fq.gz, file_2: ERR022874_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022874_R1_trimmed.fq.gz and ERR022874_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022874_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022874_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 42622 (4.26%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022874_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022874_val_1.fq.gz\n",
      "Approx 5% complete for ERR022874_val_1.fq.gz\n",
      "Approx 10% complete for ERR022874_val_1.fq.gz\n",
      "Approx 15% complete for ERR022874_val_1.fq.gz\n",
      "Approx 20% complete for ERR022874_val_1.fq.gz\n",
      "Approx 25% complete for ERR022874_val_1.fq.gz\n",
      "Approx 30% complete for ERR022874_val_1.fq.gz\n",
      "Approx 35% complete for ERR022874_val_1.fq.gz\n",
      "Approx 40% complete for ERR022874_val_1.fq.gz\n",
      "Approx 45% complete for ERR022874_val_1.fq.gz\n",
      "Approx 50% complete for ERR022874_val_1.fq.gz\n",
      "Approx 55% complete for ERR022874_val_1.fq.gz\n",
      "Approx 60% complete for ERR022874_val_1.fq.gz\n",
      "Approx 65% complete for ERR022874_val_1.fq.gz\n",
      "Approx 70% complete for ERR022874_val_1.fq.gz\n",
      "Approx 75% complete for ERR022874_val_1.fq.gz\n",
      "Approx 80% complete for ERR022874_val_1.fq.gz\n",
      "Approx 85% complete for ERR022874_val_1.fq.gz\n",
      "Approx 90% complete for ERR022874_val_1.fq.gz\n",
      "Approx 95% complete for ERR022874_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022874_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022874_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022874_val_2.fq.gz\n",
      "Approx 5% complete for ERR022874_val_2.fq.gz\n",
      "Approx 10% complete for ERR022874_val_2.fq.gz\n",
      "Approx 15% complete for ERR022874_val_2.fq.gz\n",
      "Approx 20% complete for ERR022874_val_2.fq.gz\n",
      "Approx 25% complete for ERR022874_val_2.fq.gz\n",
      "Approx 30% complete for ERR022874_val_2.fq.gz\n",
      "Approx 35% complete for ERR022874_val_2.fq.gz\n",
      "Approx 40% complete for ERR022874_val_2.fq.gz\n",
      "Approx 45% complete for ERR022874_val_2.fq.gz\n",
      "Approx 50% complete for ERR022874_val_2.fq.gz\n",
      "Approx 55% complete for ERR022874_val_2.fq.gz\n",
      "Approx 60% complete for ERR022874_val_2.fq.gz\n",
      "Approx 65% complete for ERR022874_val_2.fq.gz\n",
      "Approx 70% complete for ERR022874_val_2.fq.gz\n",
      "Approx 75% complete for ERR022874_val_2.fq.gz\n",
      "Approx 80% complete for ERR022874_val_2.fq.gz\n",
      "Approx 85% complete for ERR022874_val_2.fq.gz\n",
      "Approx 90% complete for ERR022874_val_2.fq.gz\n",
      "Approx 95% complete for ERR022874_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022874_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022874_R1_trimmed.fq.gz and ERR022874_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022875<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022875_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t26841\tAGATCGGAAGAGC\t1000000\t2.68\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 26841). Second best hit was Nextera (count: 1)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022875_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022875_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022875_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022875_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 25.820 s (25.820 s/read; 2.32 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   331,796 (33.2%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               5,113,043 bp (6.7%)\n",
      "Total written (filtered):     69,760,504 bp (91.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 331796 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.7%\n",
      "  C: 28.4%\n",
      "  G: 20.1%\n",
      "  T: 21.7%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t211025\t250000.0\t0\t211025\n",
      "2\t63382\t62500.0\t0\t63382\n",
      "3\t15827\t15625.0\t0\t15827\n",
      "4\t5075\t3906.2\t0\t5075\n",
      "5\t1843\t976.6\t0\t1843\n",
      "6\t1842\t244.1\t0\t1842\n",
      "7\t1633\t61.0\t0\t1633\n",
      "8\t1059\t15.3\t0\t1059\n",
      "9\t364\t3.8\t0\t269 95\n",
      "10\t1299\t1.0\t1\t1184 115\n",
      "11\t711\t0.2\t1\t675 36\n",
      "12\t1392\t0.1\t1\t1300 92\n",
      "13\t680\t0.0\t1\t606 74\n",
      "14\t1754\t0.0\t1\t1572 182\n",
      "15\t2600\t0.0\t1\t2379 221\n",
      "16\t1375\t0.0\t1\t1273 102\n",
      "17\t1169\t0.0\t1\t1089 80\n",
      "18\t1107\t0.0\t1\t942 165\n",
      "19\t610\t0.0\t1\t565 45\n",
      "20\t1184\t0.0\t1\t1103 81\n",
      "21\t1052\t0.0\t1\t896 156\n",
      "22\t492\t0.0\t1\t446 46\n",
      "23\t1521\t0.0\t1\t1340 181\n",
      "24\t1265\t0.0\t1\t1126 139\n",
      "25\t1160\t0.0\t1\t1070 90\n",
      "26\t682\t0.0\t1\t646 36\n",
      "27\t431\t0.0\t1\t394 37\n",
      "28\t1012\t0.0\t1\t951 61\n",
      "29\t1162\t0.0\t1\t1046 116\n",
      "30\t560\t0.0\t1\t480 80\n",
      "31\t1100\t0.0\t1\t1020 80\n",
      "32\t369\t0.0\t1\t348 21\n",
      "33\t833\t0.0\t1\t792 41\n",
      "34\t511\t0.0\t1\t488 23\n",
      "35\t886\t0.0\t1\t840 46\n",
      "36\t401\t0.0\t1\t359 42\n",
      "37\t906\t0.0\t1\t850 56\n",
      "38\t343\t0.0\t1\t329 14\n",
      "39\t273\t0.0\t1\t264 9\n",
      "40\t22\t0.0\t1\t18 4\n",
      "41\t178\t0.0\t1\t171 7\n",
      "42\t175\t0.0\t1\t163 12\n",
      "43\t143\t0.0\t1\t140 3\n",
      "44\t50\t0.0\t1\t45 5\n",
      "45\t52\t0.0\t1\t45 7\n",
      "46\t30\t0.0\t1\t25 5\n",
      "47\t53\t0.0\t1\t50 3\n",
      "48\t42\t0.0\t1\t34 8\n",
      "49\t21\t0.0\t1\t17 4\n",
      "50\t14\t0.0\t1\t12 2\n",
      "51\t5\t0.0\t1\t4 1\n",
      "52\t2\t0.0\t1\t2\n",
      "53\t11\t0.0\t1\t10 1\n",
      "54\t11\t0.0\t1\t8 3\n",
      "55\t5\t0.0\t1\t5\n",
      "56\t7\t0.0\t1\t6 1\n",
      "57\t13\t0.0\t1\t11 2\n",
      "58\t1\t0.0\t1\t1\n",
      "59\t6\t0.0\t1\t4 2\n",
      "60\t8\t0.0\t1\t4 4\n",
      "61\t9\t0.0\t1\t7 2\n",
      "62\t4\t0.0\t1\t3 1\n",
      "63\t3\t0.0\t1\t1 2\n",
      "64\t1\t0.0\t1\t0 1\n",
      "65\t5\t0.0\t1\t3 2\n",
      "66\t1\t0.0\t1\t1\n",
      "67\t7\t0.0\t1\t3 4\n",
      "68\t1\t0.0\t1\t0 1\n",
      "69\t2\t0.0\t1\t2\n",
      "70\t1\t0.0\t1\t0 1\n",
      "71\t2\t0.0\t1\t0 2\n",
      "72\t1\t0.0\t1\t0 1\n",
      "73\t3\t0.0\t1\t0 3\n",
      "74\t1\t0.0\t1\t0 1\n",
      "75\t3\t0.0\t1\t0 3\n",
      "76\t13\t0.0\t1\t0 13\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022875_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022875_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022875_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022875_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022875_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.945 s (26.945 s/read; 2.23 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   380,113 (38.0%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               4,718,011 bp (6.2%)\n",
      "Total written (filtered):     69,987,237 bp (92.1%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 380113 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 31.3%\n",
      "  C: 27.2%\n",
      "  G: 19.4%\n",
      "  T: 22.0%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t250035\t250000.0\t0\t250035\n",
      "2\t68109\t62500.0\t0\t68109\n",
      "3\t16418\t15625.0\t0\t16418\n",
      "4\t5700\t3906.2\t0\t5700\n",
      "5\t1311\t976.6\t0\t1311\n",
      "6\t1366\t244.1\t0\t1366\n",
      "7\t1403\t61.0\t0\t1403\n",
      "8\t1325\t15.3\t0\t1325\n",
      "9\t988\t3.8\t0\t894 94\n",
      "10\t1661\t1.0\t1\t1563 98\n",
      "11\t978\t0.2\t1\t929 49\n",
      "12\t2056\t0.1\t1\t1861 195\n",
      "13\t552\t0.0\t1\t516 36\n",
      "14\t1982\t0.0\t1\t1768 214\n",
      "15\t1108\t0.0\t1\t1061 47\n",
      "16\t470\t0.0\t1\t438 32\n",
      "17\t1776\t0.0\t1\t1642 134\n",
      "18\t632\t0.0\t1\t588 44\n",
      "19\t1858\t0.0\t1\t1560 298\n",
      "20\t1049\t0.0\t1\t926 123\n",
      "21\t403\t0.0\t1\t382 21\n",
      "22\t1210\t0.0\t1\t1028 182\n",
      "23\t1231\t0.0\t1\t1169 62\n",
      "24\t1873\t0.0\t1\t1773 100\n",
      "25\t1065\t0.0\t1\t1029 36\n",
      "26\t1289\t0.0\t1\t1232 57\n",
      "27\t566\t0.0\t1\t536 30\n",
      "28\t1525\t0.0\t1\t1442 83\n",
      "29\t596\t0.0\t1\t571 25\n",
      "30\t1765\t0.0\t1\t1684 81\n",
      "31\t286\t0.0\t1\t273 13\n",
      "32\t1406\t0.0\t1\t1355 51\n",
      "33\t467\t0.0\t1\t449 18\n",
      "34\t575\t0.0\t1\t545 30\n",
      "35\t936\t0.0\t1\t870 66\n",
      "36\t713\t0.0\t1\t671 42\n",
      "37\t738\t0.0\t1\t714 24\n",
      "38\t387\t0.0\t1\t367 20\n",
      "39\t572\t0.0\t1\t552 20\n",
      "40\t184\t0.0\t1\t180 4\n",
      "41\t430\t0.0\t1\t415 15\n",
      "42\t425\t0.0\t1\t416 9\n",
      "43\t45\t0.0\t1\t44 1\n",
      "44\t184\t0.0\t1\t171 13\n",
      "45\t316\t0.0\t1\t300 16\n",
      "46\t21\t0.0\t1\t18 3\n",
      "47\t9\t0.0\t1\t7 2\n",
      "48\t22\t0.0\t1\t22\n",
      "49\t5\t0.0\t1\t4 1\n",
      "50\t5\t0.0\t1\t4 1\n",
      "51\t22\t0.0\t1\t20 2\n",
      "52\t2\t0.0\t1\t0 2\n",
      "54\t4\t0.0\t1\t4\n",
      "55\t1\t0.0\t1\t0 1\n",
      "57\t12\t0.0\t1\t10 2\n",
      "58\t4\t0.0\t1\t2 2\n",
      "59\t3\t0.0\t1\t3\n",
      "60\t1\t0.0\t1\t1\n",
      "61\t5\t0.0\t1\t4 1\n",
      "62\t6\t0.0\t1\t6\n",
      "63\t7\t0.0\t1\t6 1\n",
      "64\t3\t0.0\t1\t3\n",
      "65\t6\t0.0\t1\t6\n",
      "67\t2\t0.0\t1\t2\n",
      "68\t1\t0.0\t1\t0 1\n",
      "70\t1\t0.0\t1\t0 1\n",
      "71\t3\t0.0\t1\t0 3\n",
      "72\t1\t0.0\t1\t0 1\n",
      "73\t3\t0.0\t1\t0 3\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022875_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022875_1_trimmed.fq.gz and ERR022875_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022875_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022875_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022875_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022875_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022875_R1_trimmed.fq.gz, file_2: ERR022875_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022875_R1_trimmed.fq.gz and ERR022875_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022875_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022875_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 38260 (3.83%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022875_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022875_val_1.fq.gz\n",
      "Approx 5% complete for ERR022875_val_1.fq.gz\n",
      "Approx 10% complete for ERR022875_val_1.fq.gz\n",
      "Approx 15% complete for ERR022875_val_1.fq.gz\n",
      "Approx 20% complete for ERR022875_val_1.fq.gz\n",
      "Approx 25% complete for ERR022875_val_1.fq.gz\n",
      "Approx 30% complete for ERR022875_val_1.fq.gz\n",
      "Approx 35% complete for ERR022875_val_1.fq.gz\n",
      "Approx 40% complete for ERR022875_val_1.fq.gz\n",
      "Approx 45% complete for ERR022875_val_1.fq.gz\n",
      "Approx 50% complete for ERR022875_val_1.fq.gz\n",
      "Approx 55% complete for ERR022875_val_1.fq.gz\n",
      "Approx 60% complete for ERR022875_val_1.fq.gz\n",
      "Approx 65% complete for ERR022875_val_1.fq.gz\n",
      "Approx 70% complete for ERR022875_val_1.fq.gz\n",
      "Approx 75% complete for ERR022875_val_1.fq.gz\n",
      "Approx 80% complete for ERR022875_val_1.fq.gz\n",
      "Approx 85% complete for ERR022875_val_1.fq.gz\n",
      "Approx 90% complete for ERR022875_val_1.fq.gz\n",
      "Approx 95% complete for ERR022875_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022875_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022875_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022875_val_2.fq.gz\n",
      "Approx 5% complete for ERR022875_val_2.fq.gz\n",
      "Approx 10% complete for ERR022875_val_2.fq.gz\n",
      "Approx 15% complete for ERR022875_val_2.fq.gz\n",
      "Approx 20% complete for ERR022875_val_2.fq.gz\n",
      "Approx 25% complete for ERR022875_val_2.fq.gz\n",
      "Approx 30% complete for ERR022875_val_2.fq.gz\n",
      "Approx 35% complete for ERR022875_val_2.fq.gz\n",
      "Approx 40% complete for ERR022875_val_2.fq.gz\n",
      "Approx 45% complete for ERR022875_val_2.fq.gz\n",
      "Approx 50% complete for ERR022875_val_2.fq.gz\n",
      "Approx 55% complete for ERR022875_val_2.fq.gz\n",
      "Approx 60% complete for ERR022875_val_2.fq.gz\n",
      "Approx 65% complete for ERR022875_val_2.fq.gz\n",
      "Approx 70% complete for ERR022875_val_2.fq.gz\n",
      "Approx 75% complete for ERR022875_val_2.fq.gz\n",
      "Approx 80% complete for ERR022875_val_2.fq.gz\n",
      "Approx 85% complete for ERR022875_val_2.fq.gz\n",
      "Approx 90% complete for ERR022875_val_2.fq.gz\n",
      "Approx 95% complete for ERR022875_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022875_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022875_R1_trimmed.fq.gz and ERR022875_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022876<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022876_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t22641\tAGATCGGAAGAGC\t1000000\t2.26\n",
      "Nextera\t2\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 22641). Second best hit was Nextera (count: 2)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022876_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022876_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022876_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022876_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.192 s (26.192 s/read; 2.29 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   355,235 (35.5%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:                 980,171 bp (1.3%)\n",
      "Total written (filtered):     73,924,062 bp (97.3%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 355235 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 31.1%\n",
      "  C: 27.7%\n",
      "  G: 21.3%\n",
      "  T: 19.9%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t240600\t250000.0\t0\t240600\n",
      "2\t61033\t62500.0\t0\t61033\n",
      "3\t16561\t15625.0\t0\t16561\n",
      "4\t4972\t3906.2\t0\t4972\n",
      "5\t1966\t976.6\t0\t1966\n",
      "6\t1275\t244.1\t0\t1275\n",
      "7\t1146\t61.0\t0\t1146\n",
      "8\t1141\t15.3\t0\t1141\n",
      "9\t797\t3.8\t0\t771 26\n",
      "10\t1283\t1.0\t1\t1244 39\n",
      "11\t772\t0.2\t1\t754 18\n",
      "12\t1266\t0.1\t1\t1238 28\n",
      "13\t698\t0.0\t1\t685 13\n",
      "14\t1027\t0.0\t1\t1004 23\n",
      "15\t1134\t0.0\t1\t1105 29\n",
      "16\t942\t0.0\t1\t923 19\n",
      "17\t935\t0.0\t1\t932 3\n",
      "18\t900\t0.0\t1\t890 10\n",
      "19\t717\t0.0\t1\t704 13\n",
      "20\t947\t0.0\t1\t930 17\n",
      "21\t938\t0.0\t1\t922 16\n",
      "22\t655\t0.0\t1\t645 10\n",
      "23\t902\t0.0\t1\t893 9\n",
      "24\t861\t0.0\t1\t840 21\n",
      "25\t853\t0.0\t1\t836 17\n",
      "26\t873\t0.0\t1\t859 14\n",
      "27\t663\t0.0\t1\t657 6\n",
      "28\t785\t0.0\t1\t770 15\n",
      "29\t807\t0.0\t1\t793 14\n",
      "30\t654\t0.0\t1\t651 3\n",
      "31\t882\t0.0\t1\t876 6\n",
      "32\t607\t0.0\t1\t598 9\n",
      "33\t725\t0.0\t1\t721 4\n",
      "34\t589\t0.0\t1\t580 9\n",
      "35\t644\t0.0\t1\t639 5\n",
      "36\t505\t0.0\t1\t497 8\n",
      "37\t623\t0.0\t1\t615 8\n",
      "38\t573\t0.0\t1\t564 9\n",
      "39\t719\t0.0\t1\t714 5\n",
      "40\t93\t0.0\t1\t93\n",
      "41\t180\t0.0\t1\t177 3\n",
      "42\t180\t0.0\t1\t179 1\n",
      "43\t231\t0.0\t1\t226 5\n",
      "44\t102\t0.0\t1\t101 1\n",
      "45\t102\t0.0\t1\t99 3\n",
      "46\t46\t0.0\t1\t45 1\n",
      "47\t58\t0.0\t1\t56 2\n",
      "48\t52\t0.0\t1\t50 2\n",
      "49\t37\t0.0\t1\t35 2\n",
      "50\t30\t0.0\t1\t30\n",
      "51\t15\t0.0\t1\t15\n",
      "52\t13\t0.0\t1\t11 2\n",
      "53\t5\t0.0\t1\t5\n",
      "54\t5\t0.0\t1\t5\n",
      "55\t5\t0.0\t1\t3 2\n",
      "56\t7\t0.0\t1\t7\n",
      "57\t6\t0.0\t1\t6\n",
      "58\t2\t0.0\t1\t2\n",
      "59\t1\t0.0\t1\t1\n",
      "60\t5\t0.0\t1\t5\n",
      "61\t5\t0.0\t1\t5\n",
      "63\t1\t0.0\t1\t1\n",
      "64\t2\t0.0\t1\t2\n",
      "65\t4\t0.0\t1\t2 2\n",
      "66\t5\t0.0\t1\t3 2\n",
      "67\t4\t0.0\t1\t2 2\n",
      "68\t5\t0.0\t1\t0 5\n",
      "69\t5\t0.0\t1\t0 5\n",
      "70\t1\t0.0\t1\t1\n",
      "71\t1\t0.0\t1\t0 1\n",
      "72\t1\t0.0\t1\t0 1\n",
      "73\t2\t0.0\t1\t0 2\n",
      "75\t11\t0.0\t1\t0 11\n",
      "76\t43\t0.0\t1\t1 42\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022876_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022876_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022876_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022876_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022876_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 25.490 s (25.490 s/read; 2.35 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   340,987 (34.1%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               1,463,224 bp (1.9%)\n",
      "Total written (filtered):     73,478,790 bp (96.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 340987 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.3%\n",
      "  C: 29.6%\n",
      "  G: 20.5%\n",
      "  T: 20.7%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t233030\t250000.0\t0\t233030\n",
      "2\t56466\t62500.0\t0\t56466\n",
      "3\t14900\t15625.0\t0\t14900\n",
      "4\t4721\t3906.2\t0\t4721\n",
      "5\t2068\t976.6\t0\t2068\n",
      "6\t1266\t244.1\t0\t1266\n",
      "7\t1140\t61.0\t0\t1140\n",
      "8\t1068\t15.3\t0\t1068\n",
      "9\t889\t3.8\t0\t854 35\n",
      "10\t1295\t1.0\t1\t1250 45\n",
      "11\t792\t0.2\t1\t783 9\n",
      "12\t1192\t0.1\t1\t1174 18\n",
      "13\t799\t0.0\t1\t785 14\n",
      "14\t1194\t0.0\t1\t1166 28\n",
      "15\t890\t0.0\t1\t874 16\n",
      "16\t803\t0.0\t1\t795 8\n",
      "17\t1130\t0.0\t1\t1100 30\n",
      "18\t761\t0.0\t1\t748 13\n",
      "19\t1118\t0.0\t1\t1095 23\n",
      "20\t1262\t0.0\t1\t1247 15\n",
      "21\t213\t0.0\t1\t210 3\n",
      "22\t829\t0.0\t1\t820 9\n",
      "23\t910\t0.0\t1\t881 29\n",
      "24\t1129\t0.0\t1\t1110 19\n",
      "25\t744\t0.0\t1\t737 7\n",
      "26\t1110\t0.0\t1\t1094 16\n",
      "27\t404\t0.0\t1\t397 7\n",
      "28\t1286\t0.0\t1\t1279 7\n",
      "29\t246\t0.0\t1\t237 9\n",
      "30\t1012\t0.0\t1\t999 13\n",
      "31\t343\t0.0\t1\t334 9\n",
      "32\t961\t0.0\t1\t953 8\n",
      "33\t546\t0.0\t1\t535 11\n",
      "34\t344\t0.0\t1\t340 4\n",
      "35\t634\t0.0\t1\t624 10\n",
      "36\t445\t0.0\t1\t440 5\n",
      "37\t442\t0.0\t1\t437 5\n",
      "38\t416\t0.0\t1\t410 6\n",
      "39\t448\t0.0\t1\t444 4\n",
      "40\t291\t0.0\t1\t286 5\n",
      "41\t338\t0.0\t1\t334 4\n",
      "42\t353\t0.0\t1\t350 3\n",
      "43\t142\t0.0\t1\t139 3\n",
      "44\t182\t0.0\t1\t178 4\n",
      "45\t180\t0.0\t1\t178 2\n",
      "46\t34\t0.0\t1\t32 2\n",
      "47\t27\t0.0\t1\t27\n",
      "48\t41\t0.0\t1\t40 1\n",
      "49\t41\t0.0\t1\t41\n",
      "50\t14\t0.0\t1\t14\n",
      "51\t16\t0.0\t1\t16\n",
      "52\t9\t0.0\t1\t9\n",
      "53\t5\t0.0\t1\t5\n",
      "54\t6\t0.0\t1\t6\n",
      "55\t1\t0.0\t1\t1\n",
      "56\t2\t0.0\t1\t2\n",
      "57\t8\t0.0\t1\t7 1\n",
      "58\t3\t0.0\t1\t3\n",
      "59\t2\t0.0\t1\t2\n",
      "60\t2\t0.0\t1\t1 1\n",
      "61\t3\t0.0\t1\t2 1\n",
      "62\t11\t0.0\t1\t8 3\n",
      "63\t7\t0.0\t1\t7\n",
      "64\t14\t0.0\t1\t12 2\n",
      "65\t3\t0.0\t1\t1 2\n",
      "66\t1\t0.0\t1\t1\n",
      "68\t2\t0.0\t1\t0 2\n",
      "72\t2\t0.0\t1\t1 1\n",
      "73\t1\t0.0\t1\t1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022876_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022876_1_trimmed.fq.gz and ERR022876_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022876_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022876_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022876_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022876_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022876_R1_trimmed.fq.gz, file_2: ERR022876_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022876_R1_trimmed.fq.gz and ERR022876_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022876_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022876_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 6828 (0.68%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022876_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022876_val_1.fq.gz\n",
      "Approx 5% complete for ERR022876_val_1.fq.gz\n",
      "Approx 10% complete for ERR022876_val_1.fq.gz\n",
      "Approx 15% complete for ERR022876_val_1.fq.gz\n",
      "Approx 20% complete for ERR022876_val_1.fq.gz\n",
      "Approx 25% complete for ERR022876_val_1.fq.gz\n",
      "Approx 30% complete for ERR022876_val_1.fq.gz\n",
      "Approx 35% complete for ERR022876_val_1.fq.gz\n",
      "Approx 40% complete for ERR022876_val_1.fq.gz\n",
      "Approx 45% complete for ERR022876_val_1.fq.gz\n",
      "Approx 50% complete for ERR022876_val_1.fq.gz\n",
      "Approx 55% complete for ERR022876_val_1.fq.gz\n",
      "Approx 60% complete for ERR022876_val_1.fq.gz\n",
      "Approx 65% complete for ERR022876_val_1.fq.gz\n",
      "Approx 70% complete for ERR022876_val_1.fq.gz\n",
      "Approx 75% complete for ERR022876_val_1.fq.gz\n",
      "Approx 80% complete for ERR022876_val_1.fq.gz\n",
      "Approx 85% complete for ERR022876_val_1.fq.gz\n",
      "Approx 90% complete for ERR022876_val_1.fq.gz\n",
      "Approx 95% complete for ERR022876_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022876_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022876_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022876_val_2.fq.gz\n",
      "Approx 5% complete for ERR022876_val_2.fq.gz\n",
      "Approx 10% complete for ERR022876_val_2.fq.gz\n",
      "Approx 15% complete for ERR022876_val_2.fq.gz\n",
      "Approx 20% complete for ERR022876_val_2.fq.gz\n",
      "Approx 25% complete for ERR022876_val_2.fq.gz\n",
      "Approx 30% complete for ERR022876_val_2.fq.gz\n",
      "Approx 35% complete for ERR022876_val_2.fq.gz\n",
      "Approx 40% complete for ERR022876_val_2.fq.gz\n",
      "Approx 45% complete for ERR022876_val_2.fq.gz\n",
      "Approx 50% complete for ERR022876_val_2.fq.gz\n",
      "Approx 55% complete for ERR022876_val_2.fq.gz\n",
      "Approx 60% complete for ERR022876_val_2.fq.gz\n",
      "Approx 65% complete for ERR022876_val_2.fq.gz\n",
      "Approx 70% complete for ERR022876_val_2.fq.gz\n",
      "Approx 75% complete for ERR022876_val_2.fq.gz\n",
      "Approx 80% complete for ERR022876_val_2.fq.gz\n",
      "Approx 85% complete for ERR022876_val_2.fq.gz\n",
      "Approx 90% complete for ERR022876_val_2.fq.gz\n",
      "Approx 95% complete for ERR022876_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022876_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022876_R1_trimmed.fq.gz and ERR022876_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022877<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022877_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t1827\tAGATCGGAAGAGC\t1000000\t0.18\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 1827). Second best hit was Nextera (count: 1)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022877_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022877_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022877_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022877_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.435 s (26.435 s/read; 2.27 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   339,359 (33.9%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               1,171,879 bp (1.5%)\n",
      "Total written (filtered):     74,314,310 bp (97.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 339359 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 31.9%\n",
      "  C: 25.5%\n",
      "  G: 18.1%\n",
      "  T: 24.4%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t252076\t250000.0\t0\t252076\n",
      "2\t60628\t62500.0\t0\t60628\n",
      "3\t16566\t15625.0\t0\t16566\n",
      "4\t5015\t3906.2\t0\t5015\n",
      "5\t1856\t976.6\t0\t1856\n",
      "6\t367\t244.1\t0\t367\n",
      "7\t179\t61.0\t0\t179\n",
      "8\t211\t15.3\t0\t211\n",
      "9\t116\t3.8\t0\t77 39\n",
      "10\t206\t1.0\t1\t164 42\n",
      "11\t111\t0.2\t1\t104 7\n",
      "12\t176\t0.1\t1\t163 13\n",
      "13\t93\t0.0\t1\t91 2\n",
      "14\t134\t0.0\t1\t127 7\n",
      "15\t113\t0.0\t1\t107 6\n",
      "16\t108\t0.0\t1\t107 1\n",
      "17\t112\t0.0\t1\t107 5\n",
      "18\t99\t0.0\t1\t98 1\n",
      "19\t76\t0.0\t1\t76\n",
      "20\t101\t0.0\t1\t99 2\n",
      "21\t80\t0.0\t1\t76 4\n",
      "22\t47\t0.0\t1\t46 1\n",
      "23\t65\t0.0\t1\t61 4\n",
      "24\t69\t0.0\t1\t67 2\n",
      "25\t70\t0.0\t1\t67 3\n",
      "26\t52\t0.0\t1\t51 1\n",
      "27\t44\t0.0\t1\t40 4\n",
      "28\t46\t0.0\t1\t43 3\n",
      "29\t45\t0.0\t1\t44 1\n",
      "30\t27\t0.0\t1\t25 2\n",
      "31\t31\t0.0\t1\t31\n",
      "32\t29\t0.0\t1\t29\n",
      "33\t42\t0.0\t1\t39 3\n",
      "34\t28\t0.0\t1\t27 1\n",
      "35\t33\t0.0\t1\t33\n",
      "36\t19\t0.0\t1\t18 1\n",
      "37\t30\t0.0\t1\t28 2\n",
      "38\t26\t0.0\t1\t24 2\n",
      "39\t19\t0.0\t1\t19\n",
      "40\t2\t0.0\t1\t2\n",
      "41\t10\t0.0\t1\t9 1\n",
      "42\t7\t0.0\t1\t6 1\n",
      "43\t12\t0.0\t1\t12\n",
      "44\t7\t0.0\t1\t7\n",
      "45\t6\t0.0\t1\t2 4\n",
      "46\t1\t0.0\t1\t1\n",
      "47\t5\t0.0\t1\t5\n",
      "48\t3\t0.0\t1\t3\n",
      "49\t3\t0.0\t1\t3\n",
      "50\t2\t0.0\t1\t1 1\n",
      "51\t1\t0.0\t1\t0 1\n",
      "53\t4\t0.0\t1\t1 3\n",
      "54\t2\t0.0\t1\t2\n",
      "55\t1\t0.0\t1\t0 1\n",
      "57\t5\t0.0\t1\t4 1\n",
      "60\t2\t0.0\t1\t1 1\n",
      "61\t4\t0.0\t1\t1 3\n",
      "62\t2\t0.0\t1\t0 2\n",
      "64\t2\t0.0\t1\t0 2\n",
      "65\t1\t0.0\t1\t0 1\n",
      "67\t5\t0.0\t1\t0 5\n",
      "68\t6\t0.0\t1\t0 6\n",
      "69\t12\t0.0\t1\t0 12\n",
      "70\t1\t0.0\t1\t0 1\n",
      "71\t2\t0.0\t1\t0 2\n",
      "72\t3\t0.0\t1\t0 3\n",
      "73\t3\t0.0\t1\t0 3\n",
      "74\t4\t0.0\t1\t0 4\n",
      "75\t12\t0.0\t1\t0 12\n",
      "76\t84\t0.0\t1\t0 84\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022877_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022877_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022877_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022877_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022877_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 25.581 s (25.581 s/read; 2.35 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   346,498 (34.6%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               2,005,165 bp (2.6%)\n",
      "Total written (filtered):     73,488,728 bp (96.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 346498 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 33.4%\n",
      "  C: 25.7%\n",
      "  G: 17.6%\n",
      "  T: 23.2%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t262128\t250000.0\t0\t262128\n",
      "2\t58934\t62500.0\t0\t58934\n",
      "3\t15759\t15625.0\t0\t15759\n",
      "4\t4543\t3906.2\t0\t4543\n",
      "5\t1994\t976.6\t0\t1994\n",
      "6\t358\t244.1\t0\t358\n",
      "7\t176\t61.0\t0\t176\n",
      "8\t186\t15.3\t0\t186\n",
      "9\t174\t3.8\t0\t137 37\n",
      "10\t238\t1.0\t1\t196 42\n",
      "11\t121\t0.2\t1\t101 20\n",
      "12\t165\t0.1\t1\t163 2\n",
      "13\t111\t0.0\t1\t104 7\n",
      "14\t165\t0.0\t1\t162 3\n",
      "15\t86\t0.0\t1\t85 1\n",
      "16\t99\t0.0\t1\t92 7\n",
      "17\t138\t0.0\t1\t127 11\n",
      "18\t82\t0.0\t1\t79 3\n",
      "19\t138\t0.0\t1\t132 6\n",
      "20\t115\t0.0\t1\t113 2\n",
      "21\t16\t0.0\t1\t15 1\n",
      "22\t80\t0.0\t1\t77 3\n",
      "23\t67\t0.0\t1\t66 1\n",
      "24\t102\t0.0\t1\t100 2\n",
      "25\t47\t0.0\t1\t46 1\n",
      "26\t70\t0.0\t1\t68 2\n",
      "27\t27\t0.0\t1\t26 1\n",
      "28\t65\t0.0\t1\t65\n",
      "29\t14\t0.0\t1\t13 1\n",
      "30\t50\t0.0\t1\t49 1\n",
      "31\t10\t0.0\t1\t9 1\n",
      "32\t50\t0.0\t1\t49 1\n",
      "33\t16\t0.0\t1\t16\n",
      "34\t10\t0.0\t1\t10\n",
      "35\t28\t0.0\t1\t26 2\n",
      "36\t14\t0.0\t1\t14\n",
      "37\t16\t0.0\t1\t15 1\n",
      "38\t13\t0.0\t1\t13\n",
      "39\t13\t0.0\t1\t13\n",
      "40\t9\t0.0\t1\t8 1\n",
      "41\t15\t0.0\t1\t15\n",
      "42\t11\t0.0\t1\t11\n",
      "43\t2\t0.0\t1\t2\n",
      "44\t12\t0.0\t1\t12\n",
      "45\t8\t0.0\t1\t8\n",
      "47\t1\t0.0\t1\t1\n",
      "48\t4\t0.0\t1\t4\n",
      "49\t1\t0.0\t1\t1\n",
      "51\t1\t0.0\t1\t0 1\n",
      "53\t1\t0.0\t1\t1\n",
      "57\t2\t0.0\t1\t2\n",
      "58\t1\t0.0\t1\t1\n",
      "61\t1\t0.0\t1\t1\n",
      "62\t1\t0.0\t1\t1\n",
      "66\t2\t0.0\t1\t0 2\n",
      "69\t1\t0.0\t1\t0 1\n",
      "71\t1\t0.0\t1\t0 1\n",
      "72\t2\t0.0\t1\t0 2\n",
      "73\t4\t0.0\t1\t0 4\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022877_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022877_1_trimmed.fq.gz and ERR022877_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022877_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022877_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022877_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022877_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022877_R1_trimmed.fq.gz, file_2: ERR022877_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022877_R1_trimmed.fq.gz and ERR022877_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022877_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022877_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 13550 (1.35%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022877_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022877_val_1.fq.gz\n",
      "Approx 5% complete for ERR022877_val_1.fq.gz\n",
      "Approx 10% complete for ERR022877_val_1.fq.gz\n",
      "Approx 15% complete for ERR022877_val_1.fq.gz\n",
      "Approx 20% complete for ERR022877_val_1.fq.gz\n",
      "Approx 25% complete for ERR022877_val_1.fq.gz\n",
      "Approx 30% complete for ERR022877_val_1.fq.gz\n",
      "Approx 35% complete for ERR022877_val_1.fq.gz\n",
      "Approx 40% complete for ERR022877_val_1.fq.gz\n",
      "Approx 45% complete for ERR022877_val_1.fq.gz\n",
      "Approx 50% complete for ERR022877_val_1.fq.gz\n",
      "Approx 55% complete for ERR022877_val_1.fq.gz\n",
      "Approx 60% complete for ERR022877_val_1.fq.gz\n",
      "Approx 65% complete for ERR022877_val_1.fq.gz\n",
      "Approx 70% complete for ERR022877_val_1.fq.gz\n",
      "Approx 75% complete for ERR022877_val_1.fq.gz\n",
      "Approx 80% complete for ERR022877_val_1.fq.gz\n",
      "Approx 85% complete for ERR022877_val_1.fq.gz\n",
      "Approx 90% complete for ERR022877_val_1.fq.gz\n",
      "Approx 95% complete for ERR022877_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022877_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022877_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022877_val_2.fq.gz\n",
      "Approx 5% complete for ERR022877_val_2.fq.gz\n",
      "Approx 10% complete for ERR022877_val_2.fq.gz\n",
      "Approx 15% complete for ERR022877_val_2.fq.gz\n",
      "Approx 20% complete for ERR022877_val_2.fq.gz\n",
      "Approx 25% complete for ERR022877_val_2.fq.gz\n",
      "Approx 30% complete for ERR022877_val_2.fq.gz\n",
      "Approx 35% complete for ERR022877_val_2.fq.gz\n",
      "Approx 40% complete for ERR022877_val_2.fq.gz\n",
      "Approx 45% complete for ERR022877_val_2.fq.gz\n",
      "Approx 50% complete for ERR022877_val_2.fq.gz\n",
      "Approx 55% complete for ERR022877_val_2.fq.gz\n",
      "Approx 60% complete for ERR022877_val_2.fq.gz\n",
      "Approx 65% complete for ERR022877_val_2.fq.gz\n",
      "Approx 70% complete for ERR022877_val_2.fq.gz\n",
      "Approx 75% complete for ERR022877_val_2.fq.gz\n",
      "Approx 80% complete for ERR022877_val_2.fq.gz\n",
      "Approx 85% complete for ERR022877_val_2.fq.gz\n",
      "Approx 90% complete for ERR022877_val_2.fq.gz\n",
      "Approx 95% complete for ERR022877_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022877_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022877_R1_trimmed.fq.gz and ERR022877_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022878<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022878_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t9632\tAGATCGGAAGAGC\t1000000\t0.96\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 9632). Second best hit was Nextera (count: 1)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022878_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022878_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022878_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022878_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 27.045 s (27.045 s/read; 2.22 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   356,462 (35.6%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:                 657,829 bp (0.9%)\n",
      "Total written (filtered):     74,579,800 bp (98.1%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 356462 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.0%\n",
      "  C: 26.2%\n",
      "  G: 18.5%\n",
      "  T: 23.2%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t257507\t250000.0\t0\t257507\n",
      "2\t57936\t62500.0\t0\t57936\n",
      "3\t19356\t15625.0\t0\t19356\n",
      "4\t5151\t3906.2\t0\t5151\n",
      "5\t2033\t976.6\t0\t2033\n",
      "6\t855\t244.1\t0\t855\n",
      "7\t648\t61.0\t0\t648\n",
      "8\t859\t15.3\t0\t859\n",
      "9\t418\t3.8\t0\t393 25\n",
      "10\t717\t1.0\t1\t683 34\n",
      "11\t542\t0.2\t1\t527 15\n",
      "12\t703\t0.1\t1\t688 15\n",
      "13\t498\t0.0\t1\t490 8\n",
      "14\t553\t0.0\t1\t539 14\n",
      "15\t559\t0.0\t1\t541 18\n",
      "16\t473\t0.0\t1\t461 12\n",
      "17\t506\t0.0\t1\t497 9\n",
      "18\t485\t0.0\t1\t480 5\n",
      "19\t345\t0.0\t1\t340 5\n",
      "20\t427\t0.0\t1\t418 9\n",
      "21\t517\t0.0\t1\t508 9\n",
      "22\t288\t0.0\t1\t283 5\n",
      "23\t410\t0.0\t1\t399 11\n",
      "24\t351\t0.0\t1\t341 10\n",
      "25\t374\t0.0\t1\t365 9\n",
      "26\t299\t0.0\t1\t296 3\n",
      "27\t286\t0.0\t1\t273 13\n",
      "28\t279\t0.0\t1\t274 5\n",
      "29\t322\t0.0\t1\t313 9\n",
      "30\t234\t0.0\t1\t225 9\n",
      "31\t241\t0.0\t1\t235 6\n",
      "32\t189\t0.0\t1\t188 1\n",
      "33\t212\t0.0\t1\t207 5\n",
      "34\t152\t0.0\t1\t148 4\n",
      "35\t186\t0.0\t1\t182 4\n",
      "36\t132\t0.0\t1\t123 9\n",
      "37\t166\t0.0\t1\t160 6\n",
      "38\t192\t0.0\t1\t189 3\n",
      "39\t147\t0.0\t1\t146 1\n",
      "40\t21\t0.0\t1\t21\n",
      "41\t65\t0.0\t1\t65\n",
      "42\t50\t0.0\t1\t46 4\n",
      "43\t51\t0.0\t1\t50 1\n",
      "44\t48\t0.0\t1\t46 2\n",
      "45\t38\t0.0\t1\t36 2\n",
      "46\t28\t0.0\t1\t27 1\n",
      "47\t23\t0.0\t1\t22 1\n",
      "48\t25\t0.0\t1\t24 1\n",
      "49\t24\t0.0\t1\t22 2\n",
      "50\t10\t0.0\t1\t9 1\n",
      "51\t5\t0.0\t1\t5\n",
      "52\t3\t0.0\t1\t2 1\n",
      "53\t5\t0.0\t1\t2 3\n",
      "54\t7\t0.0\t1\t6 1\n",
      "55\t3\t0.0\t1\t0 3\n",
      "56\t7\t0.0\t1\t7\n",
      "57\t10\t0.0\t1\t10\n",
      "58\t2\t0.0\t1\t0 2\n",
      "59\t4\t0.0\t1\t0 4\n",
      "60\t1\t0.0\t1\t0 1\n",
      "61\t2\t0.0\t1\t1 1\n",
      "64\t6\t0.0\t1\t1 5\n",
      "65\t14\t0.0\t1\t2 12\n",
      "66\t5\t0.0\t1\t0 5\n",
      "67\t10\t0.0\t1\t1 9\n",
      "68\t22\t0.0\t1\t0 22\n",
      "69\t8\t0.0\t1\t0 8\n",
      "70\t1\t0.0\t1\t0 1\n",
      "72\t6\t0.0\t1\t0 6\n",
      "73\t5\t0.0\t1\t0 5\n",
      "74\t7\t0.0\t1\t0 7\n",
      "75\t43\t0.0\t1\t0 43\n",
      "76\t355\t0.0\t1\t0 355\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022878_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022878_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022878_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022878_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022878_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.458 s (26.458 s/read; 2.27 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   358,839 (35.9%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               1,100,921 bp (1.4%)\n",
      "Total written (filtered):     74,175,536 bp (97.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 358839 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.7%\n",
      "  C: 26.2%\n",
      "  G: 18.6%\n",
      "  T: 22.4%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t260795\t250000.0\t0\t260795\n",
      "2\t57563\t62500.0\t0\t57563\n",
      "3\t19387\t15625.0\t0\t19387\n",
      "4\t4931\t3906.2\t0\t4931\n",
      "5\t2228\t976.6\t0\t2228\n",
      "6\t881\t244.1\t0\t881\n",
      "7\t630\t61.0\t0\t630\n",
      "8\t700\t15.3\t0\t700\n",
      "9\t620\t3.8\t0\t599 21\n",
      "10\t770\t1.0\t1\t734 36\n",
      "11\t528\t0.2\t1\t507 21\n",
      "12\t744\t0.1\t1\t733 11\n",
      "13\t465\t0.0\t1\t458 7\n",
      "14\t635\t0.0\t1\t621 14\n",
      "15\t477\t0.0\t1\t465 12\n",
      "16\t390\t0.0\t1\t388 2\n",
      "17\t572\t0.0\t1\t566 6\n",
      "18\t408\t0.0\t1\t400 8\n",
      "19\t529\t0.0\t1\t519 10\n",
      "20\t541\t0.0\t1\t536 5\n",
      "21\t173\t0.0\t1\t172 1\n",
      "22\t407\t0.0\t1\t398 9\n",
      "23\t422\t0.0\t1\t410 12\n",
      "24\t429\t0.0\t1\t420 9\n",
      "25\t325\t0.0\t1\t322 3\n",
      "26\t399\t0.0\t1\t394 5\n",
      "27\t196\t0.0\t1\t187 9\n",
      "28\t387\t0.0\t1\t376 11\n",
      "29\t180\t0.0\t1\t170 10\n",
      "30\t350\t0.0\t1\t345 5\n",
      "31\t108\t0.0\t1\t104 4\n",
      "32\t263\t0.0\t1\t260 3\n",
      "33\t154\t0.0\t1\t153 1\n",
      "34\t100\t0.0\t1\t98 2\n",
      "35\t173\t0.0\t1\t171 2\n",
      "36\t133\t0.0\t1\t132 1\n",
      "37\t118\t0.0\t1\t116 2\n",
      "38\t145\t0.0\t1\t145\n",
      "39\t106\t0.0\t1\t105 1\n",
      "40\t75\t0.0\t1\t74 1\n",
      "41\t87\t0.0\t1\t86 1\n",
      "42\t81\t0.0\t1\t81\n",
      "43\t27\t0.0\t1\t27\n",
      "44\t59\t0.0\t1\t58 1\n",
      "45\t65\t0.0\t1\t63 2\n",
      "46\t7\t0.0\t1\t7\n",
      "47\t13\t0.0\t1\t13\n",
      "48\t14\t0.0\t1\t14\n",
      "49\t14\t0.0\t1\t14\n",
      "50\t4\t0.0\t1\t4\n",
      "51\t2\t0.0\t1\t2\n",
      "52\t3\t0.0\t1\t3\n",
      "54\t3\t0.0\t1\t3\n",
      "55\t3\t0.0\t1\t2 1\n",
      "57\t8\t0.0\t1\t7 1\n",
      "61\t1\t0.0\t1\t1\n",
      "62\t3\t0.0\t1\t3\n",
      "63\t2\t0.0\t1\t2\n",
      "64\t1\t0.0\t1\t1\n",
      "68\t1\t0.0\t1\t0 1\n",
      "73\t3\t0.0\t1\t0 3\n",
      "74\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022878_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022878_1_trimmed.fq.gz and ERR022878_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022878_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022878_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022878_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022878_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022878_R1_trimmed.fq.gz, file_2: ERR022878_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022878_R1_trimmed.fq.gz and ERR022878_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022878_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022878_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 9006 (0.90%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022878_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022878_val_1.fq.gz\n",
      "Approx 5% complete for ERR022878_val_1.fq.gz\n",
      "Approx 10% complete for ERR022878_val_1.fq.gz\n",
      "Approx 15% complete for ERR022878_val_1.fq.gz\n",
      "Approx 20% complete for ERR022878_val_1.fq.gz\n",
      "Approx 25% complete for ERR022878_val_1.fq.gz\n",
      "Approx 30% complete for ERR022878_val_1.fq.gz\n",
      "Approx 35% complete for ERR022878_val_1.fq.gz\n",
      "Approx 40% complete for ERR022878_val_1.fq.gz\n",
      "Approx 45% complete for ERR022878_val_1.fq.gz\n",
      "Approx 50% complete for ERR022878_val_1.fq.gz\n",
      "Approx 55% complete for ERR022878_val_1.fq.gz\n",
      "Approx 60% complete for ERR022878_val_1.fq.gz\n",
      "Approx 65% complete for ERR022878_val_1.fq.gz\n",
      "Approx 70% complete for ERR022878_val_1.fq.gz\n",
      "Approx 75% complete for ERR022878_val_1.fq.gz\n",
      "Approx 80% complete for ERR022878_val_1.fq.gz\n",
      "Approx 85% complete for ERR022878_val_1.fq.gz\n",
      "Approx 90% complete for ERR022878_val_1.fq.gz\n",
      "Approx 95% complete for ERR022878_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022878_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022878_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022878_val_2.fq.gz\n",
      "Approx 5% complete for ERR022878_val_2.fq.gz\n",
      "Approx 10% complete for ERR022878_val_2.fq.gz\n",
      "Approx 15% complete for ERR022878_val_2.fq.gz\n",
      "Approx 20% complete for ERR022878_val_2.fq.gz\n",
      "Approx 25% complete for ERR022878_val_2.fq.gz\n",
      "Approx 30% complete for ERR022878_val_2.fq.gz\n",
      "Approx 35% complete for ERR022878_val_2.fq.gz\n",
      "Approx 40% complete for ERR022878_val_2.fq.gz\n",
      "Approx 45% complete for ERR022878_val_2.fq.gz\n",
      "Approx 50% complete for ERR022878_val_2.fq.gz\n",
      "Approx 55% complete for ERR022878_val_2.fq.gz\n",
      "Approx 60% complete for ERR022878_val_2.fq.gz\n",
      "Approx 65% complete for ERR022878_val_2.fq.gz\n",
      "Approx 70% complete for ERR022878_val_2.fq.gz\n",
      "Approx 75% complete for ERR022878_val_2.fq.gz\n",
      "Approx 80% complete for ERR022878_val_2.fq.gz\n",
      "Approx 85% complete for ERR022878_val_2.fq.gz\n",
      "Approx 90% complete for ERR022878_val_2.fq.gz\n",
      "Approx 95% complete for ERR022878_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022878_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022878_R1_trimmed.fq.gz and ERR022878_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022879<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022879_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t204\tAGATCGGAAGAGC\t1000000\t0.02\n",
      "Nextera\t3\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 204). Second best hit was Nextera (count: 3)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022879_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022879_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022879_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022879_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.470 s (26.470 s/read; 2.27 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   349,920 (35.0%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               2,482,046 bp (3.3%)\n",
      "Total written (filtered):     73,056,226 bp (96.1%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 349920 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 36.1%\n",
      "  C: 24.5%\n",
      "  G: 15.3%\n",
      "  T: 24.1%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t272232\t250000.0\t0\t272232\n",
      "2\t56087\t62500.0\t0\t56087\n",
      "3\t15646\t15625.0\t0\t15646\n",
      "4\t4446\t3906.2\t0\t4446\n",
      "5\t928\t976.6\t0\t928\n",
      "6\t209\t244.1\t0\t209\n",
      "7\t57\t61.0\t0\t57\n",
      "8\t22\t15.3\t0\t22\n",
      "9\t37\t3.8\t0\t11 26\n",
      "10\t30\t1.0\t1\t10 20\n",
      "11\t11\t0.2\t1\t6 5\n",
      "12\t18\t0.1\t1\t17 1\n",
      "13\t3\t0.0\t1\t3\n",
      "14\t8\t0.0\t1\t4 4\n",
      "15\t25\t0.0\t1\t24 1\n",
      "16\t11\t0.0\t1\t10 1\n",
      "17\t10\t0.0\t1\t10\n",
      "18\t10\t0.0\t1\t10\n",
      "19\t8\t0.0\t1\t8\n",
      "20\t9\t0.0\t1\t9\n",
      "21\t8\t0.0\t1\t7 1\n",
      "22\t3\t0.0\t1\t3\n",
      "23\t5\t0.0\t1\t5\n",
      "24\t14\t0.0\t1\t14\n",
      "25\t5\t0.0\t1\t5\n",
      "26\t6\t0.0\t1\t6\n",
      "27\t3\t0.0\t1\t3\n",
      "28\t7\t0.0\t1\t7\n",
      "29\t4\t0.0\t1\t4\n",
      "30\t4\t0.0\t1\t4\n",
      "31\t4\t0.0\t1\t4\n",
      "32\t1\t0.0\t1\t1\n",
      "33\t5\t0.0\t1\t4 1\n",
      "34\t4\t0.0\t1\t4\n",
      "35\t3\t0.0\t1\t3\n",
      "36\t2\t0.0\t1\t1 1\n",
      "37\t6\t0.0\t1\t5 1\n",
      "38\t10\t0.0\t1\t7 3\n",
      "39\t3\t0.0\t1\t3\n",
      "40\t3\t0.0\t1\t2 1\n",
      "42\t1\t0.0\t1\t0 1\n",
      "44\t1\t0.0\t1\t0 1\n",
      "50\t1\t0.0\t1\t0 1\n",
      "52\t1\t0.0\t1\t0 1\n",
      "53\t1\t0.0\t1\t0 1\n",
      "65\t2\t0.0\t1\t0 2\n",
      "75\t2\t0.0\t1\t0 2\n",
      "76\t4\t0.0\t1\t0 4\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022879_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022879_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022879_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022879_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022879_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 24.583 s (24.583 s/read; 2.44 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   355,302 (35.5%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               3,776,501 bp (5.0%)\n",
      "Total written (filtered):     71,750,775 bp (94.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 355302 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 36.3%\n",
      "  C: 23.8%\n",
      "  G: 15.6%\n",
      "  T: 24.1%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t273476\t250000.0\t0\t273476\n",
      "2\t59246\t62500.0\t0\t59246\n",
      "3\t16403\t15625.0\t0\t16403\n",
      "4\t4620\t3906.2\t0\t4620\n",
      "5\t898\t976.6\t0\t898\n",
      "6\t207\t244.1\t0\t207\n",
      "7\t59\t61.0\t0\t59\n",
      "8\t28\t15.3\t0\t28\n",
      "9\t78\t3.8\t0\t14 64\n",
      "10\t29\t1.0\t1\t14 15\n",
      "11\t15\t0.2\t1\t8 7\n",
      "12\t21\t0.1\t1\t20 1\n",
      "13\t5\t0.0\t1\t5\n",
      "14\t11\t0.0\t1\t10 1\n",
      "15\t15\t0.0\t1\t13 2\n",
      "16\t8\t0.0\t1\t8\n",
      "17\t34\t0.0\t1\t32 2\n",
      "18\t10\t0.0\t1\t10\n",
      "19\t25\t0.0\t1\t25\n",
      "20\t14\t0.0\t1\t13 1\n",
      "22\t7\t0.0\t1\t6 1\n",
      "23\t9\t0.0\t1\t7 2\n",
      "24\t16\t0.0\t1\t16\n",
      "25\t5\t0.0\t1\t5\n",
      "26\t9\t0.0\t1\t7 2\n",
      "27\t5\t0.0\t1\t5\n",
      "28\t7\t0.0\t1\t7\n",
      "30\t4\t0.0\t1\t3 1\n",
      "31\t1\t0.0\t1\t1\n",
      "32\t6\t0.0\t1\t5 1\n",
      "33\t2\t0.0\t1\t2\n",
      "34\t1\t0.0\t1\t1\n",
      "35\t3\t0.0\t1\t3\n",
      "37\t1\t0.0\t1\t1\n",
      "38\t2\t0.0\t1\t2\n",
      "39\t2\t0.0\t1\t1 1\n",
      "40\t1\t0.0\t1\t1\n",
      "41\t2\t0.0\t1\t2\n",
      "42\t1\t0.0\t1\t1\n",
      "43\t1\t0.0\t1\t1\n",
      "44\t1\t0.0\t1\t1\n",
      "45\t1\t0.0\t1\t1\n",
      "46\t2\t0.0\t1\t2\n",
      "47\t1\t0.0\t1\t1\n",
      "51\t1\t0.0\t1\t1\n",
      "54\t3\t0.0\t1\t1 2\n",
      "60\t1\t0.0\t1\t1\n",
      "62\t1\t0.0\t1\t1\n",
      "63\t1\t0.0\t1\t1\n",
      "65\t1\t0.0\t1\t0 1\n",
      "66\t2\t0.0\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022879_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022879_1_trimmed.fq.gz and ERR022879_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022879_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022879_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022879_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022879_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022879_R1_trimmed.fq.gz, file_2: ERR022879_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022879_R1_trimmed.fq.gz and ERR022879_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022879_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022879_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 30099 (3.01%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022879_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022879_val_1.fq.gz\n",
      "Approx 5% complete for ERR022879_val_1.fq.gz\n",
      "Approx 10% complete for ERR022879_val_1.fq.gz\n",
      "Approx 15% complete for ERR022879_val_1.fq.gz\n",
      "Approx 20% complete for ERR022879_val_1.fq.gz\n",
      "Approx 25% complete for ERR022879_val_1.fq.gz\n",
      "Approx 30% complete for ERR022879_val_1.fq.gz\n",
      "Approx 35% complete for ERR022879_val_1.fq.gz\n",
      "Approx 40% complete for ERR022879_val_1.fq.gz\n",
      "Approx 45% complete for ERR022879_val_1.fq.gz\n",
      "Approx 50% complete for ERR022879_val_1.fq.gz\n",
      "Approx 55% complete for ERR022879_val_1.fq.gz\n",
      "Approx 60% complete for ERR022879_val_1.fq.gz\n",
      "Approx 65% complete for ERR022879_val_1.fq.gz\n",
      "Approx 70% complete for ERR022879_val_1.fq.gz\n",
      "Approx 75% complete for ERR022879_val_1.fq.gz\n",
      "Approx 80% complete for ERR022879_val_1.fq.gz\n",
      "Approx 85% complete for ERR022879_val_1.fq.gz\n",
      "Approx 90% complete for ERR022879_val_1.fq.gz\n",
      "Approx 95% complete for ERR022879_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022879_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022879_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022879_val_2.fq.gz\n",
      "Approx 5% complete for ERR022879_val_2.fq.gz\n",
      "Approx 10% complete for ERR022879_val_2.fq.gz\n",
      "Approx 15% complete for ERR022879_val_2.fq.gz\n",
      "Approx 20% complete for ERR022879_val_2.fq.gz\n",
      "Approx 25% complete for ERR022879_val_2.fq.gz\n",
      "Approx 30% complete for ERR022879_val_2.fq.gz\n",
      "Approx 35% complete for ERR022879_val_2.fq.gz\n",
      "Approx 40% complete for ERR022879_val_2.fq.gz\n",
      "Approx 45% complete for ERR022879_val_2.fq.gz\n",
      "Approx 50% complete for ERR022879_val_2.fq.gz\n",
      "Approx 55% complete for ERR022879_val_2.fq.gz\n",
      "Approx 60% complete for ERR022879_val_2.fq.gz\n",
      "Approx 65% complete for ERR022879_val_2.fq.gz\n",
      "Approx 70% complete for ERR022879_val_2.fq.gz\n",
      "Approx 75% complete for ERR022879_val_2.fq.gz\n",
      "Approx 80% complete for ERR022879_val_2.fq.gz\n",
      "Approx 85% complete for ERR022879_val_2.fq.gz\n",
      "Approx 90% complete for ERR022879_val_2.fq.gz\n",
      "Approx 95% complete for ERR022879_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022879_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022879_R1_trimmed.fq.gz and ERR022879_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022880<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022880_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t1749\tAGATCGGAAGAGC\t1000000\t0.17\n",
      "Nextera\t3\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t2\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 1749). Second best hit was Nextera (count: 3)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022880_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022880_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022880_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022880_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 27.402 s (27.402 s/read; 2.19 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   358,504 (35.9%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               1,827,851 bp (2.4%)\n",
      "Total written (filtered):     73,647,406 bp (96.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 358504 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.2%\n",
      "  C: 24.1%\n",
      "  G: 15.4%\n",
      "  T: 25.1%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t275868\t250000.0\t0\t275868\n",
      "2\t58249\t62500.0\t0\t58249\n",
      "3\t16180\t15625.0\t0\t16180\n",
      "4\t4727\t3906.2\t0\t4727\n",
      "5\t952\t976.6\t0\t952\n",
      "6\t238\t244.1\t0\t238\n",
      "7\t104\t61.0\t0\t104\n",
      "8\t83\t15.3\t0\t83\n",
      "9\t101\t3.8\t0\t59 42\n",
      "10\t127\t1.0\t1\t100 27\n",
      "11\t45\t0.2\t1\t39 6\n",
      "12\t91\t0.1\t1\t88 3\n",
      "13\t36\t0.0\t1\t36\n",
      "14\t66\t0.0\t1\t66\n",
      "15\t90\t0.0\t1\t88 2\n",
      "16\t50\t0.0\t1\t48 2\n",
      "17\t54\t0.0\t1\t54\n",
      "18\t37\t0.0\t1\t37\n",
      "19\t37\t0.0\t1\t36 1\n",
      "20\t49\t0.0\t1\t49\n",
      "21\t40\t0.0\t1\t39 1\n",
      "22\t34\t0.0\t1\t34\n",
      "23\t48\t0.0\t1\t44 4\n",
      "24\t54\t0.0\t1\t53 1\n",
      "25\t52\t0.0\t1\t51 1\n",
      "26\t67\t0.0\t1\t67\n",
      "27\t28\t0.0\t1\t24 4\n",
      "28\t39\t0.0\t1\t39\n",
      "29\t49\t0.0\t1\t48 1\n",
      "30\t30\t0.0\t1\t28 2\n",
      "31\t66\t0.0\t1\t65 1\n",
      "32\t33\t0.0\t1\t30 3\n",
      "33\t41\t0.0\t1\t39 2\n",
      "34\t32\t0.0\t1\t28 4\n",
      "35\t54\t0.0\t1\t46 8\n",
      "36\t69\t0.0\t1\t61 8\n",
      "37\t94\t0.0\t1\t81 13\n",
      "38\t120\t0.0\t1\t96 24\n",
      "39\t124\t0.0\t1\t120 4\n",
      "40\t9\t0.0\t1\t7 2\n",
      "41\t11\t0.0\t1\t7 4\n",
      "42\t9\t0.0\t1\t7 2\n",
      "43\t12\t0.0\t1\t11 1\n",
      "44\t5\t0.0\t1\t3 2\n",
      "45\t18\t0.0\t1\t8 10\n",
      "46\t2\t0.0\t1\t1 1\n",
      "47\t9\t0.0\t1\t8 1\n",
      "48\t5\t0.0\t1\t5\n",
      "49\t13\t0.0\t1\t7 6\n",
      "50\t29\t0.0\t1\t25 4\n",
      "51\t13\t0.0\t1\t9 4\n",
      "52\t17\t0.0\t1\t16 1\n",
      "53\t7\t0.0\t1\t3 4\n",
      "54\t12\t0.0\t1\t11 1\n",
      "55\t7\t0.0\t1\t4 3\n",
      "56\t16\t0.0\t1\t16\n",
      "57\t18\t0.0\t1\t18\n",
      "58\t1\t0.0\t1\t0 1\n",
      "59\t1\t0.0\t1\t0 1\n",
      "60\t2\t0.0\t1\t0 2\n",
      "61\t1\t0.0\t1\t1\n",
      "62\t1\t0.0\t1\t0 1\n",
      "63\t1\t0.0\t1\t0 1\n",
      "64\t1\t0.0\t1\t0 1\n",
      "65\t4\t0.0\t1\t0 4\n",
      "67\t2\t0.0\t1\t0 2\n",
      "68\t4\t0.0\t1\t0 4\n",
      "69\t2\t0.0\t1\t0 2\n",
      "70\t2\t0.0\t1\t0 2\n",
      "73\t2\t0.0\t1\t0 2\n",
      "74\t1\t0.0\t1\t0 1\n",
      "75\t1\t0.0\t1\t0 1\n",
      "76\t8\t0.0\t1\t0 8\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022880_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022880_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022880_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022880_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022880_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 23.861 s (23.861 s/read; 2.51 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   353,452 (35.3%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               5,452,035 bp (7.2%)\n",
      "Total written (filtered):     70,056,859 bp (92.2%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 353452 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.8%\n",
      "  C: 24.5%\n",
      "  G: 15.2%\n",
      "  T: 25.2%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t271256\t250000.0\t0\t271256\n",
      "2\t59562\t62500.0\t0\t59562\n",
      "3\t15591\t15625.0\t0\t15591\n",
      "4\t4320\t3906.2\t0\t4320\n",
      "5\t928\t976.6\t0\t928\n",
      "6\t260\t244.1\t0\t260\n",
      "7\t96\t61.0\t0\t96\n",
      "8\t68\t15.3\t0\t68\n",
      "9\t111\t3.8\t0\t68 43\n",
      "10\t115\t1.0\t1\t85 30\n",
      "11\t48\t0.2\t1\t32 16\n",
      "12\t78\t0.1\t1\t74 4\n",
      "13\t54\t0.0\t1\t51 3\n",
      "14\t80\t0.0\t1\t74 6\n",
      "15\t31\t0.0\t1\t30 1\n",
      "16\t39\t0.0\t1\t39\n",
      "17\t80\t0.0\t1\t72 8\n",
      "18\t37\t0.0\t1\t34 3\n",
      "19\t72\t0.0\t1\t69 3\n",
      "20\t51\t0.0\t1\t51\n",
      "21\t9\t0.0\t1\t7 2\n",
      "22\t35\t0.0\t1\t32 3\n",
      "23\t42\t0.0\t1\t37 5\n",
      "24\t77\t0.0\t1\t76 1\n",
      "25\t28\t0.0\t1\t28\n",
      "26\t44\t0.0\t1\t41 3\n",
      "27\t15\t0.0\t1\t15\n",
      "28\t48\t0.0\t1\t48\n",
      "29\t5\t0.0\t1\t4 1\n",
      "30\t37\t0.0\t1\t36 1\n",
      "31\t5\t0.0\t1\t5\n",
      "32\t35\t0.0\t1\t34 1\n",
      "33\t9\t0.0\t1\t8 1\n",
      "34\t9\t0.0\t1\t8 1\n",
      "35\t18\t0.0\t1\t18\n",
      "36\t17\t0.0\t1\t17\n",
      "37\t17\t0.0\t1\t16 1\n",
      "38\t15\t0.0\t1\t15\n",
      "39\t14\t0.0\t1\t14\n",
      "40\t7\t0.0\t1\t7\n",
      "41\t9\t0.0\t1\t8 1\n",
      "42\t9\t0.0\t1\t9\n",
      "43\t3\t0.0\t1\t3\n",
      "44\t5\t0.0\t1\t5\n",
      "45\t9\t0.0\t1\t9\n",
      "47\t1\t0.0\t1\t1\n",
      "48\t2\t0.0\t1\t2\n",
      "49\t1\t0.0\t1\t1\n",
      "50\t1\t0.0\t1\t1\n",
      "51\t2\t0.0\t1\t2\n",
      "54\t1\t0.0\t1\t1\n",
      "56\t3\t0.0\t1\t3\n",
      "57\t30\t0.0\t1\t24 6\n",
      "58\t5\t0.0\t1\t5\n",
      "59\t1\t0.0\t1\t1\n",
      "60\t3\t0.0\t1\t3\n",
      "61\t1\t0.0\t1\t1\n",
      "62\t3\t0.0\t1\t3\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022880_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022880_1_trimmed.fq.gz and ERR022880_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022880_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022880_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022880_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022880_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022880_R1_trimmed.fq.gz, file_2: ERR022880_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022880_R1_trimmed.fq.gz and ERR022880_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022880_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022880_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 23858 (2.39%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022880_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022880_val_1.fq.gz\n",
      "Approx 5% complete for ERR022880_val_1.fq.gz\n",
      "Approx 10% complete for ERR022880_val_1.fq.gz\n",
      "Approx 15% complete for ERR022880_val_1.fq.gz\n",
      "Approx 20% complete for ERR022880_val_1.fq.gz\n",
      "Approx 25% complete for ERR022880_val_1.fq.gz\n",
      "Approx 30% complete for ERR022880_val_1.fq.gz\n",
      "Approx 35% complete for ERR022880_val_1.fq.gz\n",
      "Approx 40% complete for ERR022880_val_1.fq.gz\n",
      "Approx 45% complete for ERR022880_val_1.fq.gz\n",
      "Approx 50% complete for ERR022880_val_1.fq.gz\n",
      "Approx 55% complete for ERR022880_val_1.fq.gz\n",
      "Approx 60% complete for ERR022880_val_1.fq.gz\n",
      "Approx 65% complete for ERR022880_val_1.fq.gz\n",
      "Approx 70% complete for ERR022880_val_1.fq.gz\n",
      "Approx 75% complete for ERR022880_val_1.fq.gz\n",
      "Approx 80% complete for ERR022880_val_1.fq.gz\n",
      "Approx 85% complete for ERR022880_val_1.fq.gz\n",
      "Approx 90% complete for ERR022880_val_1.fq.gz\n",
      "Approx 95% complete for ERR022880_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022880_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022880_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022880_val_2.fq.gz\n",
      "Approx 5% complete for ERR022880_val_2.fq.gz\n",
      "Approx 10% complete for ERR022880_val_2.fq.gz\n",
      "Approx 15% complete for ERR022880_val_2.fq.gz\n",
      "Approx 20% complete for ERR022880_val_2.fq.gz\n",
      "Approx 25% complete for ERR022880_val_2.fq.gz\n",
      "Approx 30% complete for ERR022880_val_2.fq.gz\n",
      "Approx 35% complete for ERR022880_val_2.fq.gz\n",
      "Approx 40% complete for ERR022880_val_2.fq.gz\n",
      "Approx 45% complete for ERR022880_val_2.fq.gz\n",
      "Approx 50% complete for ERR022880_val_2.fq.gz\n",
      "Approx 55% complete for ERR022880_val_2.fq.gz\n",
      "Approx 60% complete for ERR022880_val_2.fq.gz\n",
      "Approx 65% complete for ERR022880_val_2.fq.gz\n",
      "Approx 70% complete for ERR022880_val_2.fq.gz\n",
      "Approx 75% complete for ERR022880_val_2.fq.gz\n",
      "Approx 80% complete for ERR022880_val_2.fq.gz\n",
      "Approx 85% complete for ERR022880_val_2.fq.gz\n",
      "Approx 90% complete for ERR022880_val_2.fq.gz\n",
      "Approx 95% complete for ERR022880_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022880_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022880_R1_trimmed.fq.gz and ERR022880_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022881<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022881_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t187\tAGATCGGAAGAGC\t1000000\t0.02\n",
      "smallRNA\t3\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 187). Second best hit was smallRNA (count: 3)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022881_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022881_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022881_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022881_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 26.653 s (26.653 s/read; 2.25 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   356,227 (35.6%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               1,789,103 bp (2.4%)\n",
      "Total written (filtered):     73,736,740 bp (97.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 356227 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.5%\n",
      "  C: 24.4%\n",
      "  G: 15.3%\n",
      "  T: 25.6%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t273607\t250000.0\t0\t273607\n",
      "2\t60624\t62500.0\t0\t60624\n",
      "3\t15973\t15625.0\t0\t15973\n",
      "4\t4665\t3906.2\t0\t4665\n",
      "5\t877\t976.6\t0\t877\n",
      "6\t146\t244.1\t0\t146\n",
      "7\t35\t61.0\t0\t35\n",
      "8\t13\t15.3\t0\t13\n",
      "9\t48\t3.8\t0\t4 44\n",
      "10\t30\t1.0\t1\t11 19\n",
      "11\t9\t0.2\t1\t3 6\n",
      "12\t7\t0.1\t1\t5 2\n",
      "13\t3\t0.0\t1\t3\n",
      "14\t5\t0.0\t1\t4 1\n",
      "15\t17\t0.0\t1\t15 2\n",
      "16\t2\t0.0\t1\t2\n",
      "17\t5\t0.0\t1\t5\n",
      "18\t3\t0.0\t1\t3\n",
      "19\t3\t0.0\t1\t2 1\n",
      "20\t4\t0.0\t1\t4\n",
      "21\t3\t0.0\t1\t2 1\n",
      "22\t5\t0.0\t1\t5\n",
      "23\t7\t0.0\t1\t6 1\n",
      "24\t6\t0.0\t1\t6\n",
      "25\t3\t0.0\t1\t3\n",
      "26\t6\t0.0\t1\t5 1\n",
      "27\t3\t0.0\t1\t3\n",
      "28\t5\t0.0\t1\t5\n",
      "29\t9\t0.0\t1\t8 1\n",
      "30\t5\t0.0\t1\t4 1\n",
      "31\t6\t0.0\t1\t6\n",
      "32\t2\t0.0\t1\t2\n",
      "33\t3\t0.0\t1\t3\n",
      "34\t4\t0.0\t1\t4\n",
      "35\t3\t0.0\t1\t3\n",
      "36\t7\t0.0\t1\t5 2\n",
      "37\t14\t0.0\t1\t12 2\n",
      "38\t9\t0.0\t1\t7 2\n",
      "39\t15\t0.0\t1\t15\n",
      "40\t2\t0.0\t1\t2\n",
      "41\t1\t0.0\t1\t0 1\n",
      "42\t1\t0.0\t1\t1\n",
      "43\t2\t0.0\t1\t2\n",
      "44\t1\t0.0\t1\t1\n",
      "45\t3\t0.0\t1\t3\n",
      "46\t1\t0.0\t1\t1\n",
      "47\t1\t0.0\t1\t0 1\n",
      "49\t1\t0.0\t1\t1\n",
      "50\t2\t0.0\t1\t2\n",
      "51\t2\t0.0\t1\t2\n",
      "52\t4\t0.0\t1\t3 1\n",
      "53\t1\t0.0\t1\t0 1\n",
      "54\t1\t0.0\t1\t1\n",
      "56\t5\t0.0\t1\t4 1\n",
      "57\t2\t0.0\t1\t2\n",
      "61\t1\t0.0\t1\t0 1\n",
      "64\t1\t0.0\t1\t1\n",
      "65\t1\t0.0\t1\t0 1\n",
      "75\t1\t0.0\t1\t0 1\n",
      "76\t2\t0.0\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022881_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022881_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022881_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022881_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022881_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 24.564 s (24.564 s/read; 2.44 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   347,272 (34.7%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               5,803,431 bp (7.6%)\n",
      "Total written (filtered):     69,736,529 bp (91.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 347272 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.0%\n",
      "  C: 24.8%\n",
      "  G: 15.2%\n",
      "  T: 25.7%\n",
      "  none/other: 0.3%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t266053\t250000.0\t0\t266053\n",
      "2\t60962\t62500.0\t0\t60962\n",
      "3\t14904\t15625.0\t0\t14904\n",
      "4\t3991\t3906.2\t0\t3991\n",
      "5\t880\t976.6\t0\t880\n",
      "6\t184\t244.1\t0\t184\n",
      "7\t35\t61.0\t0\t35\n",
      "8\t14\t15.3\t0\t14\n",
      "9\t60\t3.8\t0\t5 55\n",
      "10\t30\t1.0\t1\t10 20\n",
      "11\t10\t0.2\t1\t3 7\n",
      "12\t6\t0.1\t1\t5 1\n",
      "13\t4\t0.0\t1\t3 1\n",
      "14\t9\t0.0\t1\t6 3\n",
      "15\t6\t0.0\t1\t6\n",
      "16\t1\t0.0\t1\t0 1\n",
      "17\t7\t0.0\t1\t6 1\n",
      "18\t4\t0.0\t1\t2 2\n",
      "19\t7\t0.0\t1\t6 1\n",
      "20\t1\t0.0\t1\t1\n",
      "21\t3\t0.0\t1\t3\n",
      "22\t9\t0.0\t1\t9\n",
      "23\t7\t0.0\t1\t7\n",
      "24\t6\t0.0\t1\t6\n",
      "25\t1\t0.0\t1\t1\n",
      "26\t2\t0.0\t1\t2\n",
      "27\t3\t0.0\t1\t2 1\n",
      "28\t7\t0.0\t1\t6 1\n",
      "29\t2\t0.0\t1\t2\n",
      "30\t5\t0.0\t1\t4 1\n",
      "32\t2\t0.0\t1\t1 1\n",
      "34\t4\t0.0\t1\t3 1\n",
      "35\t3\t0.0\t1\t3\n",
      "36\t4\t0.0\t1\t3 1\n",
      "37\t2\t0.0\t1\t2\n",
      "39\t1\t0.0\t1\t1\n",
      "40\t2\t0.0\t1\t2\n",
      "41\t1\t0.0\t1\t1\n",
      "42\t3\t0.0\t1\t3\n",
      "43\t4\t0.0\t1\t2 2\n",
      "44\t1\t0.0\t1\t1\n",
      "45\t4\t0.0\t1\t4\n",
      "46\t3\t0.0\t1\t1 2\n",
      "49\t1\t0.0\t1\t0 1\n",
      "50\t2\t0.0\t1\t2\n",
      "51\t3\t0.0\t1\t3\n",
      "53\t1\t0.0\t1\t0 1\n",
      "54\t1\t0.0\t1\t1\n",
      "55\t1\t0.0\t1\t1\n",
      "56\t1\t0.0\t1\t0 1\n",
      "57\t5\t0.0\t1\t5\n",
      "58\t2\t0.0\t1\t2\n",
      "59\t3\t0.0\t1\t3\n",
      "60\t1\t0.0\t1\t1\n",
      "61\t2\t0.0\t1\t2\n",
      "62\t1\t0.0\t1\t0 1\n",
      "63\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022881_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022881_1_trimmed.fq.gz and ERR022881_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022881_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022881_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022881_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022881_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022881_R1_trimmed.fq.gz, file_2: ERR022881_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022881_R1_trimmed.fq.gz and ERR022881_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022881_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022881_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 36235 (3.62%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022881_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022881_val_1.fq.gz\n",
      "Approx 5% complete for ERR022881_val_1.fq.gz\n",
      "Approx 10% complete for ERR022881_val_1.fq.gz\n",
      "Approx 15% complete for ERR022881_val_1.fq.gz\n",
      "Approx 20% complete for ERR022881_val_1.fq.gz\n",
      "Approx 25% complete for ERR022881_val_1.fq.gz\n",
      "Approx 30% complete for ERR022881_val_1.fq.gz\n",
      "Approx 35% complete for ERR022881_val_1.fq.gz\n",
      "Approx 40% complete for ERR022881_val_1.fq.gz\n",
      "Approx 45% complete for ERR022881_val_1.fq.gz\n",
      "Approx 50% complete for ERR022881_val_1.fq.gz\n",
      "Approx 55% complete for ERR022881_val_1.fq.gz\n",
      "Approx 60% complete for ERR022881_val_1.fq.gz\n",
      "Approx 65% complete for ERR022881_val_1.fq.gz\n",
      "Approx 70% complete for ERR022881_val_1.fq.gz\n",
      "Approx 75% complete for ERR022881_val_1.fq.gz\n",
      "Approx 80% complete for ERR022881_val_1.fq.gz\n",
      "Approx 85% complete for ERR022881_val_1.fq.gz\n",
      "Approx 90% complete for ERR022881_val_1.fq.gz\n",
      "Approx 95% complete for ERR022881_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022881_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022881_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022881_val_2.fq.gz\n",
      "Approx 5% complete for ERR022881_val_2.fq.gz\n",
      "Approx 10% complete for ERR022881_val_2.fq.gz\n",
      "Approx 15% complete for ERR022881_val_2.fq.gz\n",
      "Approx 20% complete for ERR022881_val_2.fq.gz\n",
      "Approx 25% complete for ERR022881_val_2.fq.gz\n",
      "Approx 30% complete for ERR022881_val_2.fq.gz\n",
      "Approx 35% complete for ERR022881_val_2.fq.gz\n",
      "Approx 40% complete for ERR022881_val_2.fq.gz\n",
      "Approx 45% complete for ERR022881_val_2.fq.gz\n",
      "Approx 50% complete for ERR022881_val_2.fq.gz\n",
      "Approx 55% complete for ERR022881_val_2.fq.gz\n",
      "Approx 60% complete for ERR022881_val_2.fq.gz\n",
      "Approx 65% complete for ERR022881_val_2.fq.gz\n",
      "Approx 70% complete for ERR022881_val_2.fq.gz\n",
      "Approx 75% complete for ERR022881_val_2.fq.gz\n",
      "Approx 80% complete for ERR022881_val_2.fq.gz\n",
      "Approx 85% complete for ERR022881_val_2.fq.gz\n",
      "Approx 90% complete for ERR022881_val_2.fq.gz\n",
      "Approx 95% complete for ERR022881_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022881_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022881_R1_trimmed.fq.gz and ERR022881_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022882<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022882_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t108\tAGATCGGAAGAGC\t1000000\t0.01\n",
      "smallRNA\t3\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 108). Second best hit was smallRNA (count: 3)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022882_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022882_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022882_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022882_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 27.118 s (27.118 s/read; 2.21 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   362,940 (36.3%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               1,422,693 bp (1.9%)\n",
      "Total written (filtered):     74,099,913 bp (97.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 362940 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.9%\n",
      "  C: 23.9%\n",
      "  G: 14.8%\n",
      "  T: 26.3%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t281777\t250000.0\t0\t281777\n",
      "2\t58962\t62500.0\t0\t58962\n",
      "3\t16110\t15625.0\t0\t16110\n",
      "4\t4794\t3906.2\t0\t4794\n",
      "5\t917\t976.6\t0\t917\n",
      "6\t153\t244.1\t0\t153\n",
      "7\t32\t61.0\t0\t32\n",
      "8\t10\t15.3\t0\t10\n",
      "9\t46\t3.8\t0\t7 39\n",
      "10\t28\t1.0\t1\t2 26\n",
      "11\t10\t0.2\t1\t1 9\n",
      "12\t2\t0.1\t1\t1 1\n",
      "13\t2\t0.0\t1\t2\n",
      "14\t2\t0.0\t1\t2\n",
      "15\t3\t0.0\t1\t3\n",
      "16\t2\t0.0\t1\t2\n",
      "17\t3\t0.0\t1\t2 1\n",
      "18\t2\t0.0\t1\t2\n",
      "19\t1\t0.0\t1\t1\n",
      "20\t4\t0.0\t1\t3 1\n",
      "21\t2\t0.0\t1\t1 1\n",
      "22\t1\t0.0\t1\t1\n",
      "24\t1\t0.0\t1\t0 1\n",
      "25\t2\t0.0\t1\t2\n",
      "26\t3\t0.0\t1\t2 1\n",
      "27\t3\t0.0\t1\t2 1\n",
      "28\t2\t0.0\t1\t2\n",
      "29\t3\t0.0\t1\t3\n",
      "30\t4\t0.0\t1\t4\n",
      "31\t2\t0.0\t1\t2\n",
      "32\t4\t0.0\t1\t4\n",
      "33\t1\t0.0\t1\t1\n",
      "34\t3\t0.0\t1\t3\n",
      "36\t3\t0.0\t1\t3\n",
      "37\t7\t0.0\t1\t7\n",
      "38\t9\t0.0\t1\t9\n",
      "39\t7\t0.0\t1\t7\n",
      "41\t2\t0.0\t1\t1 1\n",
      "42\t1\t0.0\t1\t1\n",
      "44\t1\t0.0\t1\t1\n",
      "46\t1\t0.0\t1\t1\n",
      "47\t1\t0.0\t1\t1\n",
      "48\t1\t0.0\t1\t0 1\n",
      "49\t1\t0.0\t1\t0 1\n",
      "50\t1\t0.0\t1\t1\n",
      "52\t1\t0.0\t1\t1\n",
      "58\t1\t0.0\t1\t0 1\n",
      "60\t1\t0.0\t1\t0 1\n",
      "62\t1\t0.0\t1\t0 1\n",
      "64\t1\t0.0\t1\t1\n",
      "65\t1\t0.0\t1\t0 1\n",
      "73\t3\t0.0\t1\t0 3\n",
      "76\t5\t0.0\t1\t0 5\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022882_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022882_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022882_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022882_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022882_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 24.486 s (24.486 s/read; 2.45 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   358,949 (35.9%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               3,625,276 bp (4.8%)\n",
      "Total written (filtered):     71,901,757 bp (94.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 358949 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.0%\n",
      "  C: 23.8%\n",
      "  G: 14.8%\n",
      "  T: 26.2%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t277035\t250000.0\t0\t277035\n",
      "2\t60354\t62500.0\t0\t60354\n",
      "3\t15816\t15625.0\t0\t15816\n",
      "4\t4494\t3906.2\t0\t4494\n",
      "5\t876\t976.6\t0\t876\n",
      "6\t139\t244.1\t0\t139\n",
      "7\t35\t61.0\t0\t35\n",
      "8\t11\t15.3\t0\t11\n",
      "9\t48\t3.8\t0\t2 46\n",
      "10\t18\t1.0\t1\t6 12\n",
      "11\t7\t0.2\t1\t1 6\n",
      "12\t2\t0.1\t1\t1 1\n",
      "13\t5\t0.0\t1\t3 2\n",
      "14\t7\t0.0\t1\t5 2\n",
      "15\t3\t0.0\t1\t3\n",
      "16\t1\t0.0\t1\t0 1\n",
      "17\t5\t0.0\t1\t5\n",
      "18\t1\t0.0\t1\t1\n",
      "19\t7\t0.0\t1\t7\n",
      "20\t4\t0.0\t1\t4\n",
      "21\t3\t0.0\t1\t2 1\n",
      "22\t4\t0.0\t1\t4\n",
      "23\t4\t0.0\t1\t1 3\n",
      "24\t5\t0.0\t1\t5\n",
      "25\t3\t0.0\t1\t2 1\n",
      "26\t2\t0.0\t1\t1 1\n",
      "28\t6\t0.0\t1\t5 1\n",
      "30\t5\t0.0\t1\t4 1\n",
      "31\t2\t0.0\t1\t1 1\n",
      "32\t4\t0.0\t1\t2 2\n",
      "35\t2\t0.0\t1\t1 1\n",
      "36\t1\t0.0\t1\t0 1\n",
      "37\t4\t0.0\t1\t2 2\n",
      "38\t4\t0.0\t1\t3 1\n",
      "39\t1\t0.0\t1\t1\n",
      "40\t1\t0.0\t1\t1\n",
      "41\t2\t0.0\t1\t1 1\n",
      "42\t4\t0.0\t1\t3 1\n",
      "44\t4\t0.0\t1\t4\n",
      "45\t2\t0.0\t1\t2\n",
      "47\t2\t0.0\t1\t1 1\n",
      "48\t1\t0.0\t1\t1\n",
      "49\t1\t0.0\t1\t1\n",
      "51\t3\t0.0\t1\t3\n",
      "53\t1\t0.0\t1\t0 1\n",
      "54\t2\t0.0\t1\t2\n",
      "58\t2\t0.0\t1\t1 1\n",
      "60\t1\t0.0\t1\t0 1\n",
      "61\t2\t0.0\t1\t1 1\n",
      "62\t1\t0.0\t1\t1\n",
      "63\t1\t0.0\t1\t0 1\n",
      "73\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022882_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022882_1_trimmed.fq.gz and ERR022882_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022882_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022882_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022882_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022882_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022882_R1_trimmed.fq.gz, file_2: ERR022882_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022882_R1_trimmed.fq.gz and ERR022882_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022882_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022882_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 25790 (2.58%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022882_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022882_val_1.fq.gz\n",
      "Approx 5% complete for ERR022882_val_1.fq.gz\n",
      "Approx 10% complete for ERR022882_val_1.fq.gz\n",
      "Approx 15% complete for ERR022882_val_1.fq.gz\n",
      "Approx 20% complete for ERR022882_val_1.fq.gz\n",
      "Approx 25% complete for ERR022882_val_1.fq.gz\n",
      "Approx 30% complete for ERR022882_val_1.fq.gz\n",
      "Approx 35% complete for ERR022882_val_1.fq.gz\n",
      "Approx 40% complete for ERR022882_val_1.fq.gz\n",
      "Approx 45% complete for ERR022882_val_1.fq.gz\n",
      "Approx 50% complete for ERR022882_val_1.fq.gz\n",
      "Approx 55% complete for ERR022882_val_1.fq.gz\n",
      "Approx 60% complete for ERR022882_val_1.fq.gz\n",
      "Approx 65% complete for ERR022882_val_1.fq.gz\n",
      "Approx 70% complete for ERR022882_val_1.fq.gz\n",
      "Approx 75% complete for ERR022882_val_1.fq.gz\n",
      "Approx 80% complete for ERR022882_val_1.fq.gz\n",
      "Approx 85% complete for ERR022882_val_1.fq.gz\n",
      "Approx 90% complete for ERR022882_val_1.fq.gz\n",
      "Approx 95% complete for ERR022882_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022882_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022882_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022882_val_2.fq.gz\n",
      "Approx 5% complete for ERR022882_val_2.fq.gz\n",
      "Approx 10% complete for ERR022882_val_2.fq.gz\n",
      "Approx 15% complete for ERR022882_val_2.fq.gz\n",
      "Approx 20% complete for ERR022882_val_2.fq.gz\n",
      "Approx 25% complete for ERR022882_val_2.fq.gz\n",
      "Approx 30% complete for ERR022882_val_2.fq.gz\n",
      "Approx 35% complete for ERR022882_val_2.fq.gz\n",
      "Approx 40% complete for ERR022882_val_2.fq.gz\n",
      "Approx 45% complete for ERR022882_val_2.fq.gz\n",
      "Approx 50% complete for ERR022882_val_2.fq.gz\n",
      "Approx 55% complete for ERR022882_val_2.fq.gz\n",
      "Approx 60% complete for ERR022882_val_2.fq.gz\n",
      "Approx 65% complete for ERR022882_val_2.fq.gz\n",
      "Approx 70% complete for ERR022882_val_2.fq.gz\n",
      "Approx 75% complete for ERR022882_val_2.fq.gz\n",
      "Approx 80% complete for ERR022882_val_2.fq.gz\n",
      "Approx 85% complete for ERR022882_val_2.fq.gz\n",
      "Approx 90% complete for ERR022882_val_2.fq.gz\n",
      "Approx 95% complete for ERR022882_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022882_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022882_R1_trimmed.fq.gz and ERR022882_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.8\n",
      "single-core operation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /Users/rmcolq/Work/git/pathbio3/analysis/Schistosoma_mansoni/qc/\n",
      "Using user-specified basename (>>ERR022883<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> data/Schistosoma_mansoni/subsampled/ERR022883_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t128\tAGATCGGAAGAGC\t1000000\t0.01\n",
      "Nextera\t2\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 128). Second best hit was Nextera (count: 2)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022883_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022883_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022883_1.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022883_1.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 27.138 s (27.138 s/read; 2.21 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   357,280 (35.7%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               2,038,368 bp (2.7%)\n",
      "Total written (filtered):     73,490,096 bp (96.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 357280 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.3%\n",
      "  C: 23.9%\n",
      "  G: 15.0%\n",
      "  T: 25.7%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t276722\t250000.0\t0\t276722\n",
      "2\t58353\t62500.0\t0\t58353\n",
      "3\t16170\t15625.0\t0\t16170\n",
      "4\t4700\t3906.2\t0\t4700\n",
      "5\t913\t976.6\t0\t913\n",
      "6\t172\t244.1\t0\t172\n",
      "7\t31\t61.0\t0\t31\n",
      "8\t13\t15.3\t0\t13\n",
      "9\t34\t3.8\t0\t7 27\n",
      "10\t28\t1.0\t1\t6 22\n",
      "11\t7\t0.2\t1\t3 4\n",
      "12\t5\t0.1\t1\t4 1\n",
      "13\t1\t0.0\t1\t1\n",
      "14\t6\t0.0\t1\t6\n",
      "15\t5\t0.0\t1\t5\n",
      "16\t7\t0.0\t1\t7\n",
      "17\t2\t0.0\t1\t1 1\n",
      "18\t4\t0.0\t1\t4\n",
      "19\t4\t0.0\t1\t4\n",
      "20\t6\t0.0\t1\t6\n",
      "21\t2\t0.0\t1\t1 1\n",
      "22\t6\t0.0\t1\t6\n",
      "23\t1\t0.0\t1\t1\n",
      "24\t5\t0.0\t1\t5\n",
      "25\t1\t0.0\t1\t0 1\n",
      "26\t4\t0.0\t1\t4\n",
      "27\t2\t0.0\t1\t2\n",
      "28\t4\t0.0\t1\t4\n",
      "29\t1\t0.0\t1\t1\n",
      "30\t3\t0.0\t1\t3\n",
      "31\t1\t0.0\t1\t1\n",
      "32\t1\t0.0\t1\t1\n",
      "33\t3\t0.0\t1\t3\n",
      "34\t5\t0.0\t1\t5\n",
      "35\t2\t0.0\t1\t2\n",
      "36\t2\t0.0\t1\t2\n",
      "37\t5\t0.0\t1\t5\n",
      "38\t20\t0.0\t1\t18 2\n",
      "39\t9\t0.0\t1\t8 1\n",
      "41\t1\t0.0\t1\t1\n",
      "43\t2\t0.0\t1\t1 1\n",
      "45\t1\t0.0\t1\t1\n",
      "46\t1\t0.0\t1\t1\n",
      "48\t1\t0.0\t1\t0 1\n",
      "49\t1\t0.0\t1\t1\n",
      "51\t1\t0.0\t1\t1\n",
      "52\t1\t0.0\t1\t1\n",
      "54\t1\t0.0\t1\t1\n",
      "55\t1\t0.0\t1\t0 1\n",
      "56\t5\t0.0\t1\t4 1\n",
      "57\t2\t0.0\t1\t1 1\n",
      "73\t1\t0.0\t1\t0 1\n",
      "76\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022883_1.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: data/Schistosoma_mansoni/subsampled/ERR022883_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.8\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.8). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to ERR022883_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file data/Schistosoma_mansoni/subsampled/ERR022883_2.fastq.gz <<< \n",
      "This is cutadapt 4.8 with Python 3.10.14\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC data/Schistosoma_mansoni/subsampled/ERR022883_2.fastq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 24.544 s (24.544 s/read; 2.44 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               1,000,000\n",
      "Reads with adapters:                   349,011 (34.9%)\n",
      "Reads written (passing filters):     1,000,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:    76,000,000 bp\n",
      "Quality-trimmed:               4,791,663 bp (6.3%)\n",
      "Total written (filtered):     70,747,578 bp (93.1%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 349011 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.7%\n",
      "  C: 24.3%\n",
      "  G: 15.0%\n",
      "  T: 25.7%\n",
      "  none/other: 0.3%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t268927\t250000.0\t0\t268927\n",
      "2\t58974\t62500.0\t0\t58974\n",
      "3\t15378\t15625.0\t0\t15378\n",
      "4\t4409\t3906.2\t0\t4409\n",
      "5\t880\t976.6\t0\t880\n",
      "6\t188\t244.1\t0\t188\n",
      "7\t38\t61.0\t0\t38\n",
      "8\t10\t15.3\t0\t10\n",
      "9\t54\t3.8\t0\t6 48\n",
      "10\t27\t1.0\t1\t7 20\n",
      "11\t15\t0.2\t1\t4 11\n",
      "12\t7\t0.1\t1\t4 3\n",
      "13\t1\t0.0\t1\t0 1\n",
      "14\t9\t0.0\t1\t7 2\n",
      "15\t2\t0.0\t1\t2\n",
      "16\t3\t0.0\t1\t3\n",
      "17\t5\t0.0\t1\t5\n",
      "18\t3\t0.0\t1\t3\n",
      "19\t7\t0.0\t1\t7\n",
      "20\t5\t0.0\t1\t5\n",
      "22\t6\t0.0\t1\t6\n",
      "23\t4\t0.0\t1\t3 1\n",
      "24\t7\t0.0\t1\t7\n",
      "26\t3\t0.0\t1\t3\n",
      "27\t3\t0.0\t1\t2 1\n",
      "28\t7\t0.0\t1\t7\n",
      "29\t2\t0.0\t1\t1 1\n",
      "30\t2\t0.0\t1\t2\n",
      "32\t1\t0.0\t1\t1\n",
      "33\t2\t0.0\t1\t1 1\n",
      "35\t1\t0.0\t1\t1\n",
      "36\t1\t0.0\t1\t0 1\n",
      "38\t1\t0.0\t1\t1\n",
      "39\t2\t0.0\t1\t2\n",
      "40\t1\t0.0\t1\t0 1\n",
      "41\t1\t0.0\t1\t1\n",
      "43\t1\t0.0\t1\t0 1\n",
      "44\t1\t0.0\t1\t1\n",
      "45\t1\t0.0\t1\t1\n",
      "46\t1\t0.0\t1\t1\n",
      "48\t1\t0.0\t1\t1\n",
      "49\t1\t0.0\t1\t1\n",
      "50\t1\t0.0\t1\t0 1\n",
      "52\t2\t0.0\t1\t0 2\n",
      "57\t6\t0.0\t1\t4 2\n",
      "59\t2\t0.0\t1\t1 1\n",
      "60\t1\t0.0\t1\t1\n",
      "61\t2\t0.0\t1\t2\n",
      "62\t2\t0.0\t1\t1 1\n",
      "73\t2\t0.0\t1\t0 2\n",
      "74\t1\t0.0\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: data/Schistosoma_mansoni/subsampled/ERR022883_2.fastq.gz\n",
      "=============================================\n",
      "1000000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files ERR022883_1_trimmed.fq.gz and ERR022883_2_trimmed.fq.gz\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>ERR022883_1_trimmed.fq.gz<<\tRENAMING TO:>>ERR022883_R1_trimmed.fq.gz<<\n",
      "ORIGINAL FILE 2: >>ERR022883_2_trimmed.fq.gz<<\tRENAMING TO:>>ERR022883_R2_trimmed.fq.gz<<\n",
      "file_1: ERR022883_R1_trimmed.fq.gz, file_2: ERR022883_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: ERR022883_R1_trimmed.fq.gz and ERR022883_R2_trimmed.fq.gz <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to ERR022883_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to ERR022883_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 1000000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 41224 (4.12%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data ERR022883_val_1.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022883_val_1.fq.gz\n",
      "Approx 5% complete for ERR022883_val_1.fq.gz\n",
      "Approx 10% complete for ERR022883_val_1.fq.gz\n",
      "Approx 15% complete for ERR022883_val_1.fq.gz\n",
      "Approx 20% complete for ERR022883_val_1.fq.gz\n",
      "Approx 25% complete for ERR022883_val_1.fq.gz\n",
      "Approx 30% complete for ERR022883_val_1.fq.gz\n",
      "Approx 35% complete for ERR022883_val_1.fq.gz\n",
      "Approx 40% complete for ERR022883_val_1.fq.gz\n",
      "Approx 45% complete for ERR022883_val_1.fq.gz\n",
      "Approx 50% complete for ERR022883_val_1.fq.gz\n",
      "Approx 55% complete for ERR022883_val_1.fq.gz\n",
      "Approx 60% complete for ERR022883_val_1.fq.gz\n",
      "Approx 65% complete for ERR022883_val_1.fq.gz\n",
      "Approx 70% complete for ERR022883_val_1.fq.gz\n",
      "Approx 75% complete for ERR022883_val_1.fq.gz\n",
      "Approx 80% complete for ERR022883_val_1.fq.gz\n",
      "Approx 85% complete for ERR022883_val_1.fq.gz\n",
      "Approx 90% complete for ERR022883_val_1.fq.gz\n",
      "Approx 95% complete for ERR022883_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022883_val_1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  >>> Now running FastQC on the validated data ERR022883_val_2.fq.gz<<<\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of ERR022883_val_2.fq.gz\n",
      "Approx 5% complete for ERR022883_val_2.fq.gz\n",
      "Approx 10% complete for ERR022883_val_2.fq.gz\n",
      "Approx 15% complete for ERR022883_val_2.fq.gz\n",
      "Approx 20% complete for ERR022883_val_2.fq.gz\n",
      "Approx 25% complete for ERR022883_val_2.fq.gz\n",
      "Approx 30% complete for ERR022883_val_2.fq.gz\n",
      "Approx 35% complete for ERR022883_val_2.fq.gz\n",
      "Approx 40% complete for ERR022883_val_2.fq.gz\n",
      "Approx 45% complete for ERR022883_val_2.fq.gz\n",
      "Approx 50% complete for ERR022883_val_2.fq.gz\n",
      "Approx 55% complete for ERR022883_val_2.fq.gz\n",
      "Approx 60% complete for ERR022883_val_2.fq.gz\n",
      "Approx 65% complete for ERR022883_val_2.fq.gz\n",
      "Approx 70% complete for ERR022883_val_2.fq.gz\n",
      "Approx 75% complete for ERR022883_val_2.fq.gz\n",
      "Approx 80% complete for ERR022883_val_2.fq.gz\n",
      "Approx 85% complete for ERR022883_val_2.fq.gz\n",
      "Approx 90% complete for ERR022883_val_2.fq.gz\n",
      "Approx 95% complete for ERR022883_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for ERR022883_val_2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Deleting both intermediate output files ERR022883_R1_trimmed.fq.gz and ERR022883_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for accession in $(cat data/Schistosoma_mansoni/list_ids.txt)\n",
    "do\n",
    "    trim_galore \\\n",
    "      data/Schistosoma_mansoni/subsampled/$accession*.fastq.gz \\\n",
    "      --paired \\\n",
    "      --output_dir analysis/Schistosoma_mansoni/qc/ \\\n",
    "      --basename $accession \\\n",
    "      --no_report_file \\\n",
    "      --fastqc\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98aacb60-050b-4538-81b6-69a7a5fb37f4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Question:\n",
    "\n",
    "6. What do the following elements of the code above mean (Use the TrimGalore documentation to find the information)?\n",
    "\n",
    "`--paired`\n",
    "\n",
    "`--fastqc`\n",
    "\n",
    "`--basename`\n",
    "\n",
    "7. Chose a read file and compare the FastQC report before trimming to the FastQC report after adaptor trimming. What are the improvements to the data quality after trimming? Are there any remaining warnings?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "687917ad-cb1d-42b5-b8c7-da0a26140127",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    \n",
    "Answers:\n",
    "\n",
    "6. `--paired` means that the reads are generated in pairs with a set gap size and so should be considered together with their pair when performing QC. In particular, this removes entire read pairs if at least one of the two sequences became shorter than a certain threshold. `--fastqc` automatically runs FastQC after trimming. `--basename` specifies the preferred prefix of output files.\n",
    "7. For read 1 of ERR022872, the trimmed report tells us that the average read quality does not drop off as much towards the end of the read. The per-base sequence content is still more variable at the starts of the reads, but no longer at the ends of the reads. The overrepresented primers and adaptors have now been removed.\n",
    "   The read length distribution is now worse as before all reads were 76bp long and now there is a long tail with (small numbers of) much shorter reads  with lengths between 20-75bp after trimming."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68412af0-8217-4b32-837a-6f16572e7b59",
   "metadata": {},
   "source": [
    "# Mapping to the reference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "717de691-38f5-4efe-b43b-9f97c0d96a84",
   "metadata": {},
   "source": [
    "Now that we are happy with the quality of the reads and have removed adapters, we can map our sequences to the genome. This allows us to identify which genes each read came from, and which genes were <i>expressed</i> in our samples in the form of transcripts. \n",
    "\n",
    "<figure>\n",
    "    <img src=\"https://www.annualreviews.org/docserver/ahah/fulltext/biodatasci/2/1/bd020139.f4_thmb.gif\">\n",
    "</figure>\n",
    "\n",
    "In some organisms including Plasmodium, expressed transcripts may be generated by splicing together non-contiguous exons from the genome (others such as Trypanosoma do not). To handle this, we can either use a splice-aware mapper to align reads across splice junctions, or we can map directly against panels of known transcripts. In this example we are going to use [STAR](https://academic.oup.com/bioinformatics/article/29/1/15/272537) to perform splice-aware alignment to the reference genome fasta. The manual can be found [here](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf). \n",
    "\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "\n",
    "This code will be run during the class, but uses more disk space than is available in Noteable. If you want to try it yourself, it will probably work on your personal computer. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "622a41af-986a-45c2-8683-b9120d42deeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --runThreadN 4 --runMode genomeGenerate --genomeDir analysis/Schistosoma_mansoni/star/ref --genomeFastaFiles data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.genomic.fa --sjdbGTFfile data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.annotations.gtf --sjdbOverhang 75 --genomeSAindexNbases 11\n",
      "\tSTAR version: 2.7.11b   compiled:  :/Users/distiller/project/STARcompile/source\n",
      "Aug 16 15:45:17 ..... started STAR run\n",
      "Aug 16 15:45:17 ... starting to generate Genome files\n",
      "Aug 16 15:45:29 ..... processing annotations GTF\n",
      "Aug 16 15:45:31 ... starting to sort Suffix Array. This may take a long time...\n",
      "Aug 16 15:45:33 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Aug 16 15:51:21 ... loading chunks from disk, packing SA...\n",
      "Aug 16 15:51:34 ... finished generating suffix array\n",
      "Aug 16 15:51:34 ... generating Suffix Array index\n",
      "Aug 16 15:51:43 ... completed Suffix Array index\n",
      "Aug 16 15:51:43 ..... inserting junctions into the genome indices\n",
      "Aug 16 15:52:25 ... writing Genome to disk ...\n",
      "Aug 16 15:52:26 ... writing Suffix Array to disk ...\n",
      "Aug 16 15:52:27 ... writing SAindex to disk\n",
      "Aug 16 15:52:28 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir -p analysis/Schistosoma_mansoni/star/ref\n",
    "\n",
    "# unzip the reference files\n",
    "#gunzip data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.annotations.gtf.gz \n",
    "#gunzip data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.genomic.fa.gz\n",
    "\n",
    "# first we need to index the reference\n",
    "mkdir -p analysis/Schistosoma_mansoni/star/ref\n",
    "  STAR --runThreadN 4 \\\n",
    "    --runMode genomeGenerate \\\n",
    "    --genomeDir analysis/Schistosoma_mansoni/star/ref \\\n",
    "    --genomeFastaFiles data/Schistosoma_mansoni/reference/*genomic.fa \\\n",
    "    --sjdbGTFfile data/Schistosoma_mansoni/reference/*.gtf \\\n",
    "    --sjdbOverhang 75 \\\n",
    "    --genomeSAindexNbases 11"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4305dd85-dfda-4e89-976e-b2b96b470708",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "In the code above:\n",
    "\n",
    "`--runMode genomeGenerate` - directs STAR to run genome indexing\n",
    "\n",
    "`--genomeDir /path/to/genomeDir` - specifies where to store the index\n",
    "\n",
    "`--genomeFastaFiles /path/to/genome/fasta` - provides the reference genome\n",
    "\n",
    "`--sjdbGTFfile /path/to/annotations.gtf` - provides the coordinates of splice junctions in the reference genome\n",
    "\n",
    "`--sjdbOverhang ReadLength-1` - this specifies the length of sequences to include in the splice junctions database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c718a0d-d71f-4bff-9283-f9988cb0cbd8",
   "metadata": {},
   "source": [
    "We can now use this index to align each pair of readfiles against the reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "44444240-2723-4f7c-89db-133f0ff4b4c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --genomeDir analysis/Schistosoma_mansoni/star/ref --runThreadN 4 --readFilesIn /dev/fd/63 --outFileNamePrefix analysis/Schistosoma_mansoni/star/ERR022872/ERR022872 --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 4000000 --outSAMattributes Standard --quantMode TranscriptomeSAM GeneCounts\n",
      "\tSTAR version: 2.7.11b   compiled:  :/Users/distiller/project/STARcompile/source\n",
      "Aug 16 15:52:34 ..... started STAR run\n",
      "Aug 16 15:52:34 ..... loading genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Transcriptome.cpp:18:Transcriptome: exiting because of *INPUT FILE* error: could not open input file /geneInfo.tab\n",
      "Solution: check that the file exists and you have read permission for this file\n",
      "          SOLUTION: utilize --sjdbGTFfile /path/to/annotations.gtf option at the genome generation step or mapping step\n",
      "\n",
      "Aug 16 15:52:39 ...... FATAL ERROR, exiting\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --genomeDir analysis/Schistosoma_mansoni/star/ref --runThreadN 4 --readFilesIn /dev/fd/63 --outFileNamePrefix analysis/Schistosoma_mansoni/star/ERR022873/ERR022873 --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 4000000 --outSAMattributes Standard --quantMode TranscriptomeSAM GeneCounts\n",
      "\tSTAR version: 2.7.11b   compiled:  :/Users/distiller/project/STARcompile/source\n",
      "Aug 16 15:52:39 ..... started STAR run\n",
      "Aug 16 15:52:39 ..... loading genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Transcriptome.cpp:18:Transcriptome: exiting because of *INPUT FILE* error: could not open input file /geneInfo.tab\n",
      "Solution: check that the file exists and you have read permission for this file\n",
      "          SOLUTION: utilize --sjdbGTFfile /path/to/annotations.gtf option at the genome generation step or mapping step\n",
      "\n",
      "Aug 16 15:52:42 ...... FATAL ERROR, exiting\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --genomeDir analysis/Schistosoma_mansoni/star/ref --runThreadN 4 --readFilesIn /dev/fd/63 --outFileNamePrefix analysis/Schistosoma_mansoni/star/ERR022874/ERR022874 --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 4000000 --outSAMattributes Standard --quantMode TranscriptomeSAM GeneCounts\n",
      "\tSTAR version: 2.7.11b   compiled:  :/Users/distiller/project/STARcompile/source\n",
      "Aug 16 15:52:42 ..... started STAR run\n",
      "Aug 16 15:52:42 ..... loading genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Transcriptome.cpp:18:Transcriptome: exiting because of *INPUT FILE* error: could not open input file /geneInfo.tab\n",
      "Solution: check that the file exists and you have read permission for this file\n",
      "          SOLUTION: utilize --sjdbGTFfile /path/to/annotations.gtf option at the genome generation step or mapping step\n",
      "\n",
      "Aug 16 15:52:46 ...... FATAL ERROR, exiting\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tSTAR --genomeDir analysis/Schistosoma_mansoni/star/ref --runThreadN 4 --readFilesIn /dev/fd/63 --outFileNamePrefix analysis/Schistosoma_mansoni/star/ERR022875/ERR022875 --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 4000000 --outSAMattributes Standard --quantMode TranscriptomeSAM GeneCounts\n",
      "\tSTAR version: 2.7.11b   compiled:  :/Users/distiller/project/STARcompile/source\n",
      "Aug 16 15:52:46 ..... started STAR run\n",
      "Aug 16 15:52:46 ..... loading genome\n",
      "Error while terminating subprocess (pid=98231): \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for accession in $(cat data/Schistosoma_mansoni/list_ids.txt)\n",
    "do\n",
    "    mkdir -p analysis/Schistosoma_mansoni/star/$accession\n",
    "    \n",
    "    STAR \\\n",
    "      --genomeDir analysis/Schistosoma_mansoni/star/ref \\\n",
    "      --runThreadN 4 \\\n",
    "      --readFilesIn <(gunzip -c analysis/Schistosoma_mansoni/qc/$accession*.fq.gz) \\\n",
    "      --outFileNamePrefix analysis/Schistosoma_mansoni/star/$accession/$accession \\\n",
    "      --outSAMtype BAM SortedByCoordinate \\\n",
    "      --limitBAMsortRAM 4000000 \\\n",
    "      --outSAMattributes Standard \\\n",
    "      --quantMode TranscriptomeSAM GeneCounts\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "311f53ec-2b12-4c9c-99a2-77b72adf9261",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "In the code above:\n",
    "\n",
    "`<(gunzip -c reads.fq.gz)` - this uncompresses the read sequences to input to STAR which does not support compressed files\n",
    "\n",
    "`--outSAMtype BAM SortedByCoordinate` - sort and compress the output\n",
    "\n",
    "`--outSAMattributes Standard` - include some standard count information in the output file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6d8309d-77f2-4d24-a7e3-fdd604d8935a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Question:\n",
    "\n",
    "8. Using the STAR manual, what outputs are generated using the flags `--quantMode TranscriptomeSAM GeneCounts`?\n",
    "\n",
    "9. Find a python library which can load a SAM file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1121e8",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "\n",
    "Answers:\n",
    "\n",
    "8. (p18) With --quantMode TranscriptomeSAM option STAR will output alignments translated into transcript coordinates in the Aligned.toTranscriptome.out.bam file (in addition to alignments in genomic coordinates in Aligned.*.sam/bam files).\n",
    "   \n",
    "    A SAM file usually records the position and how each read lines up against the reference genome. This instead writes down how the reads map against the transcripts in the reference genome. A BAM is a compressed SAM file.\n",
    "\n",
    "   (p18) STAR outputs read counts per gene into ReadsPerGene.out.tab file with 4 columns which correspond to different strandedness options:\n",
    "\n",
    "    column 1: gene ID\n",
    "    column 2: counts for unstranded RNA-seq\n",
    "    column 3: counts for the 1st read strand aligned with RNA (htseq-count option -s yes\n",
    "    column 4: counts for the 2nd read strand aligned with RNA (htseq-count option -s reverse)\n",
    "\n",
    "    With --quantMode TranscriptomeSAM GeneCounts, it will output both the Aligned.toTranscriptome.out.bam and ReadsPerGene.out.tab outputs.\n",
    "\n",
    "9. [pysam](https://pysam.readthedocs.io/en/latest/api.html) is an example of a python library which can load a SAM file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65effdfc-c29c-448a-93a2-d9c28b794db4",
   "metadata": {},
   "source": [
    "We will be running STAR on each of the full datasets and will make the mapped read files available for the next class."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7755a167-e30e-4b20-9bfd-2c4bd87f241c",
   "metadata": {},
   "source": [
    "# Extension"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a89ef8c2-1420-4786-bb1b-866c45c0c35a",
   "metadata": {},
   "source": [
    "There also exist several methods for transcript abundance quantification using `pseudo-alignment`. These methods don't fully line up reads against the reference genome or transcript sequences, but instead count the occurance of substrings of these transcripts and use this to estimate transcript abundances. \n",
    "\n",
    "One example of this method is [Kallisto](https://www.nature.com/articles/nbt.3519)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cdee8f9-c553-4540-aaf7-7ca37412e08a",
   "metadata": {},
   "outputs": [],
   "source": [
    "! conda install --yes --quiet bioconda::kallisto=0.48"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24cababa-4f10-438e-8805-37b5c842b4c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[build] loading fasta file data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.mRNA_transcripts.fa.gz\n",
      "[build] k-mer length: 31\n",
      "[build] warning: clipped off poly-A tail (longer than 10)\n",
      "        from 1 target sequences\n",
      "[build] counting k-mers ... done.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p analysis/Schistosoma_mansoni/kallisto/\n",
    "\n",
    "kallisto index --index=analysis/Schistosoma_mansoni/kallisto/smansoni data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.mRNA_transcripts.fa.gz\n",
    "\n",
    "for accession in $(cat data/Schistosoma_mansoni/subsampled/list_ids.txt)\n",
    "do\n",
    "    kallisto quant --threads=2 \\\n",
    "      --index=analysis/Schistosoma_mansoni/kallisto/smansoni \\\n",
    "      --output-dir=analysis/Schistosoma_mansoni/kallisto \\\n",
    "      --gtf=data/Schistosoma_mansoni/reference/schistosoma_mansoni.PRJEA36577.WBPS19.annotations.gtf.gz \\\n",
    "      analysis/Schistosoma_mansoni/qc/\"$accession\"_1.trimmed.fastq.gz analysis/Schistosoma_mansoni/qc/\"$accession\"_2.trimmed.fastq.gz\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1500b255-99d2-4c6e-b507-5310918933e7",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "Question:\n",
    "\n",
    "10. Give 2 differences between the methods used above by STAR and Kallisto to quantify transcript abundances"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f78c59ba-3069-46c7-9966-4dca45f5f7d4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "\n",
    "Answers:\n",
    "\n",
    "10. - STAR performs splice aware alignment to the reference genome. Kallisto instead takes as input a file of transcripts so does not need to be aware of splicing.\n",
    "    - Kallisto uses approximate pseudo-alignment rather than full alignment so will not give you a SAM/BAM file output\n",
    "    - Kallisto is much quicker and less memory intensive than STAR.\n",
    "    - Kallisto gives transcript level expression, whilst STAR + counting only gives gene-level (transcript level can be found by following on with RSEM)\n",
    "    - Kallisto can't quantify genes or splice-variants that are not in the input transcript file. Differences between the real transcriptome and your annotation of the transcriptome will reduce accuracy. This means it cannot be used for finding new genes, transcripts or splice forms, or for any analysis other than quantification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d13a1ca3-bf6c-48d6-99e9-1087836216b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
