{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9da68c44-01ce-44f6-9de2-99c1c1b9a948",
   "metadata": {},
   "source": [
    "# What to expect\n",
    "\n",
    "In notebook 3A we ran a differential gene expression analysis on the example dataset Schistosoma mansoni and used visualisation techniques to view the most significant genes. In this notebook we will apply the same methods to our choice of dataset in this notebook. We will go on to explore the GO terms and pathways associated with them using online resources.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9094dea0-a9d0-4c6d-b766-770ef1ba4442",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pylab as plt\n",
    "\n",
    "#Install PyDESeq2 and import required classes\n",
    "! pip install --quiet pydeseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15597885-bf17-4406-a4b9-b25aafb3c36e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.79 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.23 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.30 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.93 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# load in the counts and metadata again\n",
    "prefix = \"Trypanosoma\"\n",
    "counts = pd.read_csv(f\"analysis/{prefix}/star/ReadsPerGene.csv\", index_col=0).T\n",
    "metadata = pd.read_csv(f\"data/{prefix}/metadata.csv\", index_col=0)\n",
    "counts.fillna(0)\n",
    "\n",
    "# restrict to the 2 stages we want to compare\n",
    "counts_s = counts[metadata[\"organism\"].isin([\"Trypanosoma brucei brucei\"])]\n",
    "counts_s = counts_s.loc[:, (counts_s != 0).any(axis=0)]\n",
    "metadata_s = metadata[metadata[\"organism\"].isin([\"Trypanosoma brucei brucei\"])]\n",
    "\n",
    "# create deseq2 dataset object\n",
    "dds = DeseqDataSet(\n",
    "    counts=counts_s,\n",
    "    metadata=metadata_s,\n",
    "    design_factors=\"condition\",  # compare samples based on the developmental \"stage\"\n",
    "    refit_cooks=True\n",
    ")\n",
    "\n",
    "# Run DeSeq2\n",
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bddaaecf-552e-4028-bfc0-b52d702d942f",
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir -p \"analysis/Trypanosoma/de\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d476b5e7-820c-46a2-a69e-db825dbe1125",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition peak vs ascending\n",
      "                                  baseMean  log2FoldChange     lfcSE  \\\n",
      "gene                                                                   \n",
      "Tb04.24M18.150                  197.292416        0.217076  0.190053   \n",
      "Tb04.3I12.100                   218.408392        0.124410  0.171674   \n",
      "Tb05.30F7.410                    99.278007       -1.824686  0.655682   \n",
      "Tb05.5K5.100                     16.771503        0.644534  0.565982   \n",
      "Tb05.5K5.110                    329.781049       -0.045490  0.139694   \n",
      "...                                    ...             ...       ...   \n",
      "Tb927_10_v4.snoRNA.0063:snoRNA    3.790393        0.122767  1.088272   \n",
      "Tb927_10_v4.snoRNA.0064:snoRNA    0.153151        0.709747  4.425350   \n",
      "Tb927_10_v4.snoRNA.0073:snoRNA    0.183423       -1.213827  4.425356   \n",
      "Tb927_10_v4.snoRNA.0078:snoRNA  133.952072        0.242593  0.196860   \n",
      "tmp.1.100                        48.341400        0.781146  0.795903   \n",
      "\n",
      "                                    stat    pvalue      padj  \n",
      "gene                                                          \n",
      "Tb04.24M18.150                  1.142188  0.253376  0.371602  \n",
      "Tb04.3I12.100                   0.724689  0.468643  0.597742  \n",
      "Tb05.30F7.410                  -2.782883  0.005388  0.013279  \n",
      "Tb05.5K5.100                    1.138789  0.254791  0.373442  \n",
      "Tb05.5K5.110                   -0.325640  0.744697  0.824831  \n",
      "...                                  ...       ...       ...  \n",
      "Tb927_10_v4.snoRNA.0063:snoRNA  0.112809  0.910182  0.941086  \n",
      "Tb927_10_v4.snoRNA.0064:snoRNA  0.160382  0.872580       NaN  \n",
      "Tb927_10_v4.snoRNA.0073:snoRNA -0.274289  0.783862       NaN  \n",
      "Tb927_10_v4.snoRNA.0078:snoRNA  1.232314  0.217832  0.330538  \n",
      "tmp.1.100                       0.981458  0.326367  0.453923  \n",
      "\n",
      "[9859 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.52 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Summarize results\n",
    "stat_res=DeseqStats(dds)\n",
    "stat_res.summary()\n",
    "res = stat_res.results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f44b85f7-7f6e-43b5-a60b-8ea468311ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res.to_csv(f\"analysis/{prefix}/de/slender_vs_stumpy_tbrucei.full.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fd192c3a-ea7a-4b9e-98de-344ccea5cc42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter results with baseMean<10 so that gene expressions close to zero don't skew results\n",
    "res=res[res.baseMean>=10]\n",
    "\n",
    "# Filter by padj<=0.05\n",
    "res=res[res.padj<=0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6de915e6-3998-4746-97f9-7143152edc8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tb05.30F7.410</th>\n",
       "      <td>99.278007</td>\n",
       "      <td>-1.824686</td>\n",
       "      <td>0.655682</td>\n",
       "      <td>-2.782883</td>\n",
       "      <td>5.387819e-03</td>\n",
       "      <td>1.327922e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.130</th>\n",
       "      <td>704.934698</td>\n",
       "      <td>-0.628927</td>\n",
       "      <td>0.130836</td>\n",
       "      <td>-4.806982</td>\n",
       "      <td>1.532258e-06</td>\n",
       "      <td>7.231510e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.150</th>\n",
       "      <td>356.075307</td>\n",
       "      <td>-0.569119</td>\n",
       "      <td>0.144114</td>\n",
       "      <td>-3.949087</td>\n",
       "      <td>7.844998e-05</td>\n",
       "      <td>2.813295e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.210</th>\n",
       "      <td>652.756184</td>\n",
       "      <td>-0.546718</td>\n",
       "      <td>0.123264</td>\n",
       "      <td>-4.435348</td>\n",
       "      <td>9.192383e-06</td>\n",
       "      <td>3.866796e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.270</th>\n",
       "      <td>559.894678</td>\n",
       "      <td>2.175802</td>\n",
       "      <td>0.136297</td>\n",
       "      <td>15.963716</td>\n",
       "      <td>2.287011e-57</td>\n",
       "      <td>2.794969e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9820</th>\n",
       "      <td>1610.596099</td>\n",
       "      <td>-1.227299</td>\n",
       "      <td>0.095661</td>\n",
       "      <td>-12.829598</td>\n",
       "      <td>1.119436e-37</td>\n",
       "      <td>6.301404e-36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9840</th>\n",
       "      <td>1142.358426</td>\n",
       "      <td>-0.682270</td>\n",
       "      <td>0.133546</td>\n",
       "      <td>-5.108861</td>\n",
       "      <td>3.241061e-07</td>\n",
       "      <td>1.680792e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9870</th>\n",
       "      <td>1282.657306</td>\n",
       "      <td>-0.712261</td>\n",
       "      <td>0.095751</td>\n",
       "      <td>-7.438662</td>\n",
       "      <td>1.017103e-13</td>\n",
       "      <td>1.143687e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9940</th>\n",
       "      <td>2501.918060</td>\n",
       "      <td>-1.650387</td>\n",
       "      <td>0.091116</td>\n",
       "      <td>-18.113086</td>\n",
       "      <td>2.512616e-73</td>\n",
       "      <td>5.427251e-71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927_06_v4.snoRNA.0004</th>\n",
       "      <td>37.980511</td>\n",
       "      <td>1.052091</td>\n",
       "      <td>0.362500</td>\n",
       "      <td>2.902319</td>\n",
       "      <td>3.704116e-03</td>\n",
       "      <td>9.472420e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2280 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            baseMean  log2FoldChange     lfcSE       stat  \\\n",
       "gene                                                                        \n",
       "Tb05.30F7.410              99.278007       -1.824686  0.655682  -2.782883   \n",
       "Tb05.5K5.130              704.934698       -0.628927  0.130836  -4.806982   \n",
       "Tb05.5K5.150              356.075307       -0.569119  0.144114  -3.949087   \n",
       "Tb05.5K5.210              652.756184       -0.546718  0.123264  -4.435348   \n",
       "Tb05.5K5.270              559.894678        2.175802  0.136297  15.963716   \n",
       "...                              ...             ...       ...        ...   \n",
       "Tb927.9.9820             1610.596099       -1.227299  0.095661 -12.829598   \n",
       "Tb927.9.9840             1142.358426       -0.682270  0.133546  -5.108861   \n",
       "Tb927.9.9870             1282.657306       -0.712261  0.095751  -7.438662   \n",
       "Tb927.9.9940             2501.918060       -1.650387  0.091116 -18.113086   \n",
       "Tb927_06_v4.snoRNA.0004    37.980511        1.052091  0.362500   2.902319   \n",
       "\n",
       "                               pvalue          padj  \n",
       "gene                                                 \n",
       "Tb05.30F7.410            5.387819e-03  1.327922e-02  \n",
       "Tb05.5K5.130             1.532258e-06  7.231510e-06  \n",
       "Tb05.5K5.150             7.844998e-05  2.813295e-04  \n",
       "Tb05.5K5.210             9.192383e-06  3.866796e-05  \n",
       "Tb05.5K5.270             2.287011e-57  2.794969e-55  \n",
       "...                               ...           ...  \n",
       "Tb927.9.9820             1.119436e-37  6.301404e-36  \n",
       "Tb927.9.9840             3.241061e-07  1.680792e-06  \n",
       "Tb927.9.9870             1.017103e-13  1.143687e-12  \n",
       "Tb927.9.9940             2.512616e-73  5.427251e-71  \n",
       "Tb927_06_v4.snoRNA.0004  3.704116e-03  9.472420e-03  \n",
       "\n",
       "[2280 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get list of only genes that have a fold change FC > 1.5 or FC < 0.7 \n",
    "# You can play with the exact thresholds here, these are just a guide to filter the lists\n",
    "sigs=res[abs(res.log2FoldChange)>0.5]\n",
    "sigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ee74ac7f-e05a-4573-85f8-f4aab2adae6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.270</th>\n",
       "      <td>559.894678</td>\n",
       "      <td>2.175802</td>\n",
       "      <td>0.136297</td>\n",
       "      <td>15.963716</td>\n",
       "      <td>2.287011e-57</td>\n",
       "      <td>2.794969e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.70</th>\n",
       "      <td>711.309715</td>\n",
       "      <td>0.849127</td>\n",
       "      <td>0.114743</td>\n",
       "      <td>7.400239</td>\n",
       "      <td>1.359397e-13</td>\n",
       "      <td>1.501318e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb08.27P2.60</th>\n",
       "      <td>34.502243</td>\n",
       "      <td>1.439498</td>\n",
       "      <td>0.488999</td>\n",
       "      <td>2.943764</td>\n",
       "      <td>3.242468e-03</td>\n",
       "      <td>8.454813e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb08.27P2.90</th>\n",
       "      <td>45.643900</td>\n",
       "      <td>1.285696</td>\n",
       "      <td>0.365416</td>\n",
       "      <td>3.518446</td>\n",
       "      <td>4.340823e-04</td>\n",
       "      <td>1.362080e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb09.v4.0150</th>\n",
       "      <td>900.439292</td>\n",
       "      <td>0.607229</td>\n",
       "      <td>0.126314</td>\n",
       "      <td>4.807313</td>\n",
       "      <td>1.529727e-06</td>\n",
       "      <td>7.223235e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9300</th>\n",
       "      <td>2943.061947</td>\n",
       "      <td>1.205118</td>\n",
       "      <td>0.080061</td>\n",
       "      <td>15.052421</td>\n",
       "      <td>3.328281e-51</td>\n",
       "      <td>3.323986e-49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9410</th>\n",
       "      <td>1057.658097</td>\n",
       "      <td>1.239115</td>\n",
       "      <td>0.099359</td>\n",
       "      <td>12.471079</td>\n",
       "      <td>1.073619e-35</td>\n",
       "      <td>5.478995e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9600</th>\n",
       "      <td>2503.524120</td>\n",
       "      <td>0.536295</td>\n",
       "      <td>0.081707</td>\n",
       "      <td>6.563676</td>\n",
       "      <td>5.249737e-11</td>\n",
       "      <td>4.452927e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9620</th>\n",
       "      <td>2059.820109</td>\n",
       "      <td>0.584625</td>\n",
       "      <td>0.095175</td>\n",
       "      <td>6.142626</td>\n",
       "      <td>8.116830e-10</td>\n",
       "      <td>6.013667e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927_06_v4.snoRNA.0004</th>\n",
       "      <td>37.980511</td>\n",
       "      <td>1.052091</td>\n",
       "      <td>0.362500</td>\n",
       "      <td>2.902319</td>\n",
       "      <td>3.704116e-03</td>\n",
       "      <td>9.472420e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1104 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            baseMean  log2FoldChange     lfcSE       stat  \\\n",
       "gene                                                                        \n",
       "Tb05.5K5.270              559.894678        2.175802  0.136297  15.963716   \n",
       "Tb05.5K5.70               711.309715        0.849127  0.114743   7.400239   \n",
       "Tb08.27P2.60               34.502243        1.439498  0.488999   2.943764   \n",
       "Tb08.27P2.90               45.643900        1.285696  0.365416   3.518446   \n",
       "Tb09.v4.0150              900.439292        0.607229  0.126314   4.807313   \n",
       "...                              ...             ...       ...        ...   \n",
       "Tb927.9.9300             2943.061947        1.205118  0.080061  15.052421   \n",
       "Tb927.9.9410             1057.658097        1.239115  0.099359  12.471079   \n",
       "Tb927.9.9600             2503.524120        0.536295  0.081707   6.563676   \n",
       "Tb927.9.9620             2059.820109        0.584625  0.095175   6.142626   \n",
       "Tb927_06_v4.snoRNA.0004    37.980511        1.052091  0.362500   2.902319   \n",
       "\n",
       "                               pvalue          padj  \n",
       "gene                                                 \n",
       "Tb05.5K5.270             2.287011e-57  2.794969e-55  \n",
       "Tb05.5K5.70              1.359397e-13  1.501318e-12  \n",
       "Tb08.27P2.60             3.242468e-03  8.454813e-03  \n",
       "Tb08.27P2.90             4.340823e-04  1.362080e-03  \n",
       "Tb09.v4.0150             1.529727e-06  7.223235e-06  \n",
       "...                               ...           ...  \n",
       "Tb927.9.9300             3.328281e-51  3.323986e-49  \n",
       "Tb927.9.9410             1.073619e-35  5.478995e-34  \n",
       "Tb927.9.9600             5.249737e-11  4.452927e-10  \n",
       "Tb927.9.9620             8.116830e-10  6.013667e-09  \n",
       "Tb927_06_v4.snoRNA.0004  3.704116e-03  9.472420e-03  \n",
       "\n",
       "[1104 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save the up and down regulated genes separately\n",
    "up = sigs[sigs[\"log2FoldChange\"] > 0]\n",
    "up.to_csv(f\"analysis/{prefix}/de/slender_vs_stumpy_tbrucei.up.csv\")\n",
    "up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0b65dc2e-0b54-4619-a731-a15dd247c2b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tb05.30F7.410</th>\n",
       "      <td>99.278007</td>\n",
       "      <td>-1.824686</td>\n",
       "      <td>0.655682</td>\n",
       "      <td>-2.782883</td>\n",
       "      <td>5.387819e-03</td>\n",
       "      <td>1.327922e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.130</th>\n",
       "      <td>704.934698</td>\n",
       "      <td>-0.628927</td>\n",
       "      <td>0.130836</td>\n",
       "      <td>-4.806982</td>\n",
       "      <td>1.532258e-06</td>\n",
       "      <td>7.231510e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.150</th>\n",
       "      <td>356.075307</td>\n",
       "      <td>-0.569119</td>\n",
       "      <td>0.144114</td>\n",
       "      <td>-3.949087</td>\n",
       "      <td>7.844998e-05</td>\n",
       "      <td>2.813295e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.210</th>\n",
       "      <td>652.756184</td>\n",
       "      <td>-0.546718</td>\n",
       "      <td>0.123264</td>\n",
       "      <td>-4.435348</td>\n",
       "      <td>9.192383e-06</td>\n",
       "      <td>3.866796e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb05.5K5.280</th>\n",
       "      <td>501.262274</td>\n",
       "      <td>-1.617947</td>\n",
       "      <td>0.146860</td>\n",
       "      <td>-11.016920</td>\n",
       "      <td>3.167116e-28</td>\n",
       "      <td>1.081477e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9810</th>\n",
       "      <td>1413.470872</td>\n",
       "      <td>-1.451264</td>\n",
       "      <td>0.100885</td>\n",
       "      <td>-14.385271</td>\n",
       "      <td>6.402719e-47</td>\n",
       "      <td>5.773636e-45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9820</th>\n",
       "      <td>1610.596099</td>\n",
       "      <td>-1.227299</td>\n",
       "      <td>0.095661</td>\n",
       "      <td>-12.829598</td>\n",
       "      <td>1.119436e-37</td>\n",
       "      <td>6.301404e-36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9840</th>\n",
       "      <td>1142.358426</td>\n",
       "      <td>-0.682270</td>\n",
       "      <td>0.133546</td>\n",
       "      <td>-5.108861</td>\n",
       "      <td>3.241061e-07</td>\n",
       "      <td>1.680792e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9870</th>\n",
       "      <td>1282.657306</td>\n",
       "      <td>-0.712261</td>\n",
       "      <td>0.095751</td>\n",
       "      <td>-7.438662</td>\n",
       "      <td>1.017103e-13</td>\n",
       "      <td>1.143687e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.9.9940</th>\n",
       "      <td>2501.918060</td>\n",
       "      <td>-1.650387</td>\n",
       "      <td>0.091116</td>\n",
       "      <td>-18.113086</td>\n",
       "      <td>2.512616e-73</td>\n",
       "      <td>5.427251e-71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1176 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
       "gene                                                                            \n",
       "Tb05.30F7.410    99.278007       -1.824686  0.655682  -2.782883  5.387819e-03   \n",
       "Tb05.5K5.130    704.934698       -0.628927  0.130836  -4.806982  1.532258e-06   \n",
       "Tb05.5K5.150    356.075307       -0.569119  0.144114  -3.949087  7.844998e-05   \n",
       "Tb05.5K5.210    652.756184       -0.546718  0.123264  -4.435348  9.192383e-06   \n",
       "Tb05.5K5.280    501.262274       -1.617947  0.146860 -11.016920  3.167116e-28   \n",
       "...                    ...             ...       ...        ...           ...   \n",
       "Tb927.9.9810   1413.470872       -1.451264  0.100885 -14.385271  6.402719e-47   \n",
       "Tb927.9.9820   1610.596099       -1.227299  0.095661 -12.829598  1.119436e-37   \n",
       "Tb927.9.9840   1142.358426       -0.682270  0.133546  -5.108861  3.241061e-07   \n",
       "Tb927.9.9870   1282.657306       -0.712261  0.095751  -7.438662  1.017103e-13   \n",
       "Tb927.9.9940   2501.918060       -1.650387  0.091116 -18.113086  2.512616e-73   \n",
       "\n",
       "                       padj  \n",
       "gene                         \n",
       "Tb05.30F7.410  1.327922e-02  \n",
       "Tb05.5K5.130   7.231510e-06  \n",
       "Tb05.5K5.150   2.813295e-04  \n",
       "Tb05.5K5.210   3.866796e-05  \n",
       "Tb05.5K5.280   1.081477e-26  \n",
       "...                     ...  \n",
       "Tb927.9.9810   5.773636e-45  \n",
       "Tb927.9.9820   6.301404e-36  \n",
       "Tb927.9.9840   1.680792e-06  \n",
       "Tb927.9.9870   1.143687e-12  \n",
       "Tb927.9.9940   5.427251e-71  \n",
       "\n",
       "[1176 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "down = sigs[sigs[\"log2FoldChange\"] < 0]\n",
    "up.to_csv(f\"analysis/{prefix}/de/slender_vs_stumpy_tbrucei.down.csv\")\n",
    "down"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04c3e6f-b526-4004-847f-7a6513e2aaaf",
   "metadata": {},
   "source": [
    "# GO Analysis and Metabolic pathways analysis\n",
    "\n",
    "We will now use these results for GO and metabolic pathway analysis using the TriTrypDB and PlasmoDB websites. \n",
    "\n",
    "You are welcome to perform either GO or metabolic pathway analysis on your chosen dataset, or both if you have time.\n",
    "\n",
    "## GO Analysis\n",
    "For this you will need either a text list of gene IDs which can be pasted into the website field, or by uploading a text file. The DBs return a .csv file that includes the names (other annotations, such as cellular locations or GO terms, can be included, if desired):\n",
    "\n",
    "https://tritrypdb.org/tritrypdb/app/search/transcript/GeneByLocusTag\n",
    "https://plasmodb.org/plasmo/app/search/transcript/GeneByLocusTag\n",
    "\n",
    "Note that these functions have a glitch: some genes get duplicated in the returned list (this seems to be random). In Excel it is straightforward to identify such duplications and remove them.\n",
    "\n",
    "## Metabolic Pathways\n",
    "Submit the filtered .csv files for ‘up’ and ‘down’ separately to the respective DB tools for metabolic pathway analysis:\n",
    "\n",
    "https://tritrypdb.org/tritrypdb/app/search/pathway/PathwaysByGeneList\n",
    "https://plasmodb.org/plasmo/app/search/pathway/PathwaysByGeneList\n",
    "\n",
    "Note that the same glitch described above affects these searches, i.e. a bunch of duplicates need to be subsequently removed.\n",
    "The search function offers a few different ‘Pathway Sources’ (KEGG, LeishCyc, MetaCyc and TrypanoCyc for T. brucei). I’ve included all of them, the hit list can be subsequently filtered by source, if desired.\n",
    "Pathways with hits can be downloaded as .csv file, and the results can be customized in terms of the columns that should be included, such as ‘Total Pathway Enzymes’, ‘Unique Gene Counts’ (i.e. hits in that pathway), EC numbers etc.\n",
    "I’ve included a separate column with ‘% Hits’ (Unique Gene counts * 100 / Total Pathway Enzymes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e2d44cd7-e514-4154-a022-92f7be21c739",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"analysis/{prefix}/de/slender_vs_stumpy_tbrucei.genes.txt\", \"w\") as f:\n",
    "    for gene in sigs.index:\n",
    "        f.write(f\"{gene}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49cb0602-46eb-4c60-a2d8-c74b4ea46d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"analysis/{prefix}/de/slender_vs_stumpy_tbrucei.up_genes.txt\", \"w\") as f:\n",
    "    for gene in up.index:\n",
    "        f.write(f\"{gene}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9e34c9b3-b3ca-4422-9858-6dbb82497d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"analysis/{prefix}/de/slender_vs_stumpy_tbrucei.down_genes.txt\", \"w\") as f:\n",
    "    for gene in down.index:\n",
    "        f.write(f\"{gene}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "264b6a51-a5e3-425d-83b6-a5b798fab17f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
